{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6a14ecbd-e4f4-4b8a-921c-398bce3e1653",
   "metadata": {},
   "source": [
    "# Kallisto quantification\n",
    "\n",
    "The purpose of this notebook is to develop a framework where I can automate the Kallisto quantification. \n",
    "\n",
    "The ground rules that I plan on setting:\n",
    "- I will use Kallisto indices which are already prepared\n",
    "- I am using human data as a test case, however I really think I can easily generalise this to other species\n",
    "- I will specify the directory where the FASTQ files are prepared."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f59a5fa6-5407-48d8-af32-f8bb2dc0f87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "from pydantic import BaseModel\n",
    "from typing import List, Dict\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f1ef49e-43a5-441e-b49d-434c8892a576",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Water is composed of two hydrogen atoms covalently bonded to one oxygen atom, represented by the chemical formula Hâ‚‚O. This molecular structure results in polar molecules, contributing to water's unique properties.\n"
     ]
    }
   ],
   "source": [
    "# Notebookly test\n",
    "\n",
    "load_dotenv('../../.env')\n",
    "\n",
    "openai_api_key = os.getenv('OPENAI_API_KEY')\n",
    "\n",
    "# Test OpenAI API...\n",
    "\n",
    "client = OpenAI(\n",
    "  api_key=openai_api_key,  # this is also the default, it can be omitted\n",
    ")\n",
    "\n",
    "chat_completion = client.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": \"In two sentences or fewer, can you describe the molecular composition of water?\",\n",
    "        }\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    ")\n",
    "\n",
    "result = chat_completion.choices[0].message.content\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e471760d-184c-4de7-9b3a-b0ed65762ca5",
   "metadata": {},
   "source": [
    "I appreciate the formatting."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caa25f47-10e6-4e48-904e-28ff55de653f",
   "metadata": {},
   "source": [
    "# Plan of action\n",
    "\n",
    "- Specify the location of the directory containing the FASTQ files\n",
    "- Specify the location of the metadata (EDIT: I think I will develop this later - I need to be linking SRX/SRA IDs if I want this to have any chance of working)\n",
    "- Develop a prompt for Kallisto quantification\n",
    "- The outcomes should be:\n",
    "  - Kallisto code that appropriately identifies: the correct pair of FASTQ files, correct parameters, correct index selection\n",
    "  - Outputs the results in an appropriate directory\n",
    "  - Outputs the code and logic used\n",
    "  - Outputs the... output... which can be used as an evaluation tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ba65681-b93b-40f3-bd56-1ea2c53cc716",
   "metadata": {},
   "outputs": [],
   "source": [
    "class KallistoCode(BaseModel):\n",
    "    command: str\n",
    "    kallisto_param_logic: str\n",
    "    other_command_logic: str\n",
    "\n",
    "# Intention is to print out logic associated with the Kallisto parameter and any other commands. In my mind, I anticipate a for loop, with the logic behind this decision belonging in other_command\n",
    "\n",
    "prompt = f\"\"\"\n",
    "\n",
    "## IDENTITY AND PURPOSE\n",
    "\n",
    "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
    "\n",
    "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
    "\n",
    "## STEPS\n",
    "\n",
    "1. Consider the list of FASTQ files. Use this to determine which files are matched with eachother - for example, the R1 and R2 files that correspond to eachother.\n",
    "2. Take into account that the path to the index file is ~/work/data/kallisto_indices/human/index.idx\n",
    "3. Take into consideration this documentation for kallisto quant:\n",
    "\n",
    "kallisto 0.50.0\n",
    "Computes equivalence classes for reads and quantifies abundances\n",
    "\n",
    "Usage: kallisto quant [arguments] FASTQ-files\n",
    "\n",
    "Required arguments:\n",
    "-i, --index=STRING            Filename for the kallisto index to be used for\n",
    "                              quantification\n",
    "-o, --output-dir=STRING       Directory to write output to\n",
    "\n",
    "Optional arguments:\n",
    "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
    "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
    "    --plaintext               Output plaintext instead of HDF5\n",
    "    --single                  Quantify single-end reads\n",
    "    --single-overhang         Include reads where unobserved rest of fragment is\n",
    "                              predicted to lie outside a transcript\n",
    "    --fr-stranded             Strand specific reads, first read forward\n",
    "    --rf-stranded             Strand specific reads, first read reverse\n",
    "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
    "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
    "                              (default: -l, -s values are estimated from paired\n",
    "                               end data, but are required when using --single)\n",
    "-t, --threads=INT             Number of threads to use (default: 1)\n",
    "\n",
    "Also note that Kallisto will not take in any of the following legacy options:\n",
    "\n",
    "Legacy options:\n",
    "\n",
    "These options are only supported in kallisto versions before 0.50.0. To use them install an older version of kallisto (recommended: 0.48.0).\n",
    "\n",
    "kallisto quant:\n",
    "\n",
    "--bias learns parameters for a model of sequences specific bias and corrects the abundances accordlingly.\n",
    "\n",
    "--fusion does normal quantification, but additionally looks for reads that do not pseudoalign because they are potentially from fusion genes. All output is written to the file fusion.txt in the output folder.\n",
    "\n",
    "--pseudobam outputs all pseudoalignments to a file pseudoalignments.bam in the output directory. This BAM file contains the pseudoalignments in BAM format, ordered by reads so that each pseudoalignment of a read is adjacent in the BAM file.\n",
    "\n",
    "--genomebam constructs the pseudoalignments to the transcriptome, but projects the transcript alignments to genome coordinates, resulting in split-read alignments. When the --genomebam option is supplied at GTF file must be given with the --gtf option. The GTF file, which can be plain text or gzipped, translates transcripts into genomic coordinates. We recommend downloading a the cdna FASTA files and GTF files from the same data source. The --chromosomes option can provide a length of the genomic chromosomes, this option is not neccessary, but gives a more consistent BAM header, some programs may require this for downstream analysis. kallisto does not require the genome sequence to do pseudoalignment, but downstream tools such as genome browsers will probably need it.\n",
    "\n",
    "4. Carefully consider what you know about using Kallisto for quantification, and provide code to quantify ALL provided FASTQ files in a way that is biologically meaningful. Take into account:\n",
    "- relevant output directories\n",
    "- that all FASTQ files should be quantified in a biologically accurate manner\n",
    "5. Provide a SINGLE command that can be used to perform quantification. This command should be something that can be copy-pasted into the command line\n",
    "6. For each parameter that is specified in the kallisto command, include a justification for why the parameter was included\n",
    "7. For any other part of the command (e.g. wrapping in a for loop), include a justification for why this is necesssary\n",
    "\n",
    "## OUTPUT\n",
    "\n",
    "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
    "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
    "2. kallisto parameter logic - for each kallisto parameter that was specified, include the reasoning for selecting the value that was used\n",
    "3. other command logic - for anything that is used to support the kallisto command (e.g. a for loop), justify why these were needed.\n",
    "\n",
    "## INPUT\n",
    "List of FASTQ files, relative to current directory: \n",
    "\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "be4bf85a-af56-4d30-8315-cd90d02a5851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "command='mkdir -p output && for R1 in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do R2=${R1/_1/_2}; if [ -f \"$R2\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o output/$(basename ${R1%_1.fastq.gz}) --bias --pseudobam --threads=4 $R1 $R2; fi; done' kallisto_param_logic=\"-i ~/work/data/kallisto_indices/human/index.idx: This specifies the path to the Kallisto index for quantification.\\\\n-o output/$(basename ${R1%_1.fastq.gz}): Creates an output directory for each sample using the basename of the R1 file minus the '_1.fastq.gz' suffix. The output structure is important for organizing results.\\\\n--bias: This flag helps to correct for sequence-specific bias, ensuring more accurate quantification.\\\\n--pseudobam: This option generates a BAM file with pseudoalignments, which is useful for further analysis and validation of read alignments in downstream tools.\\\\n--threads=4: Utilizes 4 threads to speed up the quantification process by parallelizing the computation, which is beneficial for large datasets.\" other_command_logic='mkdir -p output: This command creates an output directory to store all the results, ensuring that the workspace is organized.\\\\nfor loop with if condition: This allows us to iterate over each R1 file, find its corresponding R2 file, and run Kallisto only if both files exist, enabling a systematic processing of paired-end reads. This is necessary to ensure that no incomplete pairs are run through the analysis.'\n"
     ]
    }
   ],
   "source": [
    "chat_completion = client.beta.chat.completions.parse(\n",
    "    messages=[\n",
    "        {\"role\": \"system\", \"content\": \"You are an expert bioinformatician who specialises in generating accurate commands to achieve tasks.\"},\n",
    "        {\"role\": \"user\", \"content\": prompt}\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    "    response_format=KallistoCode\n",
    ")\n",
    "print(chat_completion.choices[0].message.parsed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1b873264-0775-4f3a-9e68-d623fc160a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "import subprocess\n",
    "\n",
    "async def execute_command(command, timeout=10):\n",
    "    # Create a process to run the shell command directly\n",
    "    process = await asyncio.create_subprocess_shell(\n",
    "        command,\n",
    "        stdout=asyncio.subprocess.PIPE,\n",
    "        stderr=asyncio.subprocess.PIPE,\n",
    "        shell=True\n",
    "    )\n",
    "    \n",
    "    try:\n",
    "        # Wait for initial output or timeout\n",
    "        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout)\n",
    "        stdout = stdout.decode()\n",
    "        stderr = stderr.decode()\n",
    "        return_code = process.returncode\n",
    "    except asyncio.TimeoutError:\n",
    "        # If we timeout, it means the process is still running\n",
    "        stdout = \"Process started and running in the background.\"\n",
    "        stderr = \"\"\n",
    "        return_code = \"Running\"\n",
    "    \n",
    "    return stdout, stderr, return_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7c0169ff-b41e-4d40-8fd0-506a3b654222",
   "metadata": {},
   "outputs": [],
   "source": [
    "generated_command = chat_completion.choices[0].message.parsed.command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4aa8773c-9b03-4cbd-b05c-5b73c92d7e77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mkdir -p output && for R1 in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do R2=${R1/_1/_2}; if [ -f \"$R2\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o output/$(basename ${R1%_1.fastq.gz}) --bias --pseudobam --threads=4 $R1 $R2; fi; done'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generated_command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "00832689-132e-4c81-9cfc-3e87d3e6c052",
   "metadata": {},
   "outputs": [],
   "source": [
    "class KallistoCode(BaseModel):\n",
    "    command: str\n",
    "    kallisto_param_logic: str\n",
    "    other_command_logic: str\n",
    "\n",
    "# Intention is to print out logic associated with the Kallisto parameter and any other commands. In my mind, I anticipate a for loop, with the logic behind this decision belonging in other_command\n",
    "\n",
    "prompt = f\"\"\"\n",
    "\n",
    "## IDENTITY AND PURPOSE\n",
    "\n",
    "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
    "\n",
    "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
    "\n",
    "## STEPS\n",
    "\n",
    "1. Consider the list of FASTQ files. Use this to determine which files are matched with eachother - for example, the R1 and R2 files that correspond to eachother.\n",
    "2. Take into account that the path to the index file is ~/work/data/kallisto_indices/human/index.idx\n",
    "3. Take into consideration this documentation for kallisto quant:\n",
    "\n",
    "kallisto 0.50.0\n",
    "Computes equivalence classes for reads and quantifies abundances\n",
    "\n",
    "Usage: kallisto quant [arguments] FASTQ-files\n",
    "\n",
    "Required arguments:\n",
    "-i, --index=STRING            Filename for the kallisto index to be used for\n",
    "                              quantification\n",
    "-o, --output-dir=STRING       Directory to write output to\n",
    "\n",
    "Optional arguments:\n",
    "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
    "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
    "    --plaintext               Output plaintext instead of HDF5\n",
    "    --single                  Quantify single-end reads\n",
    "    --single-overhang         Include reads where unobserved rest of fragment is\n",
    "                              predicted to lie outside a transcript\n",
    "    --fr-stranded             Strand specific reads, first read forward\n",
    "    --rf-stranded             Strand specific reads, first read reverse\n",
    "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
    "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
    "                              (default: -l, -s values are estimated from paired\n",
    "                               end data, but are required when using --single)\n",
    "-t, --threads=INT             Number of threads to use (default: 1)\n",
    "\n",
    "Also note that Kallisto will not take in any of the following legacy options:\n",
    "\n",
    "Legacy options:\n",
    "\n",
    "These options are only supported in kallisto versions before 0.50.0. To use them install an older version of kallisto (recommended: 0.48.0).\n",
    "\n",
    "kallisto quant:\n",
    "\n",
    "--bias learns parameters for a model of sequences specific bias and corrects the abundances accordlingly.\n",
    "\n",
    "--fusion does normal quantification, but additionally looks for reads that do not pseudoalign because they are potentially from fusion genes. All output is written to the file fusion.txt in the output folder.\n",
    "\n",
    "--pseudobam outputs all pseudoalignments to a file pseudoalignments.bam in the output directory. This BAM file contains the pseudoalignments in BAM format, ordered by reads so that each pseudoalignment of a read is adjacent in the BAM file.\n",
    "\n",
    "--genomebam constructs the pseudoalignments to the transcriptome, but projects the transcript alignments to genome coordinates, resulting in split-read alignments. When the --genomebam option is supplied at GTF file must be given with the --gtf option. The GTF file, which can be plain text or gzipped, translates transcripts into genomic coordinates. We recommend downloading a the cdna FASTA files and GTF files from the same data source. The --chromosomes option can provide a length of the genomic chromosomes, this option is not neccessary, but gives a more consistent BAM header, some programs may require this for downstream analysis. kallisto does not require the genome sequence to do pseudoalignment, but downstream tools such as genome browsers will probably need it.\n",
    "\n",
    "4. Carefully consider what you know about using Kallisto for quantification, and provide code to quantify ALL provided FASTQ files in a way that is biologically meaningful. Take into account:\n",
    "- relevant output directories\n",
    "- that all FASTQ files should be quantified in a biologically accurate manner\n",
    "5. Provide a SINGLE command that can be used to perform quantification. This command should be something that can be copy-pasted into the command line. Do not include any placeholder values - ensure all values are accurate based on the information provided.\n",
    "- you should ensure checks that any relevant files are included\n",
    "- include robust error handling\n",
    "- include informative print messages\n",
    "6. For each parameter that is specified in the kallisto command, include a justification for why the parameter was included\n",
    "7. For any other part of the command (e.g. wrapping in a for loop), include a justification for why this is necesssary\n",
    "\n",
    "## OUTPUT\n",
    "\n",
    "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
    "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
    "2. kallisto parameter logic - for each kallisto parameter that was specified, include the reasoning for selecting the value that was used\n",
    "3. other command logic - for anything that is used to support the kallisto command (e.g. a for loop), justify why these were needed.\n",
    "\n",
    "## INPUT\n",
    "List of FASTQ files, relative to current directory: \n",
    "\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\"\"\"\n",
    "\n",
    "async def execute_command(command, timeout=10):\n",
    "    # Create a process to run the shell command directly\n",
    "    process = await asyncio.create_subprocess_shell(\n",
    "        command,\n",
    "        stdout=asyncio.subprocess.PIPE,\n",
    "        stderr=asyncio.subprocess.PIPE,\n",
    "        shell=True\n",
    "    )\n",
    "    \n",
    "    try:\n",
    "        # Wait for initial output or timeout\n",
    "        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout)\n",
    "        stdout = stdout.decode()\n",
    "        stderr = stderr.decode()\n",
    "        return_code = process.returncode\n",
    "    except asyncio.TimeoutError:\n",
    "        # If we timeout, it means the process is still running\n",
    "        stdout = \"Process started and running in the background.\"\n",
    "        stderr = \"\"\n",
    "        return_code = \"Running\"\n",
    "    \n",
    "    return stdout, stderr, return_code\n",
    "\n",
    "async def analyze_and_correct_code(command, error_output):\n",
    "    analysis_prompt = f\"\"\"\n",
    "    Analyze the following shell command and its error output. Suggest corrections or improvements to fix the issue.\n",
    "\n",
    "    Command:\n",
    "    {command}\n",
    "\n",
    "    Error Output:\n",
    "    {error_output}\n",
    "\n",
    "    Begin by providing the input command, and stating why this did not work. Then, provide the corrected command or detailed suggestions on what changes are needed.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Use OpenAI to analyze the error and suggest a correction\n",
    "    response = client.chat.completions.create(\n",
    "        model=\"gpt-4o-mini\",  # Adjust the model as needed\n",
    "        messages=[\n",
    "            {\"role\": \"system\", \"content\": \"You are an expert in shell scripting and command-line operations.\"},\n",
    "            {\"role\": \"user\", \"content\": analysis_prompt}\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    suggestions = response.choices[0].message.content\n",
    "    return suggestions\n",
    "\n",
    "async def main_workflow(prompt, max_iterations=3):\n",
    "    for iteration in range(max_iterations):\n",
    "        # Step 1: Generate the command using OpenAI API\n",
    "        chat_completion = client.beta.chat.completions.parse(\n",
    "            messages=[\n",
    "                {\"role\": \"system\", \"content\": \"You are an expert bioinformatician who specialises in generating accurate commands to achieve tasks.\"},\n",
    "                {\"role\": \"user\", \"content\": prompt}\n",
    "            ],\n",
    "            model=\"gpt-4o-mini\",\n",
    "            response_format=KallistoCode\n",
    "        )\n",
    "\n",
    "        # Extract the generated command\n",
    "        generated_command = chat_completion.choices[0].message.parsed.command\n",
    "        \n",
    "        # Step 2: Execute the generated command\n",
    "        print(f\"Executing command (Iteration {iteration + 1}): {generated_command}\")\n",
    "        stdout, stderr, return_code = await execute_command(generated_command)\n",
    "        \n",
    "        if return_code == 0:\n",
    "            print(f\"Command executed successfully:\\n{stdout}\")\n",
    "            break\n",
    "        else:\n",
    "            print(f\"Error executing command:\\n{stderr}\")\n",
    "            # Step 3: Analyze and attempt to correct the command if there is an error\n",
    "            suggestions = await analyze_and_correct_code(generated_command, stderr)\n",
    "            print(f\"Suggested corrections:\\n{suggestions}\")\n",
    "            # Update the prompt for the next iteration with suggestions or the corrected command\n",
    "            prompt = suggestions\n",
    "\n",
    "            if iteration == max_iterations - 1:\n",
    "                print(\"Max iterations reached. The command might still have issues.\")\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8f6cf53e-048a-462d-9988-920531166c80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing command (Iteration 1): for f in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do base=$(basename ${f} _1.fastq.gz); kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o ${base}_kallisto_out --bootstrap-samples=100 --threads=4 ${f} ${f/_1/_2}; done\n",
      "Error executing command:\n",
      "/bin/sh: 1: Bad substitution\n",
      "\n",
      "Suggested corrections:\n",
      "The error you're encountering, `/bin/sh: 1: Bad substitution`, typically arises from the use of the `${f/_1/_2}` string substitution syntax, which is supported in `bash` but not in `sh`. The default shell for many systems, including when you execute scripts or commands without specifying the shell, may be `sh`, which doesn't recognize this syntax.\n",
      "\n",
      "Hereâ€™s a corrected command. The main changes involve using a different syntax for the string manipulation to be compatible with `sh`, or ensuring that the script is executed in `bash`. Let's go through both approaches:\n",
      "\n",
      "### Option 1: Use `bash` explicitly\n",
      "If you know your script or command should run in `bash`, specify it at the start of your script, or invoke the command with `bash`:\n",
      "\n",
      "```bash\n",
      "bash -c 'for f in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do base=$(basename ${f} _1.fastq.gz); kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o ${base}_kallisto_out --bootstrap-samples=100 --threads=4 ${f} ${f/_1/_2}; done'\n",
      "```\n",
      "\n",
      "### Option 2: Modify string substitution to be compatible with `sh`\n",
      "If you want to keep your script compatible with `sh`, use `sed` or `awk` to replace the substring. Hereâ€™s a modified version of your command that should work in `sh`:\n",
      "\n",
      "```bash\n",
      "for f in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do \n",
      "    base=$(basename \"$f\" _1.fastq.gz)\n",
      "    f2=$(echo \"$f\" | sed 's/_1/_2/')\n",
      "    kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o \"${base}_kallisto_out\" --bootstrap-samples=100 --threads=4 \"$f\" \"$f2\"\n",
      "done\n",
      "```\n",
      "\n",
      "### Key Changes Made\n",
      "1. **Changed the substring replacement**: Instead of using `${f/_1/_2}`, I used `sed` to create a new variable `f2` that correctly replaces `_1` with `_2`.\n",
      "\n",
      "2. **Quoting Variables**: It's a good practice to use double quotes around variable expansions to handle filenames with spaces or special characters properly (e.g., `\"$f\"` and `\"$f2\"`).\n",
      "\n",
      "With these corrections made in either option, your command should work without generating the \"Bad substitution\" error.\n",
      "Executing command (Iteration 2): for f in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do base=$(basename \"$f\" _1.fastq.gz); f2=$(echo \"$f\" | sed 's/_1/_2/'); kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o \"${base}_kallisto_out\" --bootstrap-samples=100 --threads=4 \"$f\" \"$f2\"; done\n",
      "Error executing command:\n",
      "\n",
      "Suggested corrections:\n",
      "To analyze the given shell command, let's break it down and identify potential issues that could lead to errors. The command is designed to iterate over a set of `.fastq.gz` files, process them with the `kallisto quant` command, and output results into directories based on the basename of the input files.\n",
      "\n",
      "Here are the components of the command, along with possible issues:\n",
      "\n",
      "1. **File Paths**: The command accesses files at `/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz`. If no files match that pattern, the `for` loop will not run, and the script may give errors later when trying to process non-existent variables.\n",
      "\n",
      "2. **Base Extraction**: The line `base=$(basename \"$f\" _1.fastq.gz)` uses the `basename` command incorrectly. The second argument to `basename` should be a suffix, but if the latter part is not specified properly or if the file path does not contain this suffix, it may lead to unexpected values. Ensure the files actually have this exact suffix.\n",
      "\n",
      "3. **File Pairing**: The line `f2=$(echo \"$f\" | sed 's/_1/_2/')` is intended to create a paired filename for the second read (e.g., converting from `_1` to `_2`). Ensure that `_2.fastq.gz` files exist, as the command will fail if they are absent.\n",
      "\n",
      "4. **Output Directory Creation**: The command does not create an output directory named `${base}_kallisto_out` before running the `kallisto` command. If the output directory does not exist, `kallisto` may fail to write results.\n",
      "\n",
      "5. **Error Handling**: The command does not handle errors explicitly. If `kallisto quant` fails for any reason (e.g., missing input files, or issues with the `index.idx`), there is no output or logging of such errors.\n",
      "\n",
      "### Suggested Corrections and Improvements\n",
      "Here is a corrected version of the command with considerations for the points above:\n",
      "\n",
      "```bash\n",
      "for f in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do\n",
      "    base=$(basename \"$f\" _1.fastq.gz)  # Extract base name\n",
      "    f2=$(echo \"$f\" | sed 's/_1/_2/')    # Get the second fastq file name\n",
      "\n",
      "    # Check if the second fastq file exists\n",
      "    if [[ ! -f \"$f2\" ]]; then\n",
      "        echo \"Warning: File $f2 does not exist. Skipping.\"\n",
      "        continue\n",
      "    fi\n",
      "\n",
      "    # Create output directory if it does not exist\n",
      "    output_dir=\"${base}_kallisto_out\"\n",
      "    mkdir -p \"$output_dir\"\n",
      "\n",
      "    # Run Kallisto quantification\n",
      "    kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o \"$output_dir\" --bootstrap-samples=100 --threads=4 \"$f\" \"$f2\" || {\n",
      "        echo \"Error: Kallisto failed on files $f and $f2.\"\n",
      "        continue\n",
      "    }\n",
      "done\n",
      "```\n",
      "\n",
      "### Key Changes Made:\n",
      "1. **File Existence Check**: Added a check to ensure that `f2` exists before proceeding with the `kallisto` command. If not, it issues a warning and continues to the next iteration.\n",
      "\n",
      "2. **Output Directory Creation**: Used `mkdir -p` to create the output directory, avoiding errors if it already exists.\n",
      "\n",
      "3. **Error Handling**: Included error handling for the `kallisto quant` command to report issues directly, allowing the user to diagnose problems.\n",
      "\n",
      "4. **Echoing Messages**: Added echo statements to inform the user of any issues encountered during processing.\n",
      "\n",
      "This modified version will help ensure smoother operation and clearer information regarding any issues that arise.\n",
      "Executing command (Iteration 3): for f in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do\n",
      "    base=$(basename \"$f\" _1.fastq.gz)  # Extract base name\n",
      "    f2=$(echo \"$f\" | sed 's/_1/_2/')    # Get the second fastq file name\n",
      "\n",
      "    # Check if the second fastq file exists\n",
      "    if [[ ! -f \"$f2\" ]]; then\n",
      "        echo \"Warning: File $f2 does not exist. Skipping.\"\n",
      "        continue\n",
      "    fi\n",
      "\n",
      "    # Create output directory if it does not exist\n",
      "    output_dir=\"${base}_kallisto_out\"\n",
      "    mkdir -p \"$output_dir\"\n",
      "\n",
      "    # Run Kallisto quantification\n",
      "    kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o \"$output_dir\" --bootstrap-samples=100 --threads=4 \"$f\" \"$f2\" || {\n",
      "        echo \"Error: Kallisto failed on files $f and $f2.\"\n",
      "        continue\n",
      "    }\n",
      "done\n",
      "Error executing command:\n",
      "\n",
      "Suggested corrections:\n",
      "To effectively troubleshoot and suggest corrections for the given shell command, we need to analyze its components and consider any potential issues that might arise during its execution. The command processes `.fastq.gz` files to run Kallisto quantification. Hereâ€™s the command along with potential issues and suggested improvements.\n",
      "\n",
      "### Issues to Analyze\n",
      "\n",
      "1. **File Existence Check**:\n",
      "   - The command checks for the existence of the second fastq file (with `_2` suffix) but uses `sed` to replace `_1` with `_2`. This operation may not correctly handle various edge cases (like additional underscores, unexpected naming patterns, etc.).\n",
      "\n",
      "2. **echo Command**:\n",
      "   - Using `echo \"$f\" | sed 's/_1/_2/'` might unintentionally replace multiple instances of `_1` if the base filename has more than one `_1`. Consider using parameter expansion for better reliability.\n",
      "\n",
      "3. **Output Directory**:\n",
      "   - The output directory is created using the base name derived from the `_1` suffix, which may lead to naming conflicts or unintended directories if `_1` is part of a longer string. Ensure uniqueness or clarity in naming.\n",
      "\n",
      "4. **Kallisto Command**:\n",
      "   - Ensure that the path provided for the `kallisto` executable and the index file is correct. Also, check if the required arguments, like threads and bootstrap-samples, are appropriate for your system.\n",
      "\n",
      "### Suggested Improvements\n",
      "\n",
      "Here are the corrections and more robust implementations for the command:\n",
      "\n",
      "```bash\n",
      "for f in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do\n",
      "    base=\"${f%_1.fastq.gz}\"  # Extract base name correctly\n",
      "    f2=\"${f/_1/_2}\"  # Get the second fastq file name using parameter expansion\n",
      "\n",
      "    # Check if the second fastq file exists\n",
      "    if [[ ! -f \"$f2\" ]]; then\n",
      "        echo \"Warning: File $f2 does not exist. Skipping.\"\n",
      "        continue\n",
      "    fi\n",
      "\n",
      "    # Create output directory if it does not exist\n",
      "    output_dir=\"${base##*/}_kallisto_out\"  # Ensure to get the last part of the path for clarity\n",
      "    mkdir -p \"$output_dir\"\n",
      "\n",
      "    # Run Kallisto quantification\n",
      "    kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o \"$output_dir\" --bootstrap-samples=100 --threads=4 \"$f\" \"$f2\" || {\n",
      "        echo \"Error: Kallisto failed on files $f and $f2.\"\n",
      "        continue\n",
      "    }\n",
      "done\n",
      "```\n",
      "\n",
      "### Explanation of Changes\n",
      "\n",
      "1. **Parameter Expansion**:\n",
      "   - Used `${f%_1.fastq.gz}` and `${f/_1/_2}` for cleaner and safer string manipulation. This handles only the last occurrence of `_1` when creating the second file name.\n",
      "\n",
      "2. **Output Directory Naming**:\n",
      "   - Used `${base##*/}` to ensure only the final component of the path is used for the output directory name. This prevents conflicts if the full path might resemble other directory names.\n",
      "\n",
      "3. **Error Handling**:\n",
      "   - Maintained the existing error handling in the Kallisto command but emphasize the importance of checking the logs if the command fails to provide insights into the source of the error (further improvements can involve capturing standard error and logging it).\n",
      "\n",
      "### Additional Recommendations\n",
      "\n",
      "- It's generally helpful to include debug prints (like `echo \"Processing $f and $f2\"`) within your loop to follow along with the script's progress during execution, especially on larger directories.\n",
      "- Ensure that the `kallisto` command is available in your PATH or define its full path if necessary.\n",
      "- Consider logging errors and warnings to a separate log file for easier troubleshooting in cases of large datasets.\n",
      "Max iterations reached. The command might still have issues.\n"
     ]
    }
   ],
   "source": [
    "await main_workflow(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e606c704-0f1d-42b0-b95d-cdda95e3eaf1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (Poetry)",
   "language": "python",
   "name": "poetry-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
