{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "932a495e-6d11-47e6-95ff-fe4aefba2d17",
   "metadata": {},
   "source": [
    "# Description\n",
    "\n",
    "This notebook is me doing a bit of interactive testing for the two sample datasets I will be working with. Specifically, I want to get an idea of how I can be using R to explore datasets.\n",
    "\n",
    "It has seemed to me that the R tools to explore GEO datasets seem better suited than Python. Perhaps the command line is also useful in some cases, but in any case, I'd like to do some experimentation with R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d5a894b9-8544-4277-9b8f-91180f864d93",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Setting options('download.file.method.GEOquery'='auto')\n",
      "\n",
      "Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(GEOquery)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fb143ea9-ce5f-4bd4-81f5-266405d29598",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found 3 file(s)\n",
      "\n",
      "GSE16146-GPL8523_series_matrix.txt.gz\n",
      "\n",
      "GSE16146-GPL8561_series_matrix.txt.gz\n",
      "\n",
      "GSE16146-GPL8562_series_matrix.txt.gz\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 3\n",
      " $ GSE16146-GPL8523_series_matrix.txt.gz:Formal class 'ExpressionSet' [package \"Biobase\"] with 7 slots\n",
      "  .. ..@ experimentData   :Formal class 'MIAME' [package \"Biobase\"] with 13 slots\n",
      "  .. .. .. ..@ name             : chr \"Martin,I,Voskuil\"\n",
      "  .. .. .. ..@ lab              : chr \"\"\n",
      "  .. .. .. ..@ contact          : chr \"martin.voskuil@ucdenver.edu\"\n",
      "  .. .. .. ..@ title            : chr \"The response of Mycobacterium tuberculosis to reactive oxygen and nitrogen species\"\n",
      "  .. .. .. ..@ abstract         : chr \"The bacteriostatic and bactericidal effects and the corresponding expression profiles of Mycobacterium tubercul\"| __truncated__\n",
      "  .. .. .. ..@ url              : chr \"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE16146\"\n",
      "  .. .. .. ..@ pubMedIds        : chr \"\"\n",
      "  .. .. .. ..@ samples          : list()\n",
      "  .. .. .. ..@ hybridizations   : list()\n",
      "  .. .. .. ..@ normControls     : list()\n",
      "  .. .. .. ..@ preprocessing    : list()\n",
      "  .. .. .. ..@ other            :List of 25\n",
      "  .. .. .. .. ..$ contact_address        : chr \"12800 E. 19th Ave\"\n",
      "  .. .. .. .. ..$ contact_city           : chr \"Aurora\"\n",
      "  .. .. .. .. ..$ contact_country        : chr \"USA\"\n",
      "  .. .. .. .. ..$ contact_department     : chr \"Microbiology\"\n",
      "  .. .. .. .. ..$ contact_email          : chr \"martin.voskuil@ucdenver.edu\"\n",
      "  .. .. .. .. ..$ contact_institute      : chr \"University of Colorado Denver\"\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr \"Voskuil Lab\"\n",
      "  .. .. .. .. ..$ contact_name           : chr \"Martin,I,Voskuil\"\n",
      "  .. .. .. .. ..$ contact_phone          : chr \"303-724-4219\"\n",
      "  .. .. .. .. ..$ contact_state          : chr \"CO\"\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr \"80045\"\n",
      "  .. .. .. .. ..$ contributor            : chr \"Martin,,Voskuil\\nKevin,,Visconti\"\n",
      "  .. .. .. .. ..$ geo_accession          : chr \"GSE16146\"\n",
      "  .. .. .. .. ..$ last_update_date       : chr \"Mar 21 2012\"\n",
      "  .. .. .. .. ..$ overall_design         : chr \"Set of arrays that are part of repeated experiments\\nCompound Based Treatment: H2O2 or DETA/NO treatment\"\n",
      "  .. .. .. .. ..$ platform_id            : chr \"GPL8523\\nGPL8561\\nGPL8562\"\n",
      "  .. .. .. .. ..$ platform_taxid         : chr \"1773\"\n",
      "  .. .. .. .. ..$ sample_id              : chr \"GSM404398 GSM404399 GSM404400 GSM404401 GSM404402 GSM404403 GSM404404 GSM404405 GSM404406 GSM404407 GSM404408 G\"| __truncated__\n",
      "  .. .. .. .. ..$ sample_taxid           : chr \"1773\"\n",
      "  .. .. .. .. ..$ status                 : chr \"Public on May 18 2010\"\n",
      "  .. .. .. .. ..$ submission_date        : chr \"May 18 2009\"\n",
      "  .. .. .. .. ..$ summary                : chr \"The bacteriostatic and bactericidal effects and the corresponding expression profiles of Mycobacterium tubercul\"| __truncated__\n",
      "  .. .. .. .. ..$ supplementary_file     : chr \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE16nnn/GSE16146/suppl/GSE16146_RAW.tar\"\n",
      "  .. .. .. .. ..$ title                  : chr \"The response of Mycobacterium tuberculosis to reactive oxygen and nitrogen species\"\n",
      "  .. .. .. .. ..$ type                   : chr \"Expression profiling by array\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 2\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr [1:2] \"MIAxE\" \"MIAME\"\n",
      "  .. ..@ assayData        :<environment: 0xaaab10b4e5b8> \n",
      "  .. ..@ phenoData        :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t51 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:51] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t25 obs. of  51 variables:\n",
      "  .. .. .. .. ..$ title                  : chr [1:25] \"MTB strain 1254 Ctrl vs 10.0 mM H202 40min rep 5\" \"MTB strain 1254 Ctrl vs 5.0 mM H202 40min rep 4\" \"MTB strain 1254 Ctrl vs 50.0 mM H202 40min rep 5\" \"MTB strain 1254 Ctrl vs 200 mM H202 40min rep 3\" ...\n",
      "  .. .. .. .. ..$ geo_accession          : chr [1:25] \"GSM404398\" \"GSM404402\" \"GSM404406\" \"GSM404407\" ...\n",
      "  .. .. .. .. ..$ status                 : chr [1:25] \"Public on May 18 2010\" \"Public on May 18 2010\" \"Public on May 18 2010\" \"Public on May 18 2010\" ...\n",
      "  .. .. .. .. ..$ submission_date        : chr [1:25] \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" ...\n",
      "  .. .. .. .. ..$ last_update_date       : chr [1:25] \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" ...\n",
      "  .. .. .. .. ..$ type                   : chr [1:25] \"RNA\" \"RNA\" \"RNA\" \"RNA\" ...\n",
      "  .. .. .. .. ..$ channel_count          : chr [1:25] \"2\" \"2\" \"2\" \"2\" ...\n",
      "  .. .. .. .. ..$ source_name_ch1        : chr [1:25] \"Strain: MTB strain 1254 Treatment: Control\" \"Strain: MTB strain 1254 Treatment: Control\" \"Strain: MTB strain 1254 Treatment: Control\" \"Strain: MTB strain 1254 Treatment: Control\" ...\n",
      "  .. .. .. .. ..$ organism_ch1           : chr [1:25] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1    : chr [1:25] \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.1  : chr [1:25] \"treatment: Control\" \"treatment: Control\" \"treatment: Control\" \"treatment: Control\" ...\n",
      "  .. .. .. .. ..$ molecule_ch1           : chr [1:25] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1   : chr [1:25] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ label_ch1              : chr [1:25] \"Cy3\" \"Cy3\" \"Cy3\" \"Cy3\" ...\n",
      "  .. .. .. .. ..$ label_protocol_ch1     : chr [1:25] \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch1              : chr [1:25] \"1773\" \"1773\" \"1773\" \"1773\" ...\n",
      "  .. .. .. .. ..$ source_name_ch2        : chr [1:25] \"Strain: MTB strain 1254 Treatment: 10.0 mM H2O2 for 40 min\" \"Strain: MTB strain 1254 Treatment: 5.0 mM H2O2 for 40 min\" \"Strain: MTB strain 1254 Treatment: 50.0 mM H2O2 for 40 min\" \"Strain: MTB strain 1254 Treatment: 200 mM H2O2 for 40 min\" ...\n",
      "  .. .. .. .. ..$ organism_ch2           : chr [1:25] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch2    : chr [1:25] \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch2.1  : chr [1:25] \"treatment: 10.0 mM H2O2 for 40 min\" \"treatment: 5.0 mM H2O2 for 40 min\" \"treatment: 50.0 mM H2O2 for 40 min\" \"treatment: 200 mM H2O2 for 40 min\" ...\n",
      "  .. .. .. .. ..$ growth_protocol_ch2    : chr [1:25] \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ molecule_ch2           : chr [1:25] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch2   : chr [1:25] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch2.1 : chr [1:25] \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ ...\n",
      "  .. .. .. .. ..$ label_ch2              : chr [1:25] \"Cy5\" \"Cy5\" \"Cy5\" \"Cy5\" ...\n",
      "  .. .. .. .. ..$ label_protocol_ch2     : chr [1:25] \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ label_protocol_ch2.1   : chr [1:25] \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch2              : chr [1:25] \"1773\" \"1773\" \"1773\" \"1773\" ...\n",
      "  .. .. .. .. ..$ hyb_protocol           : chr [1:25] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ scan_protocol          : chr [1:25] \"GenePix\" \"GenePix\" \"GenePix\" \"GenePix\" ...\n",
      "  .. .. .. .. ..$ scan_protocol.1        : chr [1:25] \"GenePix\" \"GenePix\" \"GenePix\" \"GenePix\" ...\n",
      "  .. .. .. .. ..$ description            : chr [1:25] \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74742.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2009-05/83718.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74748.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74750.gif\" ...\n",
      "  .. .. .. .. ..$ data_processing        : chr [1:25] \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" ...\n",
      "  .. .. .. .. ..$ platform_id            : chr [1:25] \"GPL8523\" \"GPL8523\" \"GPL8523\" \"GPL8523\" ...\n",
      "  .. .. .. .. ..$ contact_name           : chr [1:25] \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" ...\n",
      "  .. .. .. .. ..$ contact_email          : chr [1:25] \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" ...\n",
      "  .. .. .. .. ..$ contact_phone          : chr [1:25] \"303-724-4219\" \"303-724-4219\" \"303-724-4219\" \"303-724-4219\" ...\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr [1:25] \"Voskuil Lab\" \"Voskuil Lab\" \"Voskuil Lab\" \"Voskuil Lab\" ...\n",
      "  .. .. .. .. ..$ contact_department     : chr [1:25] \"Microbiology\" \"Microbiology\" \"Microbiology\" \"Microbiology\" ...\n",
      "  .. .. .. .. ..$ contact_institute      : chr [1:25] \"University of Colorado Denver\" \"University of Colorado Denver\" \"University of Colorado Denver\" \"University of Colorado Denver\" ...\n",
      "  .. .. .. .. ..$ contact_address        : chr [1:25] \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" ...\n",
      "  .. .. .. .. ..$ contact_city           : chr [1:25] \"Aurora\" \"Aurora\" \"Aurora\" \"Aurora\" ...\n",
      "  .. .. .. .. ..$ contact_state          : chr [1:25] \"CO\" \"CO\" \"CO\" \"CO\" ...\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr [1:25] \"80045\" \"80045\" \"80045\" \"80045\" ...\n",
      "  .. .. .. .. ..$ contact_country        : chr [1:25] \"USA\" \"USA\" \"USA\" \"USA\" ...\n",
      "  .. .. .. .. ..$ supplementary_file     : chr [1:25] \"NONE\" \"NONE\" \"NONE\" \"NONE\" ...\n",
      "  .. .. .. .. ..$ data_row_count         : chr [1:25] \"4608\" \"4608\" \"4608\" \"4608\" ...\n",
      "  .. .. .. .. ..$ strain:ch1             : chr [1:25] \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ strain:ch2             : chr [1:25] \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ treatment:ch1          : chr [1:25] \"Control\" \"Control\" \"Control\" \"Control\" ...\n",
      "  .. .. .. .. ..$ treatment:ch2          : chr [1:25] \"10.0 mM H2O2 for 40 min\" \"5.0 mM H2O2 for 40 min\" \"50.0 mM H2O2 for 40 min\" \"200 mM H2O2 for 40 min\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ featureData      :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t13 obs. of  3 variables:\n",
      "  .. .. .. .. ..$ Column          : chr [1:13] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Description     : chr [1:13] \"ID =\" \"MetaColumn =\" \"MetaRow =\" \"Column =\" ...\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:13] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t4608 obs. of  13 variables:\n",
      "  .. .. .. .. ..$ ID                             : int [1:4608] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ MetaColumn                     : int [1:4608] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ MetaRow                        : int [1:4608] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ Column                         : int [1:4608] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ Row                            : int [1:4608] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ Reporter Name                  : chr [1:4608] \"RV0001-pcr\" \"RV0003-pcr\" \"RV0005-pcr\" \"RV0007-pcr\" ...\n",
      "  .. .. .. .. ..$ ORF                            : chr [1:4608] \"RV0001\" \"RV0003\" \"RV0005\" \"RV0007\" ...\n",
      "  .. .. .. .. ..$ Reporter Species               : chr [1:4608] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ Repoter Biosequence Type       : chr [1:4608] \"PCR_amplicon\" \"PCR_amplicon\" \"PCR_amplicon\" \"PCR_amplicon\" ...\n",
      "  .. .. .. .. ..$ Repoter BioSequence PolymerType: chr [1:4608] \"DNA\" \"DNA\" \"DNA\" \"DNA\" ...\n",
      "  .. .. .. .. ..$ Reporter Group[role]           : chr [1:4608] \"Experimental\" \"Experimental\" \"Experimental\" \"Experimental\" ...\n",
      "  .. .. .. .. ..$ Reporter ControlType           : chr [1:4608] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ SPOT_ID                        : chr [1:4608] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"featureNames\" \"featureColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ annotation       : chr \"GPL8523\"\n",
      "  .. ..@ protocolData     :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t25 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. ..@ .Data:List of 4\n",
      "  .. .. .. .. ..$ : int [1:3] 4 2 2\n",
      "  .. .. .. .. ..$ : int [1:3] 2 58 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 3 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. ..$ names: chr [1:4] \"R\" \"Biobase\" \"eSet\" \"ExpressionSet\"\n",
      " $ GSE16146-GPL8561_series_matrix.txt.gz:Formal class 'ExpressionSet' [package \"Biobase\"] with 7 slots\n",
      "  .. ..@ experimentData   :Formal class 'MIAME' [package \"Biobase\"] with 13 slots\n",
      "  .. .. .. ..@ name             : chr \"Martin,I,Voskuil\"\n",
      "  .. .. .. ..@ lab              : chr \"\"\n",
      "  .. .. .. ..@ contact          : chr \"martin.voskuil@ucdenver.edu\"\n",
      "  .. .. .. ..@ title            : chr \"The response of Mycobacterium tuberculosis to reactive oxygen and nitrogen species\"\n",
      "  .. .. .. ..@ abstract         : chr \"The bacteriostatic and bactericidal effects and the corresponding expression profiles of Mycobacterium tubercul\"| __truncated__\n",
      "  .. .. .. ..@ url              : chr \"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE16146\"\n",
      "  .. .. .. ..@ pubMedIds        : chr \"\"\n",
      "  .. .. .. ..@ samples          : list()\n",
      "  .. .. .. ..@ hybridizations   : list()\n",
      "  .. .. .. ..@ normControls     : list()\n",
      "  .. .. .. ..@ preprocessing    : list()\n",
      "  .. .. .. ..@ other            :List of 25\n",
      "  .. .. .. .. ..$ contact_address        : chr \"12800 E. 19th Ave\"\n",
      "  .. .. .. .. ..$ contact_city           : chr \"Aurora\"\n",
      "  .. .. .. .. ..$ contact_country        : chr \"USA\"\n",
      "  .. .. .. .. ..$ contact_department     : chr \"Microbiology\"\n",
      "  .. .. .. .. ..$ contact_email          : chr \"martin.voskuil@ucdenver.edu\"\n",
      "  .. .. .. .. ..$ contact_institute      : chr \"University of Colorado Denver\"\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr \"Voskuil Lab\"\n",
      "  .. .. .. .. ..$ contact_name           : chr \"Martin,I,Voskuil\"\n",
      "  .. .. .. .. ..$ contact_phone          : chr \"303-724-4219\"\n",
      "  .. .. .. .. ..$ contact_state          : chr \"CO\"\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr \"80045\"\n",
      "  .. .. .. .. ..$ contributor            : chr \"Martin,,Voskuil\\nKevin,,Visconti\"\n",
      "  .. .. .. .. ..$ geo_accession          : chr \"GSE16146\"\n",
      "  .. .. .. .. ..$ last_update_date       : chr \"Mar 21 2012\"\n",
      "  .. .. .. .. ..$ overall_design         : chr \"Set of arrays that are part of repeated experiments\\nCompound Based Treatment: H2O2 or DETA/NO treatment\"\n",
      "  .. .. .. .. ..$ platform_id            : chr \"GPL8523\\nGPL8561\\nGPL8562\"\n",
      "  .. .. .. .. ..$ platform_taxid         : chr \"1773\"\n",
      "  .. .. .. .. ..$ sample_id              : chr \"GSM404398 GSM404399 GSM404400 GSM404401 GSM404402 GSM404403 GSM404404 GSM404405 GSM404406 GSM404407 GSM404408 G\"| __truncated__\n",
      "  .. .. .. .. ..$ sample_taxid           : chr \"1773\"\n",
      "  .. .. .. .. ..$ status                 : chr \"Public on May 18 2010\"\n",
      "  .. .. .. .. ..$ submission_date        : chr \"May 18 2009\"\n",
      "  .. .. .. .. ..$ summary                : chr \"The bacteriostatic and bactericidal effects and the corresponding expression profiles of Mycobacterium tubercul\"| __truncated__\n",
      "  .. .. .. .. ..$ supplementary_file     : chr \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE16nnn/GSE16146/suppl/GSE16146_RAW.tar\"\n",
      "  .. .. .. .. ..$ title                  : chr \"The response of Mycobacterium tuberculosis to reactive oxygen and nitrogen species\"\n",
      "  .. .. .. .. ..$ type                   : chr \"Expression profiling by array\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 2\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr [1:2] \"MIAxE\" \"MIAME\"\n",
      "  .. ..@ assayData        :<environment: 0xaaab10c91cb8> \n",
      "  .. ..@ phenoData        :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t51 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:51] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t32 obs. of  51 variables:\n",
      "  .. .. .. .. ..$ title                  : chr [1:32] \"MTB strain 1254 Ctrl vs 0.5 mM DETA/NO 4hrs rep 3\" \"MTB strain 1254 Ctrl vs 0.5 mM DETA/NO 40min rep 4\" \"MTB strain1254 vs MTB strain 1254 for DETA/NO exp control rep 3\" \"MTB strain 1254 Ctrl vs 0.05 mM DETA/NO 40min rep 2\" ...\n",
      "  .. .. .. .. ..$ geo_accession          : chr [1:32] \"GSM404400\" \"GSM404403\" \"GSM404404\" \"GSM404410\" ...\n",
      "  .. .. .. .. ..$ status                 : chr [1:32] \"Public on May 18 2010\" \"Public on May 18 2010\" \"Public on May 18 2010\" \"Public on May 18 2010\" ...\n",
      "  .. .. .. .. ..$ submission_date        : chr [1:32] \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" ...\n",
      "  .. .. .. .. ..$ last_update_date       : chr [1:32] \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" ...\n",
      "  .. .. .. .. ..$ type                   : chr [1:32] \"RNA\" \"RNA\" \"RNA\" \"RNA\" ...\n",
      "  .. .. .. .. ..$ channel_count          : chr [1:32] \"2\" \"2\" \"2\" \"2\" ...\n",
      "  .. .. .. .. ..$ source_name_ch1        : chr [1:32] \"Strain: MTB strain 1254 Treatment: Control\" \"Strain: MTB strain 1254 Treatment: Control\" \"Strain: MTB strain 1254 Tretment: No Treatment\" \"Strain: MTB strain 1254 Treatment: Control\" ...\n",
      "  .. .. .. .. ..$ organism_ch1           : chr [1:32] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1    : chr [1:32] \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.1  : chr [1:32] \"treatment: Control\" \"treatment: Control\" \"treatment: No Treatment\" \"treatment: Control\" ...\n",
      "  .. .. .. .. ..$ molecule_ch1           : chr [1:32] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1   : chr [1:32] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ label_ch1              : chr [1:32] \"Cy3\" \"Cy3\" \"Cy3\" \"Cy3\" ...\n",
      "  .. .. .. .. ..$ label_protocol_ch1     : chr [1:32] \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch1              : chr [1:32] \"1773\" \"1773\" \"1773\" \"1773\" ...\n",
      "  .. .. .. .. ..$ source_name_ch2        : chr [1:32] \"Strain: MTB strain 1254 Treatment: 0.5 mM DETA/NO for 4hrs\" \"Strain: MTB strain 1254 Treatment: 0.5 mM DETA/NO for 40min\" \"Strain: MTB strain 1254 Tretment: No Treatment\" \"Strain: MTB strain 1254 Treatment: 0.05 mM DETA/NO for 40min\" ...\n",
      "  .. .. .. .. ..$ organism_ch2           : chr [1:32] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch2    : chr [1:32] \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch2.1  : chr [1:32] \"treatment: 0.5 mM DETA/NO for 4hrs\" \"treatment: 0.5 mM DETA/NO for 40min\" \"treatment: No Treatment\" \"treatment: 0.05 mM DETA/NO for 40min\" ...\n",
      "  .. .. .. .. ..$ growth_protocol_ch2    : chr [1:32] \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ molecule_ch2           : chr [1:32] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch2   : chr [1:32] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch2.1 : chr [1:32] \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ ...\n",
      "  .. .. .. .. ..$ label_ch2              : chr [1:32] \"Cy5\" \"Cy5\" \"Cy5\" \"Cy5\" ...\n",
      "  .. .. .. .. ..$ label_protocol_ch2     : chr [1:32] \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ label_protocol_ch2.1   : chr [1:32] \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch2              : chr [1:32] \"1773\" \"1773\" \"1773\" \"1773\" ...\n",
      "  .. .. .. .. ..$ hyb_protocol           : chr [1:32] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ scan_protocol          : chr [1:32] \"GenePix\" \"GenePix\" \"GenePix\" \"GenePix\" ...\n",
      "  .. .. .. .. ..$ scan_protocol.1        : chr [1:32] \"GenePix\" \"GenePix\" \"GenePix\" \"\" ...\n",
      "  .. .. .. .. ..$ description            : chr [1:32] \"Image: http://smd.stanford.edu/MicroArray/gifs/2009-05/83726.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74760.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74765.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2001-07/18439.gif\" ...\n",
      "  .. .. .. .. ..$ data_processing        : chr [1:32] \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" ...\n",
      "  .. .. .. .. ..$ platform_id            : chr [1:32] \"GPL8561\" \"GPL8561\" \"GPL8561\" \"GPL8561\" ...\n",
      "  .. .. .. .. ..$ contact_name           : chr [1:32] \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" ...\n",
      "  .. .. .. .. ..$ contact_email          : chr [1:32] \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" ...\n",
      "  .. .. .. .. ..$ contact_phone          : chr [1:32] \"303-724-4219\" \"303-724-4219\" \"303-724-4219\" \"303-724-4219\" ...\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr [1:32] \"Voskuil Lab\" \"Voskuil Lab\" \"Voskuil Lab\" \"Voskuil Lab\" ...\n",
      "  .. .. .. .. ..$ contact_department     : chr [1:32] \"Microbiology\" \"Microbiology\" \"Microbiology\" \"Microbiology\" ...\n",
      "  .. .. .. .. ..$ contact_institute      : chr [1:32] \"University of Colorado Denver\" \"University of Colorado Denver\" \"University of Colorado Denver\" \"University of Colorado Denver\" ...\n",
      "  .. .. .. .. ..$ contact_address        : chr [1:32] \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" ...\n",
      "  .. .. .. .. ..$ contact_city           : chr [1:32] \"Aurora\" \"Aurora\" \"Aurora\" \"Aurora\" ...\n",
      "  .. .. .. .. ..$ contact_state          : chr [1:32] \"CO\" \"CO\" \"CO\" \"CO\" ...\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr [1:32] \"80045\" \"80045\" \"80045\" \"80045\" ...\n",
      "  .. .. .. .. ..$ contact_country        : chr [1:32] \"USA\" \"USA\" \"USA\" \"USA\" ...\n",
      "  .. .. .. .. ..$ supplementary_file     : chr [1:32] \"NONE\" \"NONE\" \"NONE\" \"NONE\" ...\n",
      "  .. .. .. .. ..$ data_row_count         : chr [1:32] \"5760\" \"5760\" \"5760\" \"5760\" ...\n",
      "  .. .. .. .. ..$ strain:ch1             : chr [1:32] \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ strain:ch2             : chr [1:32] \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ treatment:ch1          : chr [1:32] \"Control\" \"Control\" \"No Treatment\" \"Control\" ...\n",
      "  .. .. .. .. ..$ treatment:ch2          : chr [1:32] \"0.5 mM DETA/NO for 4hrs\" \"0.5 mM DETA/NO for 40min\" \"No Treatment\" \"0.05 mM DETA/NO for 40min\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ featureData      :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t13 obs. of  3 variables:\n",
      "  .. .. .. .. ..$ Column          : chr [1:13] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Description     : chr [1:13] \"ID =\" \"MetaColumn =\" \"MetaRow =\" \"Column =\" ...\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:13] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t5760 obs. of  13 variables:\n",
      "  .. .. .. .. ..$ ID                             : int [1:5760] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ MetaColumn                     : int [1:5760] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ MetaRow                        : int [1:5760] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ Column                         : int [1:5760] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ Row                            : int [1:5760] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ Reporter Name                  : chr [1:5760] \"UNKNOWN\" \"UNKNOWN\" \"UNKNOWN\" \"UNKNOWN\" ...\n",
      "  .. .. .. .. ..$ ORF                            : chr [1:5760] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Reporter Species               : chr [1:5760] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ Repoter Biosequence Type       : chr [1:5760] \"unknown_sequence\" \"unknown_sequence\" \"unknown_sequence\" \"unknown_sequence\" ...\n",
      "  .. .. .. .. ..$ Repoter BioSequence PolymerType: chr [1:5760] \"DNA\" \"DNA\" \"DNA\" \"DNA\" ...\n",
      "  .. .. .. .. ..$ Reporter Group[role]           : chr [1:5760] \"Control\" \"Control\" \"Control\" \"Control\" ...\n",
      "  .. .. .. .. ..$ Reporter ControlType           : chr [1:5760] \"control_unknown_type\" \"control_unknown_type\" \"control_unknown_type\" \"control_unknown_type\" ...\n",
      "  .. .. .. .. ..$ SPOT_ID                        : chr [1:5760] \"CONTROL\" \"CONTROL\" \"CONTROL\" \"CONTROL\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"featureNames\" \"featureColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ annotation       : chr \"GPL8561\"\n",
      "  .. ..@ protocolData     :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t32 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. ..@ .Data:List of 4\n",
      "  .. .. .. .. ..$ : int [1:3] 4 2 2\n",
      "  .. .. .. .. ..$ : int [1:3] 2 58 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 3 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. ..$ names: chr [1:4] \"R\" \"Biobase\" \"eSet\" \"ExpressionSet\"\n",
      " $ GSE16146-GPL8562_series_matrix.txt.gz:Formal class 'ExpressionSet' [package \"Biobase\"] with 7 slots\n",
      "  .. ..@ experimentData   :Formal class 'MIAME' [package \"Biobase\"] with 13 slots\n",
      "  .. .. .. ..@ name             : chr \"Martin,I,Voskuil\"\n",
      "  .. .. .. ..@ lab              : chr \"\"\n",
      "  .. .. .. ..@ contact          : chr \"martin.voskuil@ucdenver.edu\"\n",
      "  .. .. .. ..@ title            : chr \"The response of Mycobacterium tuberculosis to reactive oxygen and nitrogen species\"\n",
      "  .. .. .. ..@ abstract         : chr \"The bacteriostatic and bactericidal effects and the corresponding expression profiles of Mycobacterium tubercul\"| __truncated__\n",
      "  .. .. .. ..@ url              : chr \"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE16146\"\n",
      "  .. .. .. ..@ pubMedIds        : chr \"\"\n",
      "  .. .. .. ..@ samples          : list()\n",
      "  .. .. .. ..@ hybridizations   : list()\n",
      "  .. .. .. ..@ normControls     : list()\n",
      "  .. .. .. ..@ preprocessing    : list()\n",
      "  .. .. .. ..@ other            :List of 25\n",
      "  .. .. .. .. ..$ contact_address        : chr \"12800 E. 19th Ave\"\n",
      "  .. .. .. .. ..$ contact_city           : chr \"Aurora\"\n",
      "  .. .. .. .. ..$ contact_country        : chr \"USA\"\n",
      "  .. .. .. .. ..$ contact_department     : chr \"Microbiology\"\n",
      "  .. .. .. .. ..$ contact_email          : chr \"martin.voskuil@ucdenver.edu\"\n",
      "  .. .. .. .. ..$ contact_institute      : chr \"University of Colorado Denver\"\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr \"Voskuil Lab\"\n",
      "  .. .. .. .. ..$ contact_name           : chr \"Martin,I,Voskuil\"\n",
      "  .. .. .. .. ..$ contact_phone          : chr \"303-724-4219\"\n",
      "  .. .. .. .. ..$ contact_state          : chr \"CO\"\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr \"80045\"\n",
      "  .. .. .. .. ..$ contributor            : chr \"Martin,,Voskuil\\nKevin,,Visconti\"\n",
      "  .. .. .. .. ..$ geo_accession          : chr \"GSE16146\"\n",
      "  .. .. .. .. ..$ last_update_date       : chr \"Mar 21 2012\"\n",
      "  .. .. .. .. ..$ overall_design         : chr \"Set of arrays that are part of repeated experiments\\nCompound Based Treatment: H2O2 or DETA/NO treatment\"\n",
      "  .. .. .. .. ..$ platform_id            : chr \"GPL8523\\nGPL8561\\nGPL8562\"\n",
      "  .. .. .. .. ..$ platform_taxid         : chr \"1773\"\n",
      "  .. .. .. .. ..$ sample_id              : chr \"GSM404398 GSM404399 GSM404400 GSM404401 GSM404402 GSM404403 GSM404404 GSM404405 GSM404406 GSM404407 GSM404408 G\"| __truncated__\n",
      "  .. .. .. .. ..$ sample_taxid           : chr \"1773\"\n",
      "  .. .. .. .. ..$ status                 : chr \"Public on May 18 2010\"\n",
      "  .. .. .. .. ..$ submission_date        : chr \"May 18 2009\"\n",
      "  .. .. .. .. ..$ summary                : chr \"The bacteriostatic and bactericidal effects and the corresponding expression profiles of Mycobacterium tubercul\"| __truncated__\n",
      "  .. .. .. .. ..$ supplementary_file     : chr \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE16nnn/GSE16146/suppl/GSE16146_RAW.tar\"\n",
      "  .. .. .. .. ..$ title                  : chr \"The response of Mycobacterium tuberculosis to reactive oxygen and nitrogen species\"\n",
      "  .. .. .. .. ..$ type                   : chr \"Expression profiling by array\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 2\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr [1:2] \"MIAxE\" \"MIAME\"\n",
      "  .. ..@ assayData        :<environment: 0xaaaaf86d60b8> \n",
      "  .. ..@ phenoData        :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t51 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:51] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t23 obs. of  51 variables:\n",
      "  .. .. .. .. ..$ title                  : chr [1:23] \"MTB strain 1254 Ctrl vs 0.5 mM H202 40min rep 1\" \"MTB strain1254 vs MTB strain 1254 for H2O2 exp control rep 5\" \"MTB strain 1254 Ctrl vs 50.0 mM H202 40min rep 1\" \"MTB strain 1254 Ctrl vs 0.05 mM H202 40min rep 3\" ...\n",
      "  .. .. .. .. ..$ geo_accession          : chr [1:23] \"GSM404399\" \"GSM404401\" \"GSM404405\" \"GSM404414\" ...\n",
      "  .. .. .. .. ..$ status                 : chr [1:23] \"Public on May 18 2010\" \"Public on May 18 2010\" \"Public on May 18 2010\" \"Public on May 18 2010\" ...\n",
      "  .. .. .. .. ..$ submission_date        : chr [1:23] \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" ...\n",
      "  .. .. .. .. ..$ last_update_date       : chr [1:23] \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" \"May 18 2009\" ...\n",
      "  .. .. .. .. ..$ type                   : chr [1:23] \"RNA\" \"RNA\" \"RNA\" \"RNA\" ...\n",
      "  .. .. .. .. ..$ channel_count          : chr [1:23] \"2\" \"2\" \"2\" \"2\" ...\n",
      "  .. .. .. .. ..$ source_name_ch1        : chr [1:23] \"Strain: MTB strain 1254 Treatment: Control\" \"Strain: MTB strain 1254 Tretment: No Treatment\" \"Strain: MTB strain 1254 Treatment: Control\" \"Strain: MTB strain 1254 Treatment: Control\" ...\n",
      "  .. .. .. .. ..$ organism_ch1           : chr [1:23] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1    : chr [1:23] \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.1  : chr [1:23] \"treatment: Control\" \"treatment: No Treatment\" \"treatment: Control\" \"treatment: Control\" ...\n",
      "  .. .. .. .. ..$ molecule_ch1           : chr [1:23] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1   : chr [1:23] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ label_ch1              : chr [1:23] \"Cy3\" \"Cy3\" \"Cy3\" \"Cy3\" ...\n",
      "  .. .. .. .. ..$ label_protocol_ch1     : chr [1:23] \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ \"TB-Synthesis and Labeling of cDNA SKlab 01; 1. Bring 0.5-5 ?g RNA (Typically 3 ?g) to 11 ?l and add 2 ?l (~2 mg\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch1              : chr [1:23] \"1773\" \"1773\" \"1773\" \"1773\" ...\n",
      "  .. .. .. .. ..$ source_name_ch2        : chr [1:23] \"Strain: MTB strain 1254 Treatment: 0.5 mM H2O2 for 40 min\" \"Strain: MTB strain 1254 Tretment: No Treatment\" \"Strain: MTB strain 1254 Treatment: 50.0 mM H2O2 for 40 min\" \"Strain: MTB strain 1254 Treatment: 0.05 mM H2O2 for 40 min\" ...\n",
      "  .. .. .. .. ..$ organism_ch2           : chr [1:23] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch2    : chr [1:23] \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" \"strain: MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch2.1  : chr [1:23] \"treatment: 0.5 mM H2O2 for 40 min\" \"treatment: No Treatment\" \"treatment: 50.0 mM H2O2 for 40 min\" \"treatment: 0.05 mM H2O2 for 40 min\" ...\n",
      "  .. .. .. .. ..$ growth_protocol_ch2    : chr [1:23] \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ \"Growth Conditions; Parameters that describe the conditions under which the biological sample was grown or propa\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ molecule_ch2           : chr [1:23] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch2   : chr [1:23] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch2.1 : chr [1:23] \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ \"TB-RNA Isolation and Purification SKlab 01; 1. Centrifuge 20-30 ml of early- to mid-log culture (O.D. 0.1-0.2) \"| __truncated__ ...\n",
      "  .. .. .. .. ..$ label_ch2              : chr [1:23] \"Cy5\" \"Cy5\" \"Cy5\" \"Cy5\" ...\n",
      "  .. .. .. .. ..$ label_protocol_ch2     : chr [1:23] \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ \"TB-Slide Preparation SKlab 01; 1. Prepare NaOH-ethanol solution ?  dissolve NaOH in ddH2O 70 g / 280 mL 175 g /\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ label_protocol_ch2.1   : chr [1:23] \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ \"Labeling - Channel 2; Methods for labeling extracted molecules that are used in hybridizations and scanned in C\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch2              : chr [1:23] \"1773\" \"1773\" \"1773\" \"1773\" ...\n",
      "  .. .. .. .. ..$ hyb_protocol           : chr [1:23] \"not provided\" \"not provided\" \"not provided\" \"not provided\" ...\n",
      "  .. .. .. .. ..$ scan_protocol          : chr [1:23] \"GenePix\" \"GenePix\" \"GenePix\" \"GenePix\" ...\n",
      "  .. .. .. .. ..$ scan_protocol.1        : chr [1:23] \"GenePix\" \"GenePix\" \"GenePix\" \"GenePix\" ...\n",
      "  .. .. .. .. ..$ description            : chr [1:23] \"Image: http://smd.stanford.edu/MicroArray/gifs/2009-05/83717.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74755.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74746.gif\" \"Image: http://smd.stanford.edu/MicroArray/gifs/2007-06/74722.gif\" ...\n",
      "  .. .. .. .. ..$ data_processing        : chr [1:23] \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" \"VALUE is Log (base 2) of the ratio of the mean of Channel 2 (usually 635 nm) to Channel 1 (usually 532 nm)\" ...\n",
      "  .. .. .. .. ..$ platform_id            : chr [1:23] \"GPL8562\" \"GPL8562\" \"GPL8562\" \"GPL8562\" ...\n",
      "  .. .. .. .. ..$ contact_name           : chr [1:23] \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" \"Martin,I,Voskuil\" ...\n",
      "  .. .. .. .. ..$ contact_email          : chr [1:23] \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" \"martin.voskuil@ucdenver.edu\" ...\n",
      "  .. .. .. .. ..$ contact_phone          : chr [1:23] \"303-724-4219\" \"303-724-4219\" \"303-724-4219\" \"303-724-4219\" ...\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr [1:23] \"Voskuil Lab\" \"Voskuil Lab\" \"Voskuil Lab\" \"Voskuil Lab\" ...\n",
      "  .. .. .. .. ..$ contact_department     : chr [1:23] \"Microbiology\" \"Microbiology\" \"Microbiology\" \"Microbiology\" ...\n",
      "  .. .. .. .. ..$ contact_institute      : chr [1:23] \"University of Colorado Denver\" \"University of Colorado Denver\" \"University of Colorado Denver\" \"University of Colorado Denver\" ...\n",
      "  .. .. .. .. ..$ contact_address        : chr [1:23] \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" \"12800 E. 19th Ave\" ...\n",
      "  .. .. .. .. ..$ contact_city           : chr [1:23] \"Aurora\" \"Aurora\" \"Aurora\" \"Aurora\" ...\n",
      "  .. .. .. .. ..$ contact_state          : chr [1:23] \"CO\" \"CO\" \"CO\" \"CO\" ...\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr [1:23] \"80045\" \"80045\" \"80045\" \"80045\" ...\n",
      "  .. .. .. .. ..$ contact_country        : chr [1:23] \"USA\" \"USA\" \"USA\" \"USA\" ...\n",
      "  .. .. .. .. ..$ supplementary_file     : chr [1:23] \"NONE\" \"NONE\" \"NONE\" \"NONE\" ...\n",
      "  .. .. .. .. ..$ data_row_count         : chr [1:23] \"4608\" \"4608\" \"4608\" \"4608\" ...\n",
      "  .. .. .. .. ..$ strain:ch1             : chr [1:23] \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ strain:ch2             : chr [1:23] \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" \"MTB strain 1254\" ...\n",
      "  .. .. .. .. ..$ treatment:ch1          : chr [1:23] \"Control\" \"No Treatment\" \"Control\" \"Control\" ...\n",
      "  .. .. .. .. ..$ treatment:ch2          : chr [1:23] \"0.5 mM H2O2 for 40 min\" \"No Treatment\" \"50.0 mM H2O2 for 40 min\" \"0.05 mM H2O2 for 40 min\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ featureData      :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t15 obs. of  3 variables:\n",
      "  .. .. .. .. ..$ Column          : chr [1:15] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Description     : chr [1:15] \"ID =\" \"MetaColumn =\" \"MetaRow =\" \"Column =\" ...\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:15] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t4608 obs. of  15 variables:\n",
      "  .. .. .. .. ..$ ID                             : int [1:4608] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ MetaColumn                     : int [1:4608] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ MetaRow                        : int [1:4608] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ Column                         : int [1:4608] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ Row                            : int [1:4608] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ Reporter Name                  : chr [1:4608] \"RV0001-pcr\" \"RV0003-pcr\" \"RV0005-pcr\" \"RV0007-pcr\" ...\n",
      "  .. .. .. .. ..$ ORF                            : chr [1:4608] \"RV0001\" \"RV0003\" \"RV0005\" \"RV0007\" ...\n",
      "  .. .. .. .. ..$ Reporter Species               : chr [1:4608] \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" \"Mycobacterium tuberculosis\" ...\n",
      "  .. .. .. .. ..$ Reporter BioSequence [length]  : int [1:4608] NA NA NA NA NA NA NA NA NA NA ...\n",
      "  .. .. .. .. ..$ SEQUENCE                       : chr [1:4608] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Repoter Biosequence Type       : chr [1:4608] \"PCR_amplicon\" \"PCR_amplicon\" \"PCR_amplicon\" \"PCR_amplicon\" ...\n",
      "  .. .. .. .. ..$ Repoter BioSequence PolymerType: chr [1:4608] \"DNA\" \"DNA\" \"DNA\" \"DNA\" ...\n",
      "  .. .. .. .. ..$ Reporter Group[role]           : chr [1:4608] \"Experimental\" \"Experimental\" \"Experimental\" \"Experimental\" ...\n",
      "  .. .. .. .. ..$ Reporter ControlType           : chr [1:4608] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ SPOT_ID                        : chr [1:4608] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"featureNames\" \"featureColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ annotation       : chr \"GPL8562\"\n",
      "  .. ..@ protocolData     :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t23 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. ..@ .Data:List of 4\n",
      "  .. .. .. .. ..$ : int [1:3] 4 2 2\n",
      "  .. .. .. .. ..$ : int [1:3] 2 58 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 3 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. ..$ names: chr [1:4] \"R\" \"Biobase\" \"eSet\" \"ExpressionSet\"\n"
     ]
    }
   ],
   "source": [
    "gse <- getGEO(GEO = \"GSE16146\", destdir = \".\")\n",
    "str(gse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2ede7304-c172-4ccd-bb68-21bdafea5993",
   "metadata": {},
   "outputs": [],
   "source": [
    "gse_supp <- getGEOSuppFiles(GEO = \"GSE230269\") # This ended up downloading something...\n",
    "\n",
    "# In any case, it is clear that this is a viable approach "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5ef931fc-fc52-4462-ab68-bcab0f051317",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'data.frame':\t1 obs. of  10 variables:\n",
      " $ size  : num 913123\n",
      " $ isdir : logi FALSE\n",
      " $ mode  : 'octmode' int 644\n",
      " $ mtime : POSIXct, format: \"2024-08-08 08:23:05\"\n",
      " $ ctime : POSIXct, format: \"2024-08-08 08:23:05\"\n",
      " $ atime : POSIXct, format: \"2024-08-08 08:22:57\"\n",
      " $ uid   : int 1001\n",
      " $ gid   : int 100\n",
      " $ uname : chr \"myuser\"\n",
      " $ grname: chr \"users\"\n"
     ]
    }
   ],
   "source": [
    "str(gse_supp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43d340f6-6418-4d07-8b55-9c4ac1399544",
   "metadata": {},
   "source": [
    "I'd like to contrast the above with some other GEO accessions..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9d079ee4-2232-406c-9757-2044b3ed409f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found 1 file(s)\n",
      "\n",
      "GSE178333_series_matrix.txt.gz\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ GSE178333_series_matrix.txt.gz:Formal class 'ExpressionSet' [package \"Biobase\"] with 7 slots\n",
      "  .. ..@ experimentData   :Formal class 'MIAME' [package \"Biobase\"] with 13 slots\n",
      "  .. .. .. ..@ name             : chr \"Mart,Matthias,Lamers\"\n",
      "  .. .. .. ..@ lab              : chr \"\"\n",
      "  .. .. .. ..@ contact          : chr \"m.lamers@erasmusmc.nl\"\n",
      "  .. .. .. ..@ title            : chr \"Human organoid systems reveal in vitro correlates of fitness for SARS-CoV-2 B.1.1.7\"\n",
      "  .. .. .. ..@ abstract         : chr \"A new phase of the COVID-19 pandemic has started as SARS-CoV-2 variants with increased transmissibility are eme\"| __truncated__\n",
      "  .. .. .. ..@ url              : chr \"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE178333\"\n",
      "  .. .. .. ..@ pubMedIds        : chr \"\"\n",
      "  .. .. .. ..@ samples          : list()\n",
      "  .. .. .. ..@ hybridizations   : list()\n",
      "  .. .. .. ..@ normControls     : list()\n",
      "  .. .. .. ..@ preprocessing    : list()\n",
      "  .. .. .. ..@ other            :List of 24\n",
      "  .. .. .. .. ..$ contact_address        : chr \"Wytemaweg 80\"\n",
      "  .. .. .. .. ..$ contact_city           : chr \"Rotterdam\"\n",
      "  .. .. .. .. ..$ contact_country        : chr \"Netherlands\"\n",
      "  .. .. .. .. ..$ contact_department     : chr \"Viroscience\"\n",
      "  .. .. .. .. ..$ contact_email          : chr \"m.lamers@erasmusmc.nl\"\n",
      "  .. .. .. .. ..$ contact_institute      : chr \"Erasmus MC\"\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr \"Haagmans Lab\"\n",
      "  .. .. .. .. ..$ contact_name           : chr \"Mart,Matthias,Lamers\"\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr \"3015CN\"\n",
      "  .. .. .. .. ..$ contributor            : chr \"Mart,,Lamers\\nBart,,Haagmans\"\n",
      "  .. .. .. .. ..$ geo_accession          : chr \"GSE178333\"\n",
      "  .. .. .. .. ..$ last_update_date       : chr \"Jun 13 2024\"\n",
      "  .. .. .. .. ..$ overall_design         : chr \"AT2 cells were grown in 3D in alveolar medium, mechanically disrupted, infected with Bavpat-1 or B.1.1.7\"\n",
      "  .. .. .. .. ..$ platform_id            : chr \"GPL18573\"\n",
      "  .. .. .. .. ..$ platform_taxid         : chr \"9606\"\n",
      "  .. .. .. .. ..$ relation               : chr \"BioProject: https://www.ncbi.nlm.nih.gov/bioproject/PRJNA738504\"\n",
      "  .. .. .. .. ..$ sample_id              : chr \"GSM5387875 GSM5387876 GSM5387877 GSM5387878 GSM5387879 GSM5387880 GSM5387881 GSM5387882 GSM5387883 GSM5387884 G\"| __truncated__\n",
      "  .. .. .. .. ..$ sample_taxid           : chr \"9606\"\n",
      "  .. .. .. .. ..$ status                 : chr \"Public on Jun 13 2024\"\n",
      "  .. .. .. .. ..$ submission_date        : chr \"Jun 16 2021\"\n",
      "  .. .. .. .. ..$ summary                : chr \"A new phase of the COVID-19 pandemic has started as SARS-CoV-2 variants with increased transmissibility are eme\"| __truncated__\n",
      "  .. .. .. .. ..$ supplementary_file     : chr \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nnn/GSE178333/suppl/GSE178333_Lamers_2021_counts_annotated.tsv.gz\"\n",
      "  .. .. .. .. ..$ title                  : chr \"Human organoid systems reveal in vitro correlates of fitness for SARS-CoV-2 B.1.1.7\"\n",
      "  .. .. .. .. ..$ type                   : chr \"Expression profiling by high throughput sequencing\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 2\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr [1:2] \"MIAxE\" \"MIAME\"\n",
      "  .. ..@ assayData        :<environment: 0xaaaafe3966e8> \n",
      "  .. ..@ phenoData        :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t45 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:45] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t12 obs. of  45 variables:\n",
      "  .. .. .. .. ..$ title                  : chr [1:12] \"Alveolar type 2 cells infected with Bavpat-1 rep1\" \"Alveolar type 2 cells infected with Bavpat-1 rep2\" \"Alveolar type 2 cells infected with Bavpat-1 rep3\" \"Alveolar type 2 cells infected with B.1.1.7 rep1\" ...\n",
      "  .. .. .. .. ..$ geo_accession          : chr [1:12] \"GSM5387875\" \"GSM5387876\" \"GSM5387877\" \"GSM5387878\" ...\n",
      "  .. .. .. .. ..$ status                 : chr [1:12] \"Public on Jun 13 2024\" \"Public on Jun 13 2024\" \"Public on Jun 13 2024\" \"Public on Jun 13 2024\" ...\n",
      "  .. .. .. .. ..$ submission_date        : chr [1:12] \"Jun 16 2021\" \"Jun 16 2021\" \"Jun 16 2021\" \"Jun 16 2021\" ...\n",
      "  .. .. .. .. ..$ last_update_date       : chr [1:12] \"Jun 13 2024\" \"Jun 13 2024\" \"Jun 13 2024\" \"Jun 13 2024\" ...\n",
      "  .. .. .. .. ..$ type                   : chr [1:12] \"SRA\" \"SRA\" \"SRA\" \"SRA\" ...\n",
      "  .. .. .. .. ..$ channel_count          : chr [1:12] \"1\" \"1\" \"1\" \"1\" ...\n",
      "  .. .. .. .. ..$ source_name_ch1        : chr [1:12] \"Alveolar type 2 cells\" \"Alveolar type 2 cells\" \"Alveolar type 2 cells\" \"Alveolar type 2 cells\" ...\n",
      "  .. .. .. .. ..$ organism_ch1           : chr [1:12] \"Homo sapiens\" \"Homo sapiens\" \"Homo sapiens\" \"Homo sapiens\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1    : chr [1:12] \"cell type: AT2 cells\" \"cell type: AT2 cells\" \"cell type: AT2 cells\" \"cell type: AT2 cells\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.1  : chr [1:12] \"tissue: Lung\" \"tissue: Lung\" \"tissue: Lung\" \"tissue: Lung\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.2  : chr [1:12] \"treatment: Infection with SARS-CoV-2 Bavpat-1\" \"treatment: Infection with SARS-CoV-2 Bavpat-1\" \"treatment: Infection with SARS-CoV-2 Bavpat-1\" \"treatment: Infection with SARS-CoV-2 B.1.1.7\" ...\n",
      "  .. .. .. .. ..$ treatment_protocol_ch1 : chr [1:12] \"AT2 cells were grown in 3D in alveolar medium, mechanically disrupted, infected with Bavpat-1 or B.1.1.7 in sus\"| __truncated__ \"AT2 cells were grown in 3D in alveolar medium, mechanically disrupted, infected with Bavpat-1 or B.1.1.7 in sus\"| __truncated__ \"AT2 cells were grown in 3D in alveolar medium, mechanically disrupted, infected with Bavpat-1 or B.1.1.7 in sus\"| __truncated__ \"AT2 cells were grown in 3D in alveolar medium, mechanically disrupted, infected with Bavpat-1 or B.1.1.7 in sus\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ growth_protocol_ch1    : chr [1:12] \"AT2 cells were grown in 3D in alveolar medium. Bronchiolar 3D organoids were grown in AO medium, seeded on coll\"| __truncated__ \"AT2 cells were grown in 3D in alveolar medium. Bronchiolar 3D organoids were grown in AO medium, seeded on coll\"| __truncated__ \"AT2 cells were grown in 3D in alveolar medium. Bronchiolar 3D organoids were grown in AO medium, seeded on coll\"| __truncated__ \"AT2 cells were grown in 3D in alveolar medium. Bronchiolar 3D organoids were grown in AO medium, seeded on coll\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ molecule_ch1           : chr [1:12] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1   : chr [1:12] \"Extractions were performed using the trizol RNA extraction protocol.\" \"Extractions were performed using the trizol RNA extraction protocol.\" \"Extractions were performed using the trizol RNA extraction protocol.\" \"Extractions were performed using the trizol RNA extraction protocol.\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.1 : chr [1:12] \"mRNA, amplification and sequencing libraries were performed as described in {Junker et al., 2014}. Briefly, RNA\"| __truncated__ \"mRNA, amplification and sequencing libraries were performed as described in {Junker et al., 2014}. Briefly, RNA\"| __truncated__ \"mRNA, amplification and sequencing libraries were performed as described in {Junker et al., 2014}. Briefly, RNA\"| __truncated__ \"mRNA, amplification and sequencing libraries were performed as described in {Junker et al., 2014}. Briefly, RNA\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.2 : chr [1:12] \"Libraries were sequenced on an Illumina NextSeq500 using 75bp paired end sequencing.\" \"Libraries were sequenced on an Illumina NextSeq500 using 75bp paired end sequencing.\" \"Libraries were sequenced on an Illumina NextSeq500 using 75bp paired end sequencing.\" \"Libraries were sequenced on an Illumina NextSeq500 using 75bp paired end sequencing.\" ...\n",
      "  .. .. .. .. ..$ taxid_ch1              : chr [1:12] \"9606\" \"9606\" \"9606\" \"9606\" ...\n",
      "  .. .. .. .. ..$ description            : chr [1:12] \"AT2_WT_1\" \"AT2_WT_2\" \"AT2_WT_3\" \"AT2_UK_1\" ...\n",
      "  .. .. .. .. ..$ data_processing        : chr [1:12] \"Samples were mapped using MapandGO to hg38 containing SARS-CoV-2 sequences as reference tanscriptome (https://g\"| __truncated__ \"Samples were mapped using MapandGO to hg38 containing SARS-CoV-2 sequences as reference tanscriptome (https://g\"| __truncated__ \"Samples were mapped using MapandGO to hg38 containing SARS-CoV-2 sequences as reference tanscriptome (https://g\"| __truncated__ \"Samples were mapped using MapandGO to hg38 containing SARS-CoV-2 sequences as reference tanscriptome (https://g\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.1      : chr [1:12] \"Data was generated by paired-end sequencing of CEL-seq2 libraries, that contain samble barcode and UMI in Read1\"| __truncated__ \"Data was generated by paired-end sequencing of CEL-seq2 libraries, that contain samble barcode and UMI in Read1\"| __truncated__ \"Data was generated by paired-end sequencing of CEL-seq2 libraries, that contain samble barcode and UMI in Read1\"| __truncated__ \"Data was generated by paired-end sequencing of CEL-seq2 libraries, that contain samble barcode and UMI in Read1\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.2      : chr [1:12] \"Genome_build: Human: Hg38\" \"Genome_build: Human: Hg38\" \"Genome_build: Human: Hg38\" \"Genome_build: Human: Hg38\" ...\n",
      "  .. .. .. .. ..$ data_processing.3      : chr [1:12] \"Supplementary_files_format_and_content: Tsv file containing expression counts for all genes across all samples.\" \"Supplementary_files_format_and_content: Tsv file containing expression counts for all genes across all samples.\" \"Supplementary_files_format_and_content: Tsv file containing expression counts for all genes across all samples.\" \"Supplementary_files_format_and_content: Tsv file containing expression counts for all genes across all samples.\" ...\n",
      "  .. .. .. .. ..$ platform_id            : chr [1:12] \"GPL18573\" \"GPL18573\" \"GPL18573\" \"GPL18573\" ...\n",
      "  .. .. .. .. ..$ contact_name           : chr [1:12] \"Mart,Matthias,Lamers\" \"Mart,Matthias,Lamers\" \"Mart,Matthias,Lamers\" \"Mart,Matthias,Lamers\" ...\n",
      "  .. .. .. .. ..$ contact_email          : chr [1:12] \"m.lamers@erasmusmc.nl\" \"m.lamers@erasmusmc.nl\" \"m.lamers@erasmusmc.nl\" \"m.lamers@erasmusmc.nl\" ...\n",
      "  .. .. .. .. ..$ contact_laboratory     : chr [1:12] \"Haagmans Lab\" \"Haagmans Lab\" \"Haagmans Lab\" \"Haagmans Lab\" ...\n",
      "  .. .. .. .. ..$ contact_department     : chr [1:12] \"Viroscience\" \"Viroscience\" \"Viroscience\" \"Viroscience\" ...\n",
      "  .. .. .. .. ..$ contact_institute      : chr [1:12] \"Erasmus MC\" \"Erasmus MC\" \"Erasmus MC\" \"Erasmus MC\" ...\n",
      "  .. .. .. .. ..$ contact_address        : chr [1:12] \"Wytemaweg 80\" \"Wytemaweg 80\" \"Wytemaweg 80\" \"Wytemaweg 80\" ...\n",
      "  .. .. .. .. ..$ contact_city           : chr [1:12] \"Rotterdam\" \"Rotterdam\" \"Rotterdam\" \"Rotterdam\" ...\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr [1:12] \"3015CN\" \"3015CN\" \"3015CN\" \"3015CN\" ...\n",
      "  .. .. .. .. ..$ contact_country        : chr [1:12] \"Netherlands\" \"Netherlands\" \"Netherlands\" \"Netherlands\" ...\n",
      "  .. .. .. .. ..$ data_row_count         : chr [1:12] \"0\" \"0\" \"0\" \"0\" ...\n",
      "  .. .. .. .. ..$ instrument_model       : chr [1:12] \"Illumina NextSeq 500\" \"Illumina NextSeq 500\" \"Illumina NextSeq 500\" \"Illumina NextSeq 500\" ...\n",
      "  .. .. .. .. ..$ library_selection      : chr [1:12] \"cDNA\" \"cDNA\" \"cDNA\" \"cDNA\" ...\n",
      "  .. .. .. .. ..$ library_source         : chr [1:12] \"transcriptomic\" \"transcriptomic\" \"transcriptomic\" \"transcriptomic\" ...\n",
      "  .. .. .. .. ..$ library_strategy       : chr [1:12] \"RNA-Seq\" \"RNA-Seq\" \"RNA-Seq\" \"RNA-Seq\" ...\n",
      "  .. .. .. .. ..$ relation               : chr [1:12] \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN19731327\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN19731325\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN19731323\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN19731322\" ...\n",
      "  .. .. .. .. ..$ relation.1             : chr [1:12] \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX11169287\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX11169288\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX11169289\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX11169290\" ...\n",
      "  .. .. .. .. ..$ supplementary_file_1   : chr [1:12] \"NONE\" \"NONE\" \"NONE\" \"NONE\" ...\n",
      "  .. .. .. .. ..$ cell type:ch1          : chr [1:12] \"AT2 cells\" \"AT2 cells\" \"AT2 cells\" \"AT2 cells\" ...\n",
      "  .. .. .. .. ..$ tissue:ch1             : chr [1:12] \"Lung\" \"Lung\" \"Lung\" \"Lung\" ...\n",
      "  .. .. .. .. ..$ treatment:ch1          : chr [1:12] \"Infection with SARS-CoV-2 Bavpat-1\" \"Infection with SARS-CoV-2 Bavpat-1\" \"Infection with SARS-CoV-2 Bavpat-1\" \"Infection with SARS-CoV-2 B.1.1.7\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ featureData      :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  3 variables:\n",
      "  .. .. .. .. ..$ Column          : chr(0) \n",
      "  .. .. .. .. ..$ Description     : chr(0) \n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t0 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"featureNames\" \"featureColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ annotation       : chr \"GPL18573\"\n",
      "  .. ..@ protocolData     :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t12 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. ..@ .Data:List of 4\n",
      "  .. .. .. .. ..$ : int [1:3] 4 2 2\n",
      "  .. .. .. .. ..$ : int [1:3] 2 58 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 3 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. ..$ names: chr [1:4] \"R\" \"Biobase\" \"eSet\" \"ExpressionSet\"\n"
     ]
    }
   ],
   "source": [
    "gse <- getGEOfile(GEO = \"GSE178333\", )\n",
    "str(gse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d4c92625-3ecc-474e-8d8b-99fd2abcf266",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>getGEOfile {GEOquery}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Download a file from GEO soft file to the local machine</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>This function simply downloads a SOFT format file associated with the GEO\n",
       "accession number given.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>getGEOfile(\n",
       "  GEO,\n",
       "  destdir = tempdir(),\n",
       "  AnnotGPL = FALSE,\n",
       "  amount = c(\"full\", \"brief\", \"quick\", \"data\")\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr style=\"vertical-align: top;\"><td><code>GEO</code></td>\n",
       "<td>\n",
       "<p>Character string, the GEO accession for download (eg., GDS84,\n",
       "GPL96, GSE2553, or GSM10)</p>\n",
       "</td></tr>\n",
       "<tr style=\"vertical-align: top;\"><td><code>destdir</code></td>\n",
       "<td>\n",
       "<p>Directory in which to store the resulting downloaded file.\n",
       "Defaults to tempdir()</p>\n",
       "</td></tr>\n",
       "<tr style=\"vertical-align: top;\"><td><code>AnnotGPL</code></td>\n",
       "<td>\n",
       "<p>A boolean defaulting to FALSE as to whether or not to use\n",
       "the Annotation GPL information.  These files are nice to use because they\n",
       "contain up-to-date information remapped from Entrez Gene on a regular basis.\n",
       "However, they do not exist for all GPLs; in general, they are only available\n",
       "for GPLs referenced by a GDS</p>\n",
       "</td></tr>\n",
       "<tr style=\"vertical-align: top;\"><td><code>amount</code></td>\n",
       "<td>\n",
       "<p>Amount of information to pull from GEO.  Only applies to GSE,\n",
       "GPL, or GSM.  See details...</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>This function downloads GEO SOFT files based on accession number.  It does\n",
       "not do any parsing.  The first two arguments should be fairly\n",
       "self-explanatory, but the last is based on the input to the acc.cgi url at\n",
       "the geo website.  In the default &quot;full&quot; mode, the entire SOFT format file is\n",
       "downloaded.  Both &quot;brief&quot; and &quot;quick&quot; offer shortened versions of the files,\n",
       "good for &quot;peeking&quot; at the file before a big download on a slow connection.\n",
       "Finally, &quot;data&quot; downloads only the data table part of the SOFT file and is\n",
       "good for downloading a simple EXCEL-like file for use with other programs (a\n",
       "convenience).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>Invisibly returns the full path of the downloaded file.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Sean Davis\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>getGEO</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>\n",
       "  # myfile &lt;- getGEOfile('GDS10')\n",
       "\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>GEOquery</em> version 2.66.0 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{getGEOfile}{Download a file from GEO soft file to the local machine}{getGEOfile}\n",
       "\\keyword{IO}{getGEOfile}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "This function simply downloads a SOFT format file associated with the GEO\n",
       "accession number given.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "getGEOfile(\n",
       "  GEO,\n",
       "  destdir = tempdir(),\n",
       "  AnnotGPL = FALSE,\n",
       "  amount = c(\"full\", \"brief\", \"quick\", \"data\")\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{GEO}] Character string, the GEO accession for download (eg., GDS84,\n",
       "GPL96, GSE2553, or GSM10)\n",
       "\n",
       "\\item[\\code{destdir}] Directory in which to store the resulting downloaded file.\n",
       "Defaults to tempdir()\n",
       "\n",
       "\\item[\\code{AnnotGPL}] A boolean defaulting to FALSE as to whether or not to use\n",
       "the Annotation GPL information.  These files are nice to use because they\n",
       "contain up-to-date information remapped from Entrez Gene on a regular basis.\n",
       "However, they do not exist for all GPLs; in general, they are only available\n",
       "for GPLs referenced by a GDS\n",
       "\n",
       "\\item[\\code{amount}] Amount of information to pull from GEO.  Only applies to GSE,\n",
       "GPL, or GSM.  See details...\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "This function downloads GEO SOFT files based on accession number.  It does\n",
       "not do any parsing.  The first two arguments should be fairly\n",
       "self-explanatory, but the last is based on the input to the acc.cgi url at\n",
       "the geo website.  In the default \"full\" mode, the entire SOFT format file is\n",
       "downloaded.  Both \"brief\" and \"quick\" offer shortened versions of the files,\n",
       "good for \"peeking\" at the file before a big download on a slow connection.\n",
       "Finally, \"data\" downloads only the data table part of the SOFT file and is\n",
       "good for downloading a simple EXCEL-like file for use with other programs (a\n",
       "convenience).\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "Invisibly returns the full path of the downloaded file.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Sean Davis\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{getGEO}{getGEO}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "\n",
       "  # myfile <- getGEOfile('GDS10')\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "getGEOfile              package:GEOquery               R Documentation\n",
       "\n",
       "_\bD_\bo_\bw_\bn_\bl_\bo_\ba_\bd _\ba _\bf_\bi_\bl_\be _\bf_\br_\bo_\bm _\bG_\bE_\bO _\bs_\bo_\bf_\bt _\bf_\bi_\bl_\be _\bt_\bo _\bt_\bh_\be _\bl_\bo_\bc_\ba_\bl _\bm_\ba_\bc_\bh_\bi_\bn_\be\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     This function simply downloads a SOFT format file associated with\n",
       "     the GEO accession number given.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     getGEOfile(\n",
       "       GEO,\n",
       "       destdir = tempdir(),\n",
       "       AnnotGPL = FALSE,\n",
       "       amount = c(\"full\", \"brief\", \"quick\", \"data\")\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "     GEO: Character string, the GEO accession for download (eg., GDS84,\n",
       "          GPL96, GSE2553, or GSM10)\n",
       "\n",
       " destdir: Directory in which to store the resulting downloaded file.\n",
       "          Defaults to tempdir()\n",
       "\n",
       "AnnotGPL: A boolean defaulting to FALSE as to whether or not to use the\n",
       "          Annotation GPL information.  These files are nice to use\n",
       "          because they contain up-to-date information remapped from\n",
       "          Entrez Gene on a regular basis. However, they do not exist\n",
       "          for all GPLs; in general, they are only available for GPLs\n",
       "          referenced by a GDS\n",
       "\n",
       "  amount: Amount of information to pull from GEO.  Only applies to GSE,\n",
       "          GPL, or GSM.  See details...\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     This function downloads GEO SOFT files based on accession number.\n",
       "     It does not do any parsing.  The first two arguments should be\n",
       "     fairly self-explanatory, but the last is based on the input to the\n",
       "     acc.cgi url at the geo website.  In the default \"full\" mode, the\n",
       "     entire SOFT format file is downloaded.  Both \"brief\" and \"quick\"\n",
       "     offer shortened versions of the files, good for \"peeking\" at the\n",
       "     file before a big download on a slow connection. Finally, \"data\"\n",
       "     downloads only the data table part of the SOFT file and is good\n",
       "     for downloading a simple EXCEL-like file for use with other\n",
       "     programs (a convenience).\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     Invisibly returns the full path of the downloaded file.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Sean Davis\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘getGEO’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "       # myfile <- getGEOfile('GDS10')\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "help(getGEOfile\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8ea4b6e8-fc3b-4e49-b196-23c58b21d3ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>getGEOfile {GEOquery}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Download a file from GEO soft file to the local machine</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>This function simply downloads a SOFT format file associated with the GEO\n",
       "accession number given.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>getGEOfile(\n",
       "  GEO,\n",
       "  destdir = tempdir(),\n",
       "  AnnotGPL = FALSE,\n",
       "  amount = c(\"full\", \"brief\", \"quick\", \"data\")\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr style=\"vertical-align: top;\"><td><code>GEO</code></td>\n",
       "<td>\n",
       "<p>Character string, the GEO accession for download (eg., GDS84,\n",
       "GPL96, GSE2553, or GSM10)</p>\n",
       "</td></tr>\n",
       "<tr style=\"vertical-align: top;\"><td><code>destdir</code></td>\n",
       "<td>\n",
       "<p>Directory in which to store the resulting downloaded file.\n",
       "Defaults to tempdir()</p>\n",
       "</td></tr>\n",
       "<tr style=\"vertical-align: top;\"><td><code>AnnotGPL</code></td>\n",
       "<td>\n",
       "<p>A boolean defaulting to FALSE as to whether or not to use\n",
       "the Annotation GPL information.  These files are nice to use because they\n",
       "contain up-to-date information remapped from Entrez Gene on a regular basis.\n",
       "However, they do not exist for all GPLs; in general, they are only available\n",
       "for GPLs referenced by a GDS</p>\n",
       "</td></tr>\n",
       "<tr style=\"vertical-align: top;\"><td><code>amount</code></td>\n",
       "<td>\n",
       "<p>Amount of information to pull from GEO.  Only applies to GSE,\n",
       "GPL, or GSM.  See details...</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>This function downloads GEO SOFT files based on accession number.  It does\n",
       "not do any parsing.  The first two arguments should be fairly\n",
       "self-explanatory, but the last is based on the input to the acc.cgi url at\n",
       "the geo website.  In the default &quot;full&quot; mode, the entire SOFT format file is\n",
       "downloaded.  Both &quot;brief&quot; and &quot;quick&quot; offer shortened versions of the files,\n",
       "good for &quot;peeking&quot; at the file before a big download on a slow connection.\n",
       "Finally, &quot;data&quot; downloads only the data table part of the SOFT file and is\n",
       "good for downloading a simple EXCEL-like file for use with other programs (a\n",
       "convenience).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>Invisibly returns the full path of the downloaded file.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Sean Davis\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>getGEO</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>\n",
       "  # myfile &lt;- getGEOfile('GDS10')\n",
       "\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>GEOquery</em> version 2.66.0 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{getGEOfile}{Download a file from GEO soft file to the local machine}{getGEOfile}\n",
       "\\keyword{IO}{getGEOfile}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "This function simply downloads a SOFT format file associated with the GEO\n",
       "accession number given.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "getGEOfile(\n",
       "  GEO,\n",
       "  destdir = tempdir(),\n",
       "  AnnotGPL = FALSE,\n",
       "  amount = c(\"full\", \"brief\", \"quick\", \"data\")\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{GEO}] Character string, the GEO accession for download (eg., GDS84,\n",
       "GPL96, GSE2553, or GSM10)\n",
       "\n",
       "\\item[\\code{destdir}] Directory in which to store the resulting downloaded file.\n",
       "Defaults to tempdir()\n",
       "\n",
       "\\item[\\code{AnnotGPL}] A boolean defaulting to FALSE as to whether or not to use\n",
       "the Annotation GPL information.  These files are nice to use because they\n",
       "contain up-to-date information remapped from Entrez Gene on a regular basis.\n",
       "However, they do not exist for all GPLs; in general, they are only available\n",
       "for GPLs referenced by a GDS\n",
       "\n",
       "\\item[\\code{amount}] Amount of information to pull from GEO.  Only applies to GSE,\n",
       "GPL, or GSM.  See details...\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "This function downloads GEO SOFT files based on accession number.  It does\n",
       "not do any parsing.  The first two arguments should be fairly\n",
       "self-explanatory, but the last is based on the input to the acc.cgi url at\n",
       "the geo website.  In the default \"full\" mode, the entire SOFT format file is\n",
       "downloaded.  Both \"brief\" and \"quick\" offer shortened versions of the files,\n",
       "good for \"peeking\" at the file before a big download on a slow connection.\n",
       "Finally, \"data\" downloads only the data table part of the SOFT file and is\n",
       "good for downloading a simple EXCEL-like file for use with other programs (a\n",
       "convenience).\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "Invisibly returns the full path of the downloaded file.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Sean Davis\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{getGEO}{getGEO}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "\n",
       "  # myfile <- getGEOfile('GDS10')\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "getGEOfile              package:GEOquery               R Documentation\n",
       "\n",
       "_\bD_\bo_\bw_\bn_\bl_\bo_\ba_\bd _\ba _\bf_\bi_\bl_\be _\bf_\br_\bo_\bm _\bG_\bE_\bO _\bs_\bo_\bf_\bt _\bf_\bi_\bl_\be _\bt_\bo _\bt_\bh_\be _\bl_\bo_\bc_\ba_\bl _\bm_\ba_\bc_\bh_\bi_\bn_\be\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     This function simply downloads a SOFT format file associated with\n",
       "     the GEO accession number given.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     getGEOfile(\n",
       "       GEO,\n",
       "       destdir = tempdir(),\n",
       "       AnnotGPL = FALSE,\n",
       "       amount = c(\"full\", \"brief\", \"quick\", \"data\")\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "     GEO: Character string, the GEO accession for download (eg., GDS84,\n",
       "          GPL96, GSE2553, or GSM10)\n",
       "\n",
       " destdir: Directory in which to store the resulting downloaded file.\n",
       "          Defaults to tempdir()\n",
       "\n",
       "AnnotGPL: A boolean defaulting to FALSE as to whether or not to use the\n",
       "          Annotation GPL information.  These files are nice to use\n",
       "          because they contain up-to-date information remapped from\n",
       "          Entrez Gene on a regular basis. However, they do not exist\n",
       "          for all GPLs; in general, they are only available for GPLs\n",
       "          referenced by a GDS\n",
       "\n",
       "  amount: Amount of information to pull from GEO.  Only applies to GSE,\n",
       "          GPL, or GSM.  See details...\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     This function downloads GEO SOFT files based on accession number.\n",
       "     It does not do any parsing.  The first two arguments should be\n",
       "     fairly self-explanatory, but the last is based on the input to the\n",
       "     acc.cgi url at the geo website.  In the default \"full\" mode, the\n",
       "     entire SOFT format file is downloaded.  Both \"brief\" and \"quick\"\n",
       "     offer shortened versions of the files, good for \"peeking\" at the\n",
       "     file before a big download on a slow connection. Finally, \"data\"\n",
       "     downloads only the data table part of the SOFT file and is good\n",
       "     for downloading a simple EXCEL-like file for use with other\n",
       "     programs (a convenience).\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     Invisibly returns the full path of the downloaded file.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Sean Davis\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘getGEO’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "       # myfile <- getGEOfile('GDS10')\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "help(getGEOfile)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d508cf37-278b-46fd-bba1-691eefc2b5e0",
   "metadata": {},
   "source": [
    "Let's do a more thorough test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a5924e81-b5e1-4eb6-b0c6-222777bf7b1d",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in h(simpleError(msg, call)): error in evaluating the argument 'path' in selecting a method for function 'basename': $ operator is invalid for atomic vectors\n",
     "output_type": "error",
     "traceback": [
      "Error in h(simpleError(msg, call)): error in evaluating the argument 'path' in selecting a method for function 'basename': $ operator is invalid for atomic vectors\nTraceback:\n",
      "1. listAllGEOFiles(accession)",
      "2. listGEOFiles(accession)   # at line 33 of file <text>",
      "3. data.frame(file_name = basename(file_info$localfile), url = file_info$url, \n .     stringsAsFactors = FALSE)   # at line 21-25 of file <text>",
      "4. basename(file_info$localfile)",
      "5. .handleSimpleError(function (cond) \n . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"$ operator is invalid for atomic vectors\", \n .     base::quote(file_info$localfile))",
      "6. h(simpleError(msg, call))"
     ]
    }
   ],
   "source": [
    "listGEOSuppFiles <- function(accession) {\n",
    "  # Get the supplementary files without downloading them\n",
    "  files_info <- getGEOSuppFiles(accession, fetch_files = FALSE)\n",
    "  \n",
    "  # Extract file names and URLs\n",
    "  files <- data.frame(\n",
    "    file_name = basename(rownames(files_info)),\n",
    "    url = files_info$url,\n",
    "    stringsAsFactors = FALSE\n",
    "  )\n",
    "  \n",
    "  return(files)\n",
    "}\n",
    "\n",
    "# Function to list all primary files for a given GEO accession\n",
    "listGEOFiles <- function(accession) {\n",
    "  # Get the primary files without downloading them\n",
    "  file_info <- getGEOfile(accession)\n",
    "  \n",
    "  # Extract file names and URLs\n",
    "  files <- data.frame(\n",
    "    file_name = basename(file_info$localfile),\n",
    "    url = file_info$url,\n",
    "    stringsAsFactors = FALSE\n",
    "  )\n",
    "  \n",
    "  return(files)\n",
    "}\n",
    "\n",
    "# Main function to list all files for a given GEO accession\n",
    "listAllGEOFiles <- function(accession) {\n",
    "  supp_files <- listGEOSuppFiles(accession)\n",
    "  primary_files <- listGEOFiles(accession)\n",
    "  \n",
    "  all_files <- rbind(supp_files, primary_files)\n",
    "  return(all_files)\n",
    "}\n",
    "\n",
    "# Example usage\n",
    "accession <- \"GSE178333\" # Replace with your GEO accession\n",
    "all_files <- listAllGEOFiles(accession)\n",
    "\n",
    "# Print the files\n",
    "print(all_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "54b9e60f-854c-4a32-98ff-673dc8985d82",
   "metadata": {},
   "outputs": [],
   "source": [
    "accession <- \"GSE178333\"\n",
    "files_info <- getGEOSuppFiles(accession, fetch_files = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cc79a006-9fe0-4a5d-a4f9-3b14e4902393",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>fname</th><th scope=col>url</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GSE178333_Lamers_2021_counts_annotated.tsv.gz</td><td>https://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nnn/GSE178333/suppl//GSE178333_Lamers_2021_counts_annotated.tsv.gz</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 2\n",
       "\\begin{tabular}{ll}\n",
       " fname & url\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t GSE178333\\_Lamers\\_2021\\_counts\\_annotated.tsv.gz & https://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nnn/GSE178333/suppl//GSE178333\\_Lamers\\_2021\\_counts\\_annotated.tsv.gz\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 2\n",
       "\n",
       "| fname &lt;chr&gt; | url &lt;chr&gt; |\n",
       "|---|---|\n",
       "| GSE178333_Lamers_2021_counts_annotated.tsv.gz | https://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nnn/GSE178333/suppl//GSE178333_Lamers_2021_counts_annotated.tsv.gz |\n",
       "\n"
      ],
      "text/plain": [
       "  fname                                        \n",
       "1 GSE178333_Lamers_2021_counts_annotated.tsv.gz\n",
       "  url                                                                                                             \n",
       "1 https://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nnn/GSE178333/suppl//GSE178333_Lamers_2021_counts_annotated.tsv.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "files_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "65956ccd-0aa2-401f-af73-df48a92d7564",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using locally cached version of GSE178333 found here:\n",
      "/tmp/RtmpaTBzWH/GSE178333.soft.gz \n",
      "\n"
     ]
    }
   ],
   "source": [
    "file_info <- getGEOfile(accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4e9ec700-377c-43aa-805f-90b091ec32e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/tmp/RtmpaTBzWH/GSE178333.soft.gz'"
      ],
      "text/latex": [
       "'/tmp/RtmpaTBzWH/GSE178333.soft.gz'"
      ],
      "text/markdown": [
       "'/tmp/RtmpaTBzWH/GSE178333.soft.gz'"
      ],
      "text/plain": [
       "[1] \"/tmp/RtmpaTBzWH/GSE178333.soft.gz\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b40f10c-8af3-4b81-9590-bb3e006e4126",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
