{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "97744368-ece8-4ac4-a4d8-2f8cc5dc2aae",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found 1 file(s)\n",
      "\n",
      "GSE178333_series_matrix.txt.gz\n",
      "\n",
      "Using locally cached version: /tmp/RtmpLUBvu9/GSE178333_series_matrix.txt.gz\n",
      "\n",
      "Using locally cached version of GPL18573 found here:\n",
      "/tmp/RtmpLUBvu9/GPL18573.soft.gz \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Load GEOquery package\n",
    "library(GEOquery)\n",
    "library(tidyverse)\n",
    "# Fetch the GEO series data\n",
    "gse <- getGEO(\"GSE178333\", GSEMatrix = TRUE)\n",
    "\n",
    "# Access the metadata\n",
    "# gse is a list, so we need to access the first element\n",
    "gse_data <- gse[[1]]\n",
    "\n",
    "# Get the sample metadata\n",
    "sample_metadata <- pData(gse_data)\n",
    "\n",
    "dir.create(\"~/work//InputData/GSE178333_COVID\")\n",
    "\n",
    "write_csv(sample_metadata, \"~/work/InputData/GSE178333_COVID/meta.csv\")\n",
    "\n",
    "# Should be trivial to optparse this when I need to"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6d3fb4b6-fe18-4354-b5f1-fa2bdfdd50cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Setting options('download.file.method.GEOquery'='auto')\n",
      "\n",
      "Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# I think I'm being gaslit by LLMs...\n",
    "\n",
    "library(GEOquery)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5b73d63e-bdf2-4704-8103-73c095407094",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found 2 file(s)\n",
      "\n",
      "GSE273561-GPL24247_series_matrix.txt.gz\n",
      "\n",
      "GSE273561-GPL24676_series_matrix.txt.gz\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gse <- getGEO(\"GSE273561\", GSEMatrix = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "28f6ed57-b4c4-4c3c-90b2-31a28d64edea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 2\n",
      " $ GSE273561-GPL24247_series_matrix.txt.gz:Formal class 'ExpressionSet' [package \"Biobase\"] with 7 slots\n",
      "  .. ..@ experimentData   :Formal class 'MIAME' [package \"Biobase\"] with 13 slots\n",
      "  .. .. .. ..@ name             : chr \"Lopa,,Mishra\"\n",
      "  .. .. .. ..@ lab              : chr \"\"\n",
      "  .. .. .. ..@ contact          : chr \"lopamishra2@gmail.com\"\n",
      "  .. .. .. ..@ title            : chr \"Aldehydes alter TGF-β signaling and induce obesity and cancer\"\n",
      "  .. .. .. ..@ abstract         : chr \"Obesity and fatty liver diseases-metabolic dysfunction-associated steatotic liver disease (MASLD and MASH) affe\"| __truncated__\n",
      "  .. .. .. ..@ url              : chr \"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE273561\"\n",
      "  .. .. .. ..@ pubMedIds        : chr \"\"\n",
      "  .. .. .. ..@ samples          : list()\n",
      "  .. .. .. ..@ hybridizations   : list()\n",
      "  .. .. .. ..@ normControls     : list()\n",
      "  .. .. .. ..@ preprocessing    : list()\n",
      "  .. .. .. ..@ other            :List of 22\n",
      "  .. .. .. .. ..$ contact_address        : chr \"350 Community Drive\"\n",
      "  .. .. .. .. ..$ contact_city           : chr \"Manhasset\"\n",
      "  .. .. .. .. ..$ contact_country        : chr \"USA\"\n",
      "  .. .. .. .. ..$ contact_email          : chr \"lopamishra2@gmail.com\"\n",
      "  .. .. .. .. ..$ contact_institute      : chr \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\"\n",
      "  .. .. .. .. ..$ contact_name           : chr \"Lopa,,Mishra\"\n",
      "  .. .. .. .. ..$ contact_state          : chr \"NY\"\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr \"11030\"\n",
      "  .. .. .. .. ..$ contributor            : chr \"Xiaochun,,Yang\\nKrishanu,,Bhowmick\\nShuyun,,Rao\\nXiyan,,Xiang\\nLopa,,Mishra\"\n",
      "  .. .. .. .. ..$ geo_accession          : chr \"GSE273561\"\n",
      "  .. .. .. .. ..$ last_update_date       : chr \"Aug 07 2024\"\n",
      "  .. .. .. .. ..$ overall_design         : chr \"Using unbiased transcriptome methods snRNA-seq and RNA-seq to characterize liver metabolism alterations in mice\"| __truncated__\n",
      "  .. .. .. .. ..$ platform_id            : chr \"GPL24247\\nGPL24676\"\n",
      "  .. .. .. .. ..$ platform_taxid         : chr \"9606\\n10090\"\n",
      "  .. .. .. .. ..$ sample_id              : chr \"GSM8432342 GSM8432343 GSM8432344 GSM8432345 GSM8432346 GSM8432347 GSM8432348 GSM8432349 GSM8432350 GSM8432351 G\"| __truncated__\n",
      "  .. .. .. .. ..$ sample_taxid           : chr \"9606\\n10090\"\n",
      "  .. .. .. .. ..$ status                 : chr \"Public on Aug 06 2024\"\n",
      "  .. .. .. .. ..$ submission_date        : chr \"Jul 31 2024\"\n",
      "  .. .. .. .. ..$ summary                : chr \"Obesity and fatty liver diseases-metabolic dysfunction-associated steatotic liver disease (MASLD and MASH) affe\"| __truncated__\n",
      "  .. .. .. .. ..$ supplementary_file     : chr \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE273nnn/GSE273561/suppl/GSE273561_RAW.tar\\nftp://ftp.ncbi.nlm.nih.gov/g\"| __truncated__\n",
      "  .. .. .. .. ..$ title                  : chr \"Aldehydes alter TGF-β signaling and induce obesity and cancer\"\n",
      "  .. .. .. .. ..$ type                   : chr \"Expression profiling by high throughput sequencing\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 2\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr [1:2] \"MIAxE\" \"MIAME\"\n",
      "  .. ..@ assayData        :<environment: 0xaaaafb5c53d8> \n",
      "  .. ..@ phenoData        :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t53 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:53] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t18 obs. of  53 variables:\n",
      "  .. .. .. .. ..$ title                  : chr [1:18] \"Aldh2-/-, snRNAseq\" \"Aldh2-/-Sptbn1+/-, snRNAseq\" \"WT, snRNAseq\" \"Sptbn1+/-, snRNAseq\" ...\n",
      "  .. .. .. .. ..$ geo_accession          : chr [1:18] \"GSM8432342\" \"GSM8432343\" \"GSM8432344\" \"GSM8432345\" ...\n",
      "  .. .. .. .. ..$ status                 : chr [1:18] \"Public on Aug 06 2024\" \"Public on Aug 06 2024\" \"Public on Aug 06 2024\" \"Public on Aug 06 2024\" ...\n",
      "  .. .. .. .. ..$ submission_date        : chr [1:18] \"Jul 31 2024\" \"Jul 31 2024\" \"Jul 31 2024\" \"Jul 31 2024\" ...\n",
      "  .. .. .. .. ..$ last_update_date       : chr [1:18] \"Aug 06 2024\" \"Aug 06 2024\" \"Aug 06 2024\" \"Aug 06 2024\" ...\n",
      "  .. .. .. .. ..$ type                   : chr [1:18] \"SRA\" \"SRA\" \"SRA\" \"SRA\" ...\n",
      "  .. .. .. .. ..$ channel_count          : chr [1:18] \"1\" \"1\" \"1\" \"1\" ...\n",
      "  .. .. .. .. ..$ source_name_ch1        : chr [1:18] \"liver\" \"liver\" \"liver\" \"liver\" ...\n",
      "  .. .. .. .. ..$ organism_ch1           : chr [1:18] \"Mus musculus\" \"Mus musculus\" \"Mus musculus\" \"Mus musculus\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1    : chr [1:18] \"cell type: liver cells\" \"cell type: liver cells\" \"cell type: liver cells\" \"cell type: liver cells\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.1  : chr [1:18] \"tissue: liver\" \"tissue: liver\" \"tissue: liver\" \"tissue: liver\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.2  : chr [1:18] \"strain: C57BL/6\" \"strain: C57BL/6\" \"strain: C57BL/6\" \"strain: C57BL/6\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.3  : chr [1:18] \"genotype: Aldh2-/-\" \"genotype: Aldh2-/-Sptbn1+/-\" \"genotype: WT\" \"genotype: Sptbn1+/-\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.4  : chr [1:18] \"age: 12 months\" \"age: 12 months\" \"age: 12 months\" \"age: 12 months\" ...\n",
      "  .. .. .. .. ..$ molecule_ch1           : chr [1:18] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1   : chr [1:18] \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.1 : chr [1:18] \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.2 : chr [1:18] \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.3 : chr [1:18] \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch1              : chr [1:18] \"10090\" \"10090\" \"10090\" \"10090\" ...\n",
      "  .. .. .. .. ..$ description            : chr [1:18] \"A624\" \"A625\" \"A757\" \"A894\" ...\n",
      "  .. .. .. .. ..$ description.1          : chr [1:18] \"10x Genomics\" \"10x Genomics\" \"10x Genomics\" \"10x Genomics\" ...\n",
      "  .. .. .. .. ..$ data_processing        : chr [1:18] \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.1      : chr [1:18] \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" ...\n",
      "  .. .. .. .. ..$ data_processing.2      : chr [1:18] \"Assembly: mm10\" \"Assembly: mm10\" \"Assembly: mm10\" \"Assembly: mm10\" ...\n",
      "  .. .. .. .. ..$ data_processing.3      : chr [1:18] \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" ...\n",
      "  .. .. .. .. ..$ data_processing.4      : chr [1:18] \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.5      : chr [1:18] \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.6      : chr [1:18] \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" ...\n",
      "  .. .. .. .. ..$ platform_id            : chr [1:18] \"GPL24247\" \"GPL24247\" \"GPL24247\" \"GPL24247\" ...\n",
      "  .. .. .. .. ..$ contact_name           : chr [1:18] \"Lopa,,Mishra\" \"Lopa,,Mishra\" \"Lopa,,Mishra\" \"Lopa,,Mishra\" ...\n",
      "  .. .. .. .. ..$ contact_email          : chr [1:18] \"lopamishra2@gmail.com\" \"lopamishra2@gmail.com\" \"lopamishra2@gmail.com\" \"lopamishra2@gmail.com\" ...\n",
      "  .. .. .. .. ..$ contact_institute      : chr [1:18] \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" ...\n",
      "  .. .. .. .. ..$ contact_address        : chr [1:18] \"350 Community Drive\" \"350 Community Drive\" \"350 Community Drive\" \"350 Community Drive\" ...\n",
      "  .. .. .. .. ..$ contact_city           : chr [1:18] \"Manhasset\" \"Manhasset\" \"Manhasset\" \"Manhasset\" ...\n",
      "  .. .. .. .. ..$ contact_state          : chr [1:18] \"NY\" \"NY\" \"NY\" \"NY\" ...\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr [1:18] \"11030\" \"11030\" \"11030\" \"11030\" ...\n",
      "  .. .. .. .. ..$ contact_country        : chr [1:18] \"USA\" \"USA\" \"USA\" \"USA\" ...\n",
      "  .. .. .. .. ..$ data_row_count         : chr [1:18] \"0\" \"0\" \"0\" \"0\" ...\n",
      "  .. .. .. .. ..$ instrument_model       : chr [1:18] \"Illumina NovaSeq 6000\" \"Illumina NovaSeq 6000\" \"Illumina NovaSeq 6000\" \"Illumina NovaSeq 6000\" ...\n",
      "  .. .. .. .. ..$ library_selection      : chr [1:18] \"cDNA\" \"cDNA\" \"cDNA\" \"cDNA\" ...\n",
      "  .. .. .. .. ..$ library_source         : chr [1:18] \"transcriptomic single cell\" \"transcriptomic single cell\" \"transcriptomic single cell\" \"transcriptomic single cell\" ...\n",
      "  .. .. .. .. ..$ library_strategy       : chr [1:18] \"RNA-Seq\" \"RNA-Seq\" \"RNA-Seq\" \"RNA-Seq\" ...\n",
      "  .. .. .. .. ..$ relation               : chr [1:18] \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940209\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940208\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940207\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940206\" ...\n",
      "  .. .. .. .. ..$ relation.1             : chr [1:18] \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524103\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524104\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524105\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524106\" ...\n",
      "  .. .. .. .. ..$ supplementary_file_1   : chr [1:18] \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432342/suppl/GSM8432342_A624_filtered_feature_bc_matrix.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432343/suppl/GSM8432343_A625_filtered_feature_bc_matrix.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432344/suppl/GSM8432344_A757_filtered_feature_bc_matrix.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432345/suppl/GSM8432345_A894_filtered_feature_bc_matrix.h5\" ...\n",
      "  .. .. .. .. ..$ supplementary_file_2   : chr [1:18] \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432342/suppl/GSM8432342_A624_molecule_info.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432343/suppl/GSM8432343_A625_molecule_info.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432344/suppl/GSM8432344_A757_molecule_info.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432345/suppl/GSM8432345_A894_molecule_info.h5\" ...\n",
      "  .. .. .. .. ..$ supplementary_file_3   : chr [1:18] \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432342/suppl/GSM8432342_A624_raw_feature_bc_matrix.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432343/suppl/GSM8432343_A625_raw_feature_bc_matrix.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432344/suppl/GSM8432344_A757_raw_feature_bc_matrix.h5\" \"ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM8432nnn/GSM8432345/suppl/GSM8432345_A894_raw_feature_bc_matrix.h5\" ...\n",
      "  .. .. .. .. ..$ age:ch1                : chr [1:18] \"12 months\" \"12 months\" \"12 months\" \"12 months\" ...\n",
      "  .. .. .. .. ..$ cell type:ch1          : chr [1:18] \"liver cells\" \"liver cells\" \"liver cells\" \"liver cells\" ...\n",
      "  .. .. .. .. ..$ genotype:ch1           : chr [1:18] \"Aldh2-/-\" \"Aldh2-/-Sptbn1+/-\" \"WT\" \"Sptbn1+/-\" ...\n",
      "  .. .. .. .. ..$ strain:ch1             : chr [1:18] \"C57BL/6\" \"C57BL/6\" \"C57BL/6\" \"C57BL/6\" ...\n",
      "  .. .. .. .. ..$ tissue:ch1             : chr [1:18] \"liver\" \"liver\" \"liver\" \"liver\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ featureData      :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  3 variables:\n",
      "  .. .. .. .. ..$ Column          : chr(0) \n",
      "  .. .. .. .. ..$ Description     : chr(0) \n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t0 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"featureNames\" \"featureColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ annotation       : chr \"GPL24247\"\n",
      "  .. ..@ protocolData     :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t18 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. ..@ .Data:List of 4\n",
      "  .. .. .. .. ..$ : int [1:3] 4 2 2\n",
      "  .. .. .. .. ..$ : int [1:3] 2 58 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 3 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. ..$ names: chr [1:4] \"R\" \"Biobase\" \"eSet\" \"ExpressionSet\"\n",
      " $ GSE273561-GPL24676_series_matrix.txt.gz:Formal class 'ExpressionSet' [package \"Biobase\"] with 7 slots\n",
      "  .. ..@ experimentData   :Formal class 'MIAME' [package \"Biobase\"] with 13 slots\n",
      "  .. .. .. ..@ name             : chr \"Lopa,,Mishra\"\n",
      "  .. .. .. ..@ lab              : chr \"\"\n",
      "  .. .. .. ..@ contact          : chr \"lopamishra2@gmail.com\"\n",
      "  .. .. .. ..@ title            : chr \"Aldehydes alter TGF-β signaling and induce obesity and cancer\"\n",
      "  .. .. .. ..@ abstract         : chr \"Obesity and fatty liver diseases-metabolic dysfunction-associated steatotic liver disease (MASLD and MASH) affe\"| __truncated__\n",
      "  .. .. .. ..@ url              : chr \"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE273561\"\n",
      "  .. .. .. ..@ pubMedIds        : chr \"\"\n",
      "  .. .. .. ..@ samples          : list()\n",
      "  .. .. .. ..@ hybridizations   : list()\n",
      "  .. .. .. ..@ normControls     : list()\n",
      "  .. .. .. ..@ preprocessing    : list()\n",
      "  .. .. .. ..@ other            :List of 22\n",
      "  .. .. .. .. ..$ contact_address        : chr \"350 Community Drive\"\n",
      "  .. .. .. .. ..$ contact_city           : chr \"Manhasset\"\n",
      "  .. .. .. .. ..$ contact_country        : chr \"USA\"\n",
      "  .. .. .. .. ..$ contact_email          : chr \"lopamishra2@gmail.com\"\n",
      "  .. .. .. .. ..$ contact_institute      : chr \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\"\n",
      "  .. .. .. .. ..$ contact_name           : chr \"Lopa,,Mishra\"\n",
      "  .. .. .. .. ..$ contact_state          : chr \"NY\"\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr \"11030\"\n",
      "  .. .. .. .. ..$ contributor            : chr \"Xiaochun,,Yang\\nKrishanu,,Bhowmick\\nShuyun,,Rao\\nXiyan,,Xiang\\nLopa,,Mishra\"\n",
      "  .. .. .. .. ..$ geo_accession          : chr \"GSE273561\"\n",
      "  .. .. .. .. ..$ last_update_date       : chr \"Aug 07 2024\"\n",
      "  .. .. .. .. ..$ overall_design         : chr \"Using unbiased transcriptome methods snRNA-seq and RNA-seq to characterize liver metabolism alterations in mice\"| __truncated__\n",
      "  .. .. .. .. ..$ platform_id            : chr \"GPL24247\\nGPL24676\"\n",
      "  .. .. .. .. ..$ platform_taxid         : chr \"9606\\n10090\"\n",
      "  .. .. .. .. ..$ sample_id              : chr \"GSM8432342 GSM8432343 GSM8432344 GSM8432345 GSM8432346 GSM8432347 GSM8432348 GSM8432349 GSM8432350 GSM8432351 G\"| __truncated__\n",
      "  .. .. .. .. ..$ sample_taxid           : chr \"9606\\n10090\"\n",
      "  .. .. .. .. ..$ status                 : chr \"Public on Aug 06 2024\"\n",
      "  .. .. .. .. ..$ submission_date        : chr \"Jul 31 2024\"\n",
      "  .. .. .. .. ..$ summary                : chr \"Obesity and fatty liver diseases-metabolic dysfunction-associated steatotic liver disease (MASLD and MASH) affe\"| __truncated__\n",
      "  .. .. .. .. ..$ supplementary_file     : chr \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE273nnn/GSE273561/suppl/GSE273561_RAW.tar\\nftp://ftp.ncbi.nlm.nih.gov/g\"| __truncated__\n",
      "  .. .. .. .. ..$ title                  : chr \"Aldehydes alter TGF-β signaling and induce obesity and cancer\"\n",
      "  .. .. .. .. ..$ type                   : chr \"Expression profiling by high throughput sequencing\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 2\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr [1:2] \"MIAxE\" \"MIAME\"\n",
      "  .. ..@ assayData        :<environment: 0xaaaafefe2930> \n",
      "  .. ..@ phenoData        :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t49 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:49] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t9 obs. of  49 variables:\n",
      "  .. .. .. .. ..$ title                  : chr [1:9] \"siCtrl, replicate 1, RNAseq\" \"siCtrl, replicate 2, RNAseq\" \"siCtrl, replicate 3, RNAseq\" \"25nM siSptbn1, replicate 1, RNAseq\" ...\n",
      "  .. .. .. .. ..$ geo_accession          : chr [1:9] \"GSM8432346\" \"GSM8432347\" \"GSM8432348\" \"GSM8432349\" ...\n",
      "  .. .. .. .. ..$ status                 : chr [1:9] \"Public on Aug 06 2024\" \"Public on Aug 06 2024\" \"Public on Aug 06 2024\" \"Public on Aug 06 2024\" ...\n",
      "  .. .. .. .. ..$ submission_date        : chr [1:9] \"Jul 31 2024\" \"Jul 31 2024\" \"Jul 31 2024\" \"Jul 31 2024\" ...\n",
      "  .. .. .. .. ..$ last_update_date       : chr [1:9] \"Aug 06 2024\" \"Aug 06 2024\" \"Aug 06 2024\" \"Aug 06 2024\" ...\n",
      "  .. .. .. .. ..$ type                   : chr [1:9] \"SRA\" \"SRA\" \"SRA\" \"SRA\" ...\n",
      "  .. .. .. .. ..$ channel_count          : chr [1:9] \"1\" \"1\" \"1\" \"1\" ...\n",
      "  .. .. .. .. ..$ source_name_ch1        : chr [1:9] \"liver\" \"liver\" \"liver\" \"liver\" ...\n",
      "  .. .. .. .. ..$ organism_ch1           : chr [1:9] \"Homo sapiens\" \"Homo sapiens\" \"Homo sapiens\" \"Homo sapiens\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1    : chr [1:9] \"cell type: coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" \"cell type: coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" \"cell type: coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" \"cell type: coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.1  : chr [1:9] \"tissue: liver\" \"tissue: liver\" \"tissue: liver\" \"tissue: liver\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.2  : chr [1:9] \"treatment: siControl\" \"treatment: siControl\" \"treatment: siControl\" \"treatment: 25nM siSptbn1\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.3  : chr [1:9] \"time: 96h\" \"time: 96h\" \"time: 96h\" \"time: 96h\" ...\n",
      "  .. .. .. .. ..$ molecule_ch1           : chr [1:9] \"total RNA\" \"total RNA\" \"total RNA\" \"total RNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1   : chr [1:9] \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ \"snRNA-seq: Nuclei were isolated by lysing frozen mouse livers with a pestle and cold lysing buffer (10mM Tris-C\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.1 : chr [1:9] \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ \"snRNA-seq: Nuclei were labeled with 7-AAD (Sigma) and sorted on a BD FACS Aria to purify nuclei from the remain\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.2 : chr [1:9] \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ \"RNA-seq: RNA was extracted from liver tissues using the RNeasy Plus Mini kit. RNA quality and concentration wer\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1.3 : chr [1:9] \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ \"RAN-seq: About 1 μg of RNA was used to generate barcoded sequencing libraries. Novogene Corporation performed l\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ taxid_ch1              : chr [1:9] \"9606\" \"9606\" \"9606\" \"9606\" ...\n",
      "  .. .. .. .. ..$ description            : chr [1:9] \"P2A1\" \"P2A2\" \"P2A3\" \"P2B1\" ...\n",
      "  .. .. .. .. ..$ description.1          : chr [1:9] \"TPM_values_siCtrl_vs_25nM_siSptbn1_vs_50nM_siSptbn1_96h.txt\" \"TPM_values_siCtrl_vs_25nM_siSptbn1_vs_50nM_siSptbn1_96h.txt\" \"TPM_values_siCtrl_vs_25nM_siSptbn1_vs_50nM_siSptbn1_96h.txt\" \"TPM_values_siCtrl_vs_25nM_siSptbn1_vs_50nM_siSptbn1_96h.txt\" ...\n",
      "  .. .. .. .. ..$ data_processing        : chr [1:9] \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ \"snRNA-seq: The demutiplexing, barcoded processing, gene counting and aggregation were made using the Cell Range\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.1      : chr [1:9] \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" \"snRNA-seq: The preprocessing, QC, normalizing, clustering, and visulization were made using the Seurat v4.\" ...\n",
      "  .. .. .. .. ..$ data_processing.2      : chr [1:9] \"Assembly: mm10\" \"Assembly: mm10\" \"Assembly: mm10\" \"Assembly: mm10\" ...\n",
      "  .. .. .. .. ..$ data_processing.3      : chr [1:9] \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" \"Supplementary files format and content: snRNA-seq: H5, with molecule, feature and matrix files for each sample\" ...\n",
      "  .. .. .. .. ..$ data_processing.4      : chr [1:9] \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ \"RNA-seq: Sequence reads were trimmed to remove possible adapter sequences and nucleotides with poor quality usi\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.5      : chr [1:9] \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ \"RNA-seq: After extraction of gene hit counts, the gene hit counts table was used for downstream differential ex\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ data_processing.6      : chr [1:9] \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" \"Supplementary files format and content: RNA-seq: tab-delimited text file includes TPM values for each sample\" ...\n",
      "  .. .. .. .. ..$ platform_id            : chr [1:9] \"GPL24676\" \"GPL24676\" \"GPL24676\" \"GPL24676\" ...\n",
      "  .. .. .. .. ..$ contact_name           : chr [1:9] \"Lopa,,Mishra\" \"Lopa,,Mishra\" \"Lopa,,Mishra\" \"Lopa,,Mishra\" ...\n",
      "  .. .. .. .. ..$ contact_email          : chr [1:9] \"lopamishra2@gmail.com\" \"lopamishra2@gmail.com\" \"lopamishra2@gmail.com\" \"lopamishra2@gmail.com\" ...\n",
      "  .. .. .. .. ..$ contact_institute      : chr [1:9] \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" \"The Feinstein Institutes for Medical Research & Cold Spring Harbor Laboratory\" ...\n",
      "  .. .. .. .. ..$ contact_address        : chr [1:9] \"350 Community Drive\" \"350 Community Drive\" \"350 Community Drive\" \"350 Community Drive\" ...\n",
      "  .. .. .. .. ..$ contact_city           : chr [1:9] \"Manhasset\" \"Manhasset\" \"Manhasset\" \"Manhasset\" ...\n",
      "  .. .. .. .. ..$ contact_state          : chr [1:9] \"NY\" \"NY\" \"NY\" \"NY\" ...\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr [1:9] \"11030\" \"11030\" \"11030\" \"11030\" ...\n",
      "  .. .. .. .. ..$ contact_country        : chr [1:9] \"USA\" \"USA\" \"USA\" \"USA\" ...\n",
      "  .. .. .. .. ..$ data_row_count         : chr [1:9] \"0\" \"0\" \"0\" \"0\" ...\n",
      "  .. .. .. .. ..$ instrument_model       : chr [1:9] \"Illumina NovaSeq 6000\" \"Illumina NovaSeq 6000\" \"Illumina NovaSeq 6000\" \"Illumina NovaSeq 6000\" ...\n",
      "  .. .. .. .. ..$ library_selection      : chr [1:9] \"cDNA\" \"cDNA\" \"cDNA\" \"cDNA\" ...\n",
      "  .. .. .. .. ..$ library_source         : chr [1:9] \"transcriptomic\" \"transcriptomic\" \"transcriptomic\" \"transcriptomic\" ...\n",
      "  .. .. .. .. ..$ library_strategy       : chr [1:9] \"RNA-Seq\" \"RNA-Seq\" \"RNA-Seq\" \"RNA-Seq\" ...\n",
      "  .. .. .. .. ..$ relation               : chr [1:9] \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940205\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940204\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940203\" \"BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN42940202\" ...\n",
      "  .. .. .. .. ..$ relation.1             : chr [1:9] \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524107\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524108\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524109\" \"SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX25524110\" ...\n",
      "  .. .. .. .. ..$ supplementary_file_1   : chr [1:9] \"NONE\" \"NONE\" \"NONE\" \"NONE\" ...\n",
      "  .. .. .. .. ..$ cell type:ch1          : chr [1:9] \"coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" \"coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" \"coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" \"coculture human liver cells: hepatocyte, Kupffer cell, stellate cell\" ...\n",
      "  .. .. .. .. ..$ time:ch1               : chr [1:9] \"96h\" \"96h\" \"96h\" \"96h\" ...\n",
      "  .. .. .. .. ..$ tissue:ch1             : chr [1:9] \"liver\" \"liver\" \"liver\" \"liver\" ...\n",
      "  .. .. .. .. ..$ treatment:ch1          : chr [1:9] \"siControl\" \"siControl\" \"siControl\" \"25nM siSptbn1\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ featureData      :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  3 variables:\n",
      "  .. .. .. .. ..$ Column          : chr(0) \n",
      "  .. .. .. .. ..$ Description     : chr(0) \n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t0 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"featureNames\" \"featureColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ annotation       : chr \"GPL24676\"\n",
      "  .. ..@ protocolData     :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t9 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. ..@ .Data:List of 4\n",
      "  .. .. .. .. ..$ : int [1:3] 4 2 2\n",
      "  .. .. .. .. ..$ : int [1:3] 2 58 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 3 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. ..$ names: chr [1:4] \"R\" \"Biobase\" \"eSet\" \"ExpressionSet\"\n"
     ]
    }
   ],
   "source": [
    "str(gse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "11ec3c91-97a8-4e89-b6b0-91b9dfd87a8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GSE273561-GPL24247_series_matrix.txt.gz'</li><li>'GSE273561-GPL24676_series_matrix.txt.gz'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'GSE273561-GPL24247\\_series\\_matrix.txt.gz'\n",
       "\\item 'GSE273561-GPL24676\\_series\\_matrix.txt.gz'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'GSE273561-GPL24247_series_matrix.txt.gz'\n",
       "2. 'GSE273561-GPL24676_series_matrix.txt.gz'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"GSE273561-GPL24247_series_matrix.txt.gz\"\n",
       "[2] \"GSE273561-GPL24676_series_matrix.txt.gz\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(gse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bfdb20f-95fa-45f3-8dd1-c4f827672887",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
