GSE123821

Status	Public on Dec 14, 2018
Title	RNA-seq analysis for resistance to PI3K inhibition in Glioblastoma (GBM)
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	The goal of this study is to identify the pathway alterations driving the adaptive resistance to PI3K inhibition in GBM. We generated the resistant cell lines through a patient-derived in vivo glioma sphere-forming cell (GSC) model. We performed RNA-seq on the paired GSC samples including the parental and resistant groups. Libraries were sequenced with an average coverage for each tumor of 50x on the Hiseq4000 platform from Illumina, using 76 nt pair-ended reads. RNA-seq raw data were pre-processed using PRADA. PRADA aligned RNA-seq reads to a composite reference database composed of whole genome and transcriptome sequences; we used the hg19 human genome assembly, together with the Ensembl64 transcriptome version. Transcripts were filtered for size and protein-coding genes. Expression data were normalized to reads per kilobase per million reads, and these values were log2-transformed for further analyses.
4 samples



GSE198258
Status	Public on Jul 06, 2022
Title	Bulk RNAseq analysis of hypertrophic cardiomyopathy (HCM) hiPSC-derived cardiomyocytes (hiPSC-CMs) and the isogenic control
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	To investigate HCM mutation-associated molecular details at the early stage of disease development, we performed bulk RNA-seq analysis at Day 15 of differentiation using RNA isolated from the isogenic control and mutant hiPSC-CMs. We performed pair-end sequencing with >30 million reads per sample with clean read counts of more than 96% for each sample.
 	Overall design	mRNA profiles of Day15 MYH7 R723C with MYH6 R725C mutant hiPSC-CMs and isogenic control
4 samples


GSE102674
Status	Public on May 01, 2019
Title	Effect of mTORC1 and 2 inhibition on myofibroblast global gene expression
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	Background: Mechanistic target of rapamycin (mTOR) is a nodal serine/threonine protein kinase critical for the control of fundamental cellular processes. Dysregulated mTOR signalling has been shown in cancer, COPD and idiopathic pulmonary fibrosis (IPF). mTOR forms the catalytic subunit of 2 protein complexes: mTORC1 and mTORC2 which differ in upstream inputs, downstream effects and sensitivity to the allosteric inhibitor rapamycin.
Aim: To compare the effects of the mTORC1 inhibitor rapamycin with an ATP-competitive inhibitor of mTORC1/2, AZD8055, on TGFβ1 induced transcriptional responses during myofibroblast differentiation by RNA-Seq analysis.
Methods: Primary human lung fibroblasts were treated with TGFβ1 for 24h to induce myofibroblast differentiation and extra cellular matrix (ECM) production in the presence of rapamycin or AZD8055. Fibroblasts were also treated with rapamycin or AZD8055 without TGFβ1. Total mRNA was analysed by RNA-Seq.
Results: On a global transcriptomic level, TGFβ1 significantly affected the expression of ~4,200 genes by more than 1.5 fold, of which ~1,100 changes were reversed by pan-mTORC1/2 inhibition, while being insensitive to rapamycin. Only 15 gene changes were reversed exclusively by rapamycin. Analysis of the rapamycin-insensitive, mTOR sensitive myofibroblast transcriptome revealed ECM-receptor interaction, metabolism and actin cytoskeleton regulation as major affected pathways.
Conclusions: Our data support the conclusion that therapeutic approaches aimed at interfering with TGFb1 induced myofibroblast differentiation and ECM production require dual mTOR1/2 inhibition and may, in part, explain the lack of clinical efficacy of everolimus in the context of IPF.

Overall design	Primary human lung fibroblasts were treated with TGFβ1 for 24h in the presence of rapamycin or AZD8055. Each treatment was done in quadruplicates, for a total of 16 samples. Untreated cells were used as control. Paired-end RNA sequencing was performed with the NextSeq sequencing platform (Illumina, San Diego, CA, USA). Each sample was run on 4 different lanes.
24 samples



GSE93709
Status	Public on Jan 17, 2018
Title	Transcription factors operate across disease loci: EBNA2 in autoimmunity
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	Explaining the genetics of many diseases is challenging because most associations localize to regulatory regions. We present a novel computational method for discovering disease-driving mechanisms acting across multiple disease-associated, non-coding genomic regions. Application to a matrix of 213 phenotypes and 1,544 transcription factor (TF) binding datasets identifies 2,264 significant associations for hundreds of TFs in 92 phenotypes, including prostate and breast cancers. Strikingly, nearly half of the systemic lupus erythematosus risk loci are occupied by the Epstein-Barr virus EBNA2 protein and 24 human TFs, revealing an important gene-environment interaction. EBNA2-anchored associations also exist in multiple sclerosis, rheumatoid arthritis, inflammatory bowel disease, type 1 diabetes, juvenile idiopathic arthritis, and celiac disease. Instances of allele-dependent DNA binding and downstream effects on gene expression at plausibly causal autoimmune variants support a genetic mechanism of pathogenesis centered on EBNA2. Our results nominate mechanisms operating across disease risk loci, suggesting new paradigms of disease origins.
2 samples


GSE133858
Status	Public on Jul 05, 2021
Title	RNA-sequencing data from PLC/PRF/5 cells with NC or USP22 knockdown (shUSP22) under hypoxic condition
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	We use RNA-sequencing to evaluate the effect of USP22 knockdown on the gene changes in HCC cells

Overall design	examing two replicates of each group under hypoxic condition
4 samples (ok this is getting kinda ridiculous)


GSE60939
Status	Public on Aug 30, 2014
Title	Potent antitumor activity of Cabozantinib, a c-MET and VEGFR2 Inhibitor, in a Colorectal Cancer Patient-derived Tumor Explant Model
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	Anti-angiogenic therapy is commonly used for the treatment of CRC. Although patients derive some clinical benefit, treatment resistance inevitably occurs. The MET signaling pathway has been proposed to be a major contributor of resistance to anti-angiogenic therapy. MET is upregulated in response to VEGF pathway inhibition and plays an essential role in tumorigenesis and progression of tumors. In this study we set out to determine the efficacy of cabozantinib in a preclinical CRC PDTX model. We demonstrate potent inhibitory effects on tumor growth in 80% of tumors treated. The greatest antitumor effects were observed in tumors that possess a mutation in the PIK3CA gene. The underlying antitumor mechanisms of cabozantinib consisted of inhibition of angiogenesis and Akt activation and significantly decreased expression of genes involved in the PI3K pathway. These findings support further evaluation of cabozantinib in patients with CRC. PIK3CA mutation as a predictive biomarker of sensitivity is intriguing and warrants further elucidation. A clinical trial of cabozantinib in refractory metastatic CRC is being activated.
6 samples


Series GSE186063
Status	Public on Dec 28, 2022
Title	Multi-omics profiling reveals the interactions of host defense and hyper-keratinization in psoriasis
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	Objectives: To understand the crosstalk between the immune system and keratinization in psoriatic skin, using a systems biology approach based on transcriptomics, proteomics and microbiome profiling.
Methods: We collected the skin tissue biopsies and swabs in both lesion and non-lesion skin of 13 patients with psoriasis (PsO), 15 patients with psoriatic arthritis (PsA), and healthy skin from 12 patients with ankylosing spondylitis (AS). We performed transcriptome sequencing and metagenomics profiling on the local skin sites to study the similarities and differences in the molecular profiles between the three conditions. To assess the systemic nature of the disease, we performed a high-throughput proteome profiling to study the profiles of proteins circulating in the serum of the same donors.
Results: We found that lesion and non-lesional samples were remarkably different in terms of their transcriptome profiles. Functional annotation of differentially expressed genes (DEGs) showed a major enrichment in neutrophil activation. By using co-expression gene networks, we identified a gene module that was associated with local psoriasis severity at the site of biopsy. From this module, we extracted a “core” set of genes that were functionally involved in neutrophil activation, epidermal cell differentiation and response to bacteria. Skin microbiome analysis revealed that the abundance of Enhydrobacter, Micrococcus and Leptotrichia were significantly correlated with the “core network” of genes. We further identified 39 circulating proteins from the serum that were significantly correlated with the corresponding local skin gene expression, highlighting systemic aberrations due to skin disease. Integration of “core” genes identified from skin with circulating protein profiles revealed PI3 as a key biomarker for psoriasis.
Conclusions: We identified a core network that regulates inflammation and hyper-keratinization in psoriatic skin, and is associated with local disease severity and microbiome composition. Multi-omics analysis identified PI3 as a psoriasis-specific biomarker for disease severity and potential target for treatment strategies.

Overall design	Skin transcriptome profiles of both lesion and non-lesion of 13 patients with psoriasis (PsO), 15 patients with psoriatic arthritis (PsA), and healthy skin from 12 patients with ankylosing spondylitis (AS)

66 samples


GSE277707 (heh)

Status	Public on Mar 28, 2025
Title	The malignant trajectory from the notochord to chordoma and the role of the TBXT rs2305089 germline single nucleotide polymorphism
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	The rare bone cancer chordoma and its putative precursor the Benign Notochordal Cell Tumor (BNCT) express the notochordal transcription factor, TBXT. The aim of this study was to elucidate the role of rs2305089 germline single nucleotide polymorphism in TBXT in the pathogenesis of chordoma. A radiological and genotyping association study demonstrated that BNCTs are associated with chordomas (P<0.001) and with the rs2305089 polymorphism (P=0.002). We then engineered mesoderm/notochord models which differed only by their rs2305089 genotype. Expression of TBXT was higher and the gene regulatory network activated by TBXT was enriched in heterozygous compared to wild type cell lines. Heterozygous cell lines displayed enrichment of Wnt/beta-catenin and epithelial mesenchymal transition pathways while wild type cell lines were enriched for metabolic pathways and MTORC1 signalling. Moreover, the variant A allele induced faster cell migratory capacity together with changes in the expression of endoplasmic reticulum and intracellular transport mediators, thus suggesting that rs2305089 regulates notochord vacuoles as lysosome-related organelles in the aberrant regression of the notochord. In conclusion, we show for the first time that the rs2305089 variant allele is associated not only with chordoma but also with the BNCT and provide functional evidence that this SNP regulates notochordal cell differentiation in the pathogeneis of BNCTs and chordomas.

Overall design	Six cell lines which differed only by their genotype at the rs2305089 locus were engineered using CRISPR homology directed repair: three heterozygous (labelled 463N1-3) and three wild type/unedited cell lines (labelled 4610N1-3). Induced pluripotent stem cells were differentiated into mesoderm by 48 hours and mesenchymal stem cells (MSC) by 72 hours.
36 samples


GSE271756
Status	Public on Mar 01, 2025
Title	AGAL misprocessing-induced ER stress and the unfolded protein response: lysosomal storage-independent mechanism of Fabry disease pathogenesis?
Organism	Homo sapiens
Experiment type	Expression profiling by high throughput sequencing
Summary	Classic Fabry disease (FD) is caused by GLA mutations that result in enzymatic deficiency of alpha-galactosidase A (AGAL), lysosomal storage of globotriaosylceramide, and a resulting multisystemic disease. In non-classic later-onset FD, patients have some preserved AGAL activity and a milder disease course, though female carriers may also be affected. While FD pathogenesis has been mostly attributed to catalytic deficiency of mutated AGAL, lysosomal storage and impairment of lysosomal functions, other pathogenic factors may be important, especially in non-classic later-onset FD. Clinical findings in affected males revealed a milder clinical course with ~15% residual AGAL activity and borderline plasma lyso-Gb3Cer levels. Kidney biopsies did not show lysosomal storage. Laboratory investigations documented intracellular retention of mutated AGAL with resulting endoplasmic reticulum (ER) stress and the unfolded protein response (UPR), which were alleviated with BRD4780, a small molecule clearing misfolded proteins from the early secretory compartment. We observed similar findings of ER stress and UPR with several other classic and non-classic FD missense AGAL variants. We identified defective proteostasis of mutated AGAL resulting in chronic ER stress and UPR of AGAL expressing cells (hereafter referred to as AGALopathy) as an important contributor to FD pathogenesis. These findings provide insight into non-classic later-onset FD and may better explain clinical manifestations with implications for pathogenesis, clinical characterization and treatment of all FD forms.

Overall design	We characterized the genetic, clinical, biochemical, molecular, cellular and organ pathology correlates of the p.L394P AGAL variant that was identified in six individuals with end-stage kidney disease by the Czech national screening program for FD and by further screening in an additional 24 family members. Among other material we used HEK cells stably producing either wt-AGAL_FLAG or mut-AGAL_FLAG tagged.
Preparation of cell lines stably expressing AGAL_FLAG in HEK cells: HEK 293 cells were maintained in Dulbecco's Modified Eagle Medium : Nutrient Mixture F-12 (DMEM/F-12) high glucose medium supplemented with 10% (vol/vol) fetal calf serum (Gibco), 100 U/ml penicillin G/streptomycin (Sigma, Prague, Czech Republic). HEK293 cells were transfected with 2.5 µg of plasmid DNA using Lipofectamine 3000TM (Invitrogen, Paisley, UK). Three days post transfection, cells were trypsinized, diluted and cultured in selective medium containing 0.5 mg/ml G418 (Invitrogen-Gibco, Paisley, UK). For each selected clone, the presence and correct sequence of GLA_FLAG was confirmed by Sanger sequencing; the amount of GLA_FLAG transcript was determined by real-time PCR, and expression levels of AGAL_FLAG tagged proteins were assesed by Western blot with mouse monoclonal anti-FLAG antibody (F1804, SIGMA-Aldrich, Prague, Czech Republic). Clones demonstrating similar GLA_FLAG transcript and AGAL_FLAG protein amounts were selected for further analyses.
9 samples

### MOUSE DATASETS

GSE261561
Series GSE261561		Query DataSets for GSE261561
Status	Public on Apr 19, 2024
Title	Contribution of VEGF-B-induced endocardial endothelial cell lineage in physiological versus pathological cardiac hypertrophy.
Organism	Mus musculus
Experiment type	Expression profiling by high throughput sequencing
Summary	ScRNA-seq was used to investigate the effects of autocrine versus paracrine VEGF-B signaling in the heart using transgenic mouse models. The paracrine model was further investigated in pregnancy-induced cardiac hypertrophy as well as in mice with ligation of the left anterior descending (LAD) coronary artery.

Overall design	Murine hearts were isolated and digested followed by analysis of the stromovascular fraction (SVF) by scRNA-seq, or FAC sorting of the cardiac enodthelial cells (ECs) by using FITC-tagged Pecam1 antibody and analysis by scRNA-seq.



GSE112196
Status	Public on Oct 18, 2018
Title	Chd2 is necessary for neural circuit development and long-term memory
Organism	Mus musculus
Experiment type	Expression profiling by high throughput sequencing
Summary	Considerable evidence suggests loss of function mutations in the chromatin remodeler, CHD2, contribute to a broad spectrum of human neurodevelopmental disorders. However, it is unknown how CHD2 mutations lead to impaired brain function. Here we report mice with heterozygous mutations in Chd2 exhibit deficits in neuron proliferation and a shift in neuronal excitability that included divergent changes in excitatory and inhibitory synaptic function. Further in vivo experiments show Chd2+/- mice displayed aberrant cortical rhythmogenesis and severe deficits in long-term memory, consistent with phenotypes observed in humans. We identified broad, age-dependent transcriptional changes in Chd2+/- mice, including alterations in neurogenesis, synaptic transmission and disease-related genes. Deficits in interneuron density and memory caused by Chd2+/- were reproduced by Chd2 mutation restricted to a subset of inhibitory neurons and corrected by interneuron transplantation. Our results provide initial insight into how Chd2 haploinsufficiency leads to aberrant cortical network function and impaired memory.

Overall design	We examined RNA expression levels between Chd2+/- and WT from E13.5 MGE, E13.5 cortex, and P45 Hippocampus.

GSE239863
Status	Public on Oct 01, 2023
Title	The Rogdi Knockout Mouse is a Model for Kohlschütter–Tönz Syndrome
Organism	Mus musculus
Experiment type	Expression profiling by high throughput sequencing
Summary	Kohlschütter–Tönz syndrome (KTS) is a rare autosomal recessive disorder characterized by severe intellectual disability, early-onset epileptic seizures, and amelogenesis imperfecta. Loss of ROGDI expression likely produces these severe patient defects. Here we present a novel Rogdi mutant mouse demonstrating Rogdi-/- loss of function recapitulates most KTS patient symptoms. Mutants displayed pronounced pentylenetetrazol-induced seizures confirming epilepsy susceptibility. Spontaneous locomotion and circadian activity tests demonstrate Rogdi mutant hyperactivity mirroring patient spasticity. Object recognition impairment indicates memory deficits. Rogdi-/- mutant enamel was markedly less mature. Scanning electron microscopy confirmed its hypomineralized/hypomature crystallization, as well as its low mineral content. Transcriptomic RNA sequencing of postnatal day 5 lower incisors showed downregulated enamel matrix proteins Enam, Amelx, and Ambn. Enamel crystallization is highly pH-dependent as an acidic pH is required to accelerate matrix protein degradation and to promote mineralization. Rogdi-/- teeth exhibit no signs of cyclic dental acidification. Additionally, expression changes in Wdr72, Slc9a3r2, and Atp6v0c were identified as potential contributors to these tooth acidification abnormalities. These proteins interact through the acidifying V-ATPase complex. Here we present the Rogdi-/- mutant as a novel model deciphering partially KTS pathophysiology. Rogdi-/- mutant defects in acidification might explain the unusual combination of enamel and neurological rare disease symptoms.

Overall design	To investigate the role of ROGDI, A mouse mutant model of Rogdi inactivation was generated at the Institut Clinique de la Souris (ICS), Illkirch, France. independent lower incisors of PN5 male Rogdi mutants and WT controls were analyzed by high throughput RNA sequencing (RNA-seq). Bioinformatic analysis were performed selecting significant genes using p-value < 0.05 and absolute value of log2 Fold-Change > 0.3.
