{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6a14ecbd-e4f4-4b8a-921c-398bce3e1653",
   "metadata": {},
   "source": [
    "# Kallisto quantification\n",
    "\n",
    "The purpose of this notebook is to develop a framework where I can automate the Kallisto quantification. \n",
    "\n",
    "The ground rules that I plan on setting:\n",
    "- I will use Kallisto indices which are already prepared\n",
    "- I am using human data as a test case, however I really think I can easily generalise this to other species\n",
    "- I will specify the directory where the FASTQ files are prepared."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f59a5fa6-5407-48d8-af32-f8bb2dc0f87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "import openai # Probably don't need above... but this is for testing tools with structured outputs\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "from pydantic import BaseModel, Field\n",
    "from typing import List, Dict, Literal\n",
    "import subprocess\n",
    "import glob\n",
    "import asyncio\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f1ef49e-43a5-441e-b49d-434c8892a576",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Water (Hâ‚‚O) is composed of two hydrogen atoms covalently bonded to one oxygen atom. This molecular structure gives water its unique properties, including its solvent capabilities and high surface tension.\n"
     ]
    }
   ],
   "source": [
    "# Notebookly test\n",
    "\n",
    "load_dotenv('../../.env')\n",
    "\n",
    "openai_api_key = os.getenv('OPENAI_API_KEY')\n",
    "\n",
    "# Test OpenAI API...\n",
    "\n",
    "client = OpenAI(\n",
    "  api_key=openai_api_key,  # this is also the default, it can be omitted\n",
    ")\n",
    "\n",
    "chat_completion = client.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": \"In two sentences or fewer, can you describe the molecular composition of water?\",\n",
    "        }\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    ")\n",
    "\n",
    "result = chat_completion.choices[0].message.content\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e471760d-184c-4de7-9b3a-b0ed65762ca5",
   "metadata": {},
   "source": [
    "I appreciate the formatting."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caa25f47-10e6-4e48-904e-28ff55de653f",
   "metadata": {},
   "source": [
    "# Plan of action\n",
    "\n",
    "- Specify the location of the directory containing the FASTQ files\n",
    "- Specify the location of the metadata (EDIT: I think I will develop this later - I need to be linking SRX/SRA IDs if I want this to have any chance of working)\n",
    "- Develop a prompt for Kallisto quantification\n",
    "- The outcomes should be:\n",
    "  - Kallisto code that appropriately identifies: the correct pair of FASTQ files, correct parameters, correct index selection\n",
    "  - Outputs the results in an appropriate directory\n",
    "  - Outputs the code and logic used\n",
    "  - Outputs the... output... which can be used as an evaluation tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ba65681-b93b-40f3-bd56-1ea2c53cc716",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "class KallistoCode(BaseModel):\n",
    "    command: str\n",
    "    kallisto_param_logic: str\n",
    "    other_command_logic: str\n",
    "\n",
    "# Intention is to print out logic associated with the Kallisto parameter and any other commands. In my mind, I anticipate a for loop, with the logic behind this decision belonging in other_command\n",
    "\n",
    "prompt = f\"\"\"\n",
    "\n",
    "## IDENTITY AND PURPOSE\n",
    "\n",
    "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
    "\n",
    "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
    "\n",
    "## STEPS\n",
    "\n",
    "1. Consider the list of FASTQ files. Use this to determine which files are matched with eachother - for example, the R1 and R2 files that correspond to eachother.\n",
    "2. Take into account that the path to the index file is ~/work/data/kallisto_indices/human/index.idx\n",
    "3. Take into consideration this documentation for kallisto quant:\n",
    "\n",
    "kallisto 0.50.0\n",
    "Computes equivalence classes for reads and quantifies abundances\n",
    "\n",
    "Usage: kallisto quant [arguments] FASTQ-files\n",
    "\n",
    "Required arguments:\n",
    "-i, --index=STRING            Filename for the kallisto index to be used for\n",
    "                              quantification\n",
    "-o, --output-dir=STRING       Directory to write output to\n",
    "\n",
    "Optional arguments:\n",
    "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
    "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
    "    --plaintext               Output plaintext instead of HDF5\n",
    "    --single                  Quantify single-end reads\n",
    "    --single-overhang         Include reads where unobserved rest of fragment is\n",
    "                              predicted to lie outside a transcript\n",
    "    --fr-stranded             Strand specific reads, first read forward\n",
    "    --rf-stranded             Strand specific reads, first read reverse\n",
    "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
    "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
    "                              (default: -l, -s values are estimated from paired\n",
    "                               end data, but are required when using --single)\n",
    "-t, --threads=INT             Number of threads to use (default: 1)\n",
    "\n",
    "Also note that Kallisto will not take in any of the following legacy options:\n",
    "\n",
    "Legacy options:\n",
    "\n",
    "These options are only supported in kallisto versions before 0.50.0. To use them install an older version of kallisto (recommended: 0.48.0).\n",
    "\n",
    "kallisto quant:\n",
    "\n",
    "--bias learns parameters for a model of sequences specific bias and corrects the abundances accordlingly.\n",
    "\n",
    "--fusion does normal quantification, but additionally looks for reads that do not pseudoalign because they are potentially from fusion genes. All output is written to the file fusion.txt in the output folder.\n",
    "\n",
    "--pseudobam outputs all pseudoalignments to a file pseudoalignments.bam in the output directory. This BAM file contains the pseudoalignments in BAM format, ordered by reads so that each pseudoalignment of a read is adjacent in the BAM file.\n",
    "\n",
    "--genomebam constructs the pseudoalignments to the transcriptome, but projects the transcript alignments to genome coordinates, resulting in split-read alignments. When the --genomebam option is supplied at GTF file must be given with the --gtf option. The GTF file, which can be plain text or gzipped, translates transcripts into genomic coordinates. We recommend downloading a the cdna FASTA files and GTF files from the same data source. The --chromosomes option can provide a length of the genomic chromosomes, this option is not neccessary, but gives a more consistent BAM header, some programs may require this for downstream analysis. kallisto does not require the genome sequence to do pseudoalignment, but downstream tools such as genome browsers will probably need it.\n",
    "\n",
    "4. Carefully consider what you know about using Kallisto for quantification, and provide code to quantify ALL provided FASTQ files in a way that is biologically meaningful. Take into account:\n",
    "- relevant output directories\n",
    "- that all FASTQ files should be quantified in a biologically accurate manner\n",
    "5. Provide a SINGLE command that can be used to perform quantification. This command should be something that can be copy-pasted into the command line\n",
    "6. For each parameter that is specified in the kallisto command, include a justification for why the parameter was included\n",
    "7. For any other part of the command (e.g. wrapping in a for loop), include a justification for why this is necesssary\n",
    "\n",
    "## OUTPUT\n",
    "\n",
    "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
    "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
    "2. kallisto parameter logic - for each kallisto parameter that was specified, include the reasoning for selecting the value that was used\n",
    "3. other command logic - for anything that is used to support the kallisto command (e.g. a for loop), justify why these were needed.\n",
    "\n",
    "## INPUT\n",
    "List of FASTQ files, relative to current directory: \n",
    "\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "be4bf85a-af56-4d30-8315-cd90d02a5851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "command='mkdir -p output && for R1 in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do R2=${R1/_1/_2}; if [ -f \"$R2\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o output/$(basename ${R1%_1.fastq.gz}) --bias --pseudobam --threads=4 $R1 $R2; fi; done' kallisto_param_logic=\"-i ~/work/data/kallisto_indices/human/index.idx: This specifies the path to the Kallisto index for quantification.\\\\n-o output/$(basename ${R1%_1.fastq.gz}): Creates an output directory for each sample using the basename of the R1 file minus the '_1.fastq.gz' suffix. The output structure is important for organizing results.\\\\n--bias: This flag helps to correct for sequence-specific bias, ensuring more accurate quantification.\\\\n--pseudobam: This option generates a BAM file with pseudoalignments, which is useful for further analysis and validation of read alignments in downstream tools.\\\\n--threads=4: Utilizes 4 threads to speed up the quantification process by parallelizing the computation, which is beneficial for large datasets.\" other_command_logic='mkdir -p output: This command creates an output directory to store all the results, ensuring that the workspace is organized.\\\\nfor loop with if condition: This allows us to iterate over each R1 file, find its corresponding R2 file, and run Kallisto only if both files exist, enabling a systematic processing of paired-end reads. This is necessary to ensure that no incomplete pairs are run through the analysis.'\n"
     ]
    }
   ],
   "source": [
    "chat_completion = client.beta.chat.completions.parse(\n",
    "    messages=[\n",
    "        {\"role\": \"system\", \"content\": \"You are an expert bioinformatician who specialises in generating accurate commands to achieve tasks.\"},\n",
    "        {\"role\": \"user\", \"content\": prompt}\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    "    response_format=KallistoCode\n",
    ")\n",
    "print(chat_completion.choices[0].message.parsed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1b873264-0775-4f3a-9e68-d623fc160a9d",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "import asyncio\n",
    "import subprocess\n",
    "\n",
    "async def execute_command(command, timeout=10):\n",
    "    # Create a process to run the shell command directly\n",
    "    process = await asyncio.create_subprocess_shell(\n",
    "        command,\n",
    "        stdout=asyncio.subprocess.PIPE,\n",
    "        stderr=asyncio.subprocess.PIPE,\n",
    "        shell=True\n",
    "    )\n",
    "    \n",
    "    try:\n",
    "        # Wait for initial output or timeout\n",
    "        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout)\n",
    "        stdout = stdout.decode()\n",
    "        stderr = stderr.decode()\n",
    "        return_code = process.returncode\n",
    "    except asyncio.TimeoutError:\n",
    "        # If we timeout, it means the process is still running\n",
    "        stdout = \"Process started and running in the background.\"\n",
    "        stderr = \"\"\n",
    "        return_code = \"Running\"\n",
    "    \n",
    "    return stdout, stderr, return_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7c0169ff-b41e-4d40-8fd0-506a3b654222",
   "metadata": {},
   "outputs": [],
   "source": [
    "generated_command = chat_completion.choices[0].message.parsed.command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4aa8773c-9b03-4cbd-b05c-5b73c92d7e77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mkdir -p output && for R1 in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do R2=${R1/_1/_2}; if [ -f \"$R2\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o output/$(basename ${R1%_1.fastq.gz}) --bias --pseudobam --threads=4 $R1 $R2; fi; done'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generated_command"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f34cfbd-88b3-4e23-871d-dc47eb0ce6e5",
   "metadata": {},
   "source": [
    "# \"Code sub-modules\"\n",
    "\n",
    "I will be writing individual blocks for each aspect of the code. This will involve:\n",
    "- Identifying the locations of the FASTQ files\n",
    "- (Getting the metadata and other files?)\n",
    "- (Identifying the locations of the index files?)\n",
    "- (Producing the Kallisto documentation?)\n",
    "- Generating Kallisto code\n",
    "- Executing the code\n",
    "- The self-correction mechanism\n",
    "- The \"Main\" workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "847e87ea-8add-453e-8cfc-1b1e22c56ddd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChatCompletionMessage(content=None, refusal=None, role='assistant', function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_Rr4e39ENmSdYlva2A42PCO9C', function=Function(arguments='{\"extension\":[\"kallisto\",\"idx\",\"kallisto.idx\"]}', name='GetFiles'), type='function')])"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Experiment - I will instead have a generic function for finding files, with an optional parameter to specify the file extension i.e. file type\n",
    "# I hope this might enable a more agentic workflow - i.e. \"use this function to find files\" -> filter these file names\n",
    "\n",
    "def list_files(extensions):\n",
    "    files = []\n",
    "    \n",
    "    # Loop through each extension and collect the files\n",
    "    for extension in extensions:\n",
    "        matched_files = glob.glob(os.path.join(os.path.expanduser(\"~\"), '**', f'*{extension}'), recursive=True)\n",
    "        # Filter out directories and keep only files\n",
    "        files.extend([f for f in matched_files if os.path.isfile(f)])\n",
    "    \n",
    "    # Get absolute paths and sort them\n",
    "    absolute_paths = [os.path.abspath(f) for f in files]\n",
    "    sorted_paths = sorted(absolute_paths)\n",
    "    \n",
    "    return sorted_paths\n",
    "\n",
    "# Implement the function in tools\n",
    "\n",
    "class GetFiles(BaseModel):\n",
    "    extension: list[str]\n",
    "\n",
    "tools = [openai.pydantic_function_tool(GetFiles)]\n",
    "\n",
    "chat_completion = client.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": \"Can you identify pre-built Kallisto index files that were downloaded from the official Kallisto GitHub? Take a deep breath, carefully consider what likely extensions may be. Only provide responses which are specific to Kallisto indices.\",\n",
    "        }\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    "    tools=tools\n",
    ")\n",
    "\n",
    "chat_completion.choices[0].message"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "67389b79-5a65-47b7-a4c7-f746238d29eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['kallisto', 'idx', 'kallisto.idx']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'type': 'function',\n",
       "  'function': {'name': 'GetFiles',\n",
       "   'strict': True,\n",
       "   'parameters': {'properties': {'extension': {'items': {'type': 'string'},\n",
       "      'title': 'Extension',\n",
       "      'type': 'array'}},\n",
       "    'required': ['extension'],\n",
       "    'title': 'GetFiles',\n",
       "    'type': 'object',\n",
       "    'additionalProperties': False}}}]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tool_call = chat_completion.choices[0].message.tool_calls[0]\n",
    "arguments = json.loads(tool_call.function.arguments)\n",
    "extension = arguments.get('extension')\n",
    "print(extension)\n",
    "\n",
    "tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "d5c81817-4523-46b1-9f3e-1098cb7780d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kallisto is a software tool used for quantifying transcript abundance from RNA-seq data using a pseudo-alignment approach. The file extension that is typically associated with Kallisto index files is `.idx`. \n",
      "\n",
      "From the file extensions you provided:\n",
      "- **`idx`** is the correct and commonly used extension for Kallisto index files.\n",
      "- **`kallisto.idx`** could also be a valid naming convention since it includes a specific context by prefixing with 'kallisto', but it is not the standard extension used by Kallisto itself.\n",
      "- The extension **`kallisto`** on its own is not a recognized file type for Kallisto index files.\n",
      "\n",
      "In summary, the primary file extension used for Kallisto index files is `.idx`. You might come across files named `kallisto.idx`, but `idx` is the standard.\n"
     ]
    }
   ],
   "source": [
    "prompt = f\"\"\"\n",
    "\n",
    "Are any of these file extensions: {extension} used, either commonly or rarely, for Kallisto index files, particularly those downloaded from the official Kallisto GitHub?\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "chat_completion = client.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": prompt,\n",
    "        }\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\"\n",
    ")\n",
    "\n",
    "print(chat_completion.choices[0].message.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "55071796-3369-4c81-9fa2-98b62e8aeb39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/myuser/work/data/kallisto_indices/human/index.idx']"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_files(extension)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cb89f4c-e6ef-4610-bb93-0379f62567a4",
   "metadata": {},
   "source": [
    "I have done a bit of proof-of-concept-ing, I will now test a similar principle for getting documentation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "435a188b-e4e1-4ba4-9ac6-5ce5072c730b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Using LLM to automate documentation\n",
    "\n",
    "def get_documentation(command):\n",
    "    try:\n",
    "        # Execute the kallisto command\n",
    "        result = subprocess.run(command, shell=True, capture_output=True, text=True)\n",
    "        \n",
    "        # Capture the stdout\n",
    "        stdout = result.stdout\n",
    "        \n",
    "        # Return the results\n",
    "        return stdout\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        return None\n",
    "\n",
    "docs = get_documentation(\"kallisto quant --help\")\n",
    "print(docs) # What? I swear this works by accident..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "5af133b2-81c8-42af-b0c9-0b77a40649d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ChatCompletionMessage(content=None, refusal=None, role='assistant', function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='call_iQyyX0slcWAVSPCxtNklQFp2', function=Function(arguments='{\"command\": \"kallisto quant\"}', name='GetDocumentation'), type='function'), ChatCompletionMessageToolCall(id='call_UIDE8Ybjlg2ZCrSkXGNQAPfx', function=Function(arguments='{\"command\": \"kallisto index\"}', name='GetDocumentation'), type='function')])\n",
      "kallisto quant\n",
      "kallisto 0.51.0\n",
      "Computes equivalence classes for reads and quantifies abundances\n",
      "\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "\n"
     ]
    }
   ],
   "source": [
    "class GetDocumentation(BaseModel):\n",
    "    command: str\n",
    "\n",
    "tools = [openai.pydantic_function_tool(GetDocumentation)]\n",
    "\n",
    "chat_completion = client.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": \"What command can I use to extract the documentation associated with Kallisto quantification? Think carefully about what the specific Kallisto sub-commands are\",\n",
    "        }\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    "    tools=tools\n",
    ")\n",
    "\n",
    "print(chat_completion.choices[0].message)\n",
    "\n",
    "tool_call = chat_completion.choices[0].message.tool_calls[0]\n",
    "arguments = json.loads(tool_call.function.arguments)\n",
    "command = arguments.get('command')\n",
    "print(command)\n",
    "\n",
    "docs = get_documentation(command)\n",
    "print(docs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa46c829-2e7f-44b2-82bc-81e9b48b8cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine the documentation and file finding tools into a workflow to generate a prompt\n",
    "\n",
    "# This will be:\n",
    "# Use LLMs to determine inputs for finding FASTQ files and Kallisto index files... though now that I think about, I did need an additional step to ensure I get the dataset relevant FASTQs..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "975ee585-ab75-42a3-8c73-713e1e836a61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "# Identifying the locations of FASTQ files\n",
    "\n",
    "# Since the output directory is predefined in the extraction of the FASTQ files, I can use this to specify where the output FASTQ files will be.\n",
    "\n",
    "def list_fastq_gz_files(directory):\n",
    "    # Use glob to find all files ending with .fastq.gz in the specified directory\n",
    "    files = glob.glob(os.path.join(directory, '*.fastq.gz'))\n",
    "    # Convert relative paths to absolute paths\n",
    "    absolute_paths = [os.path.abspath(f) for f in files]\n",
    "    sorted_paths = sorted(absolute_paths)\n",
    "    # Print the absolute paths\n",
    "    #for path in absolute_paths:\n",
    "    #    print(path)\n",
    "    \n",
    "    return sorted_paths\n",
    "\n",
    "# Example usage\n",
    "directory = \"/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/\"\n",
    "fastq_files = list_fastq_gz_files(directory)\n",
    "print(\"\\n\".join(fastq_files))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "17bd2f23-3ca6-4a15-b85d-f6dd9519a2d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracting the additional files?\n",
    "\n",
    "# Leave for later\n",
    "\n",
    "# Intentions are for example: linking sample metadata to the Kallisto files; identifying the correct species, i.e. human or mouse index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "be817735-7ebf-45cb-8893-4d039738c74c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identifying locations of the index files\n",
    "\n",
    "# This is hard coded, I'm not sure how I want to handle creation of Kallisto indices...\n",
    "\n",
    "# For the moment, I've manually downloaded the pre-built indices (just human at the moment)\n",
    "\n",
    "def list_kallisto_indices(directory):\n",
    "    # Use glob to find all files ending with .fastq.gz in the specified directory\n",
    "    files = glob.glob(os.path.join(directory, '**', '*.idx'), recursive=True)\n",
    "    # Convert relative paths to absolute paths\n",
    "    absolute_paths = [os.path.abspath(f) for f in files]\n",
    "    \n",
    "    # Print the absolute paths\n",
    "    #for path in absolute_paths:\n",
    "    #    print(path)\n",
    "    \n",
    "    return absolute_paths\n",
    "\n",
    "directory = os.path.expanduser(\"~\")\n",
    "kallisto_indices = list_kallisto_indices(\"~\")\n",
    "kallisto_indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "20c71c82-59b5-40e2-a9a5-0172a1a9e768",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "directory = os.path.expanduser(\"~\")\n",
    "list_kallisto_indices(\"./\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a60fe292-fa0d-4530-b781-491f3ef60575",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kallisto 0.51.0\n",
      "Computes equivalence classes for reads and quantifies abundances\n",
      "\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Producing Kallisto documentation\n",
    "\n",
    "def kallisto_documentation():\n",
    "    try:\n",
    "        # Execute the kallisto command\n",
    "        result = subprocess.run(\"kallisto quant\", shell=True, capture_output=True, text=True)\n",
    "        \n",
    "        # Capture the stdout\n",
    "        stdout = result.stdout\n",
    "        \n",
    "        # Return the results\n",
    "        return stdout\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        return None\n",
    "\n",
    "kallisto_docs = kallisto_documentation()\n",
    "print(kallisto_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94a61371-fdb2-4f17-b82e-a8ff0f576a4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## IDENTITY AND PURPOSE\n",
      "\n",
      "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
      "\n",
      "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
      "\n",
      "## STEPS\n",
      "\n",
      "1. Identify and pair the provided FASTQ files (e.g., R1 and R2 pairs) for each sample.\n",
      "2. Use the Kallisto index located here:\n",
      "/home/myuser/work/data/kallisto_indices/human/index.idx\n",
      "3. Take into consideration this documentation for kallisto quant:\n",
      "\n",
      "kallisto 0.51.0\n",
      "Computes equivalence classes for reads and quantifies abundances\n",
      "\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "\n",
      "\n",
      "4. Provide a SINGLE command that can be used to perform quantification. \n",
      "- This command will be executed asyncio.create_subprocess_shell\n",
      "- Do not include any placeholder values - ensure all values are accurate based on the information provided.\n",
      "- Input FASTQ files corresponding to each sample\n",
      "- Account for output directories and file management\n",
      "- Account for error handling and output verification\n",
      "- Ensure compatibility with POSIX-compliant shell syntax and Python's `asyncio.create_subprocess_shell`.\n",
      "- Do not assume information which is not provided\n",
      "5. Justify the inclusion of each Kallisto parameter and any addition command components (e.g. loops, mkdir)\n",
      "\n",
      "## OUTPUT\n",
      "\n",
      "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
      "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
      "2. command logic - for each kallisto parameter that was specified, include the reasoning for enabling the parameter or specifying why the parameter was used. Additionally, if any auxiliary functions are used, such as mkdir or loops, include a justification. Example format:\n",
      "- The -b 100 parameter was specified to generate 100 bootstrap samples, improving reliablility of quantification\n",
      "- The --plaintext option was specified as Kallisto was not built with HDF5\n",
      "- A for loop was constructed to iteratively perform analysis on each pair of samples\n",
      "\n",
      "## INPUT\n",
      "List of FASTQ files: \n",
      "\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
      "/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Generating Kallisto code: prompt/class\n",
    "\n",
    "class KallistoCode(BaseModel):\n",
    "    command: str\n",
    "    command_logic: str\n",
    "\n",
    "# Intention is to create a prompt that can be used to facilitate generation of a Kallisto command\n",
    "\n",
    "def CreateInitialPrompt(fastq_dir, index_dir):\n",
    "    # Step 1: Prepare the necessary inputs\n",
    "    \n",
    "    fastq_files = list_fastq_gz_files(fastq_dir)\n",
    "    formatted_fastq_files = \"\\n\".join(fastq_files)\n",
    "    \n",
    "    kallisto_indices = list_kallisto_indices(index_dir)\n",
    "    formatted_kallisto_indices = \"\\n\".join(kallisto_indices)\n",
    "    kallisto_docs = kallisto_documentation()\n",
    "\n",
    "    \n",
    "    kallisto_prompt = f\"\"\"\n",
    "\n",
    "## IDENTITY AND PURPOSE\n",
    "\n",
    "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
    "\n",
    "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
    "\n",
    "## STEPS\n",
    "\n",
    "1. Identify and pair the provided FASTQ files (e.g., R1 and R2 pairs) for each sample.\n",
    "2. Use the Kallisto index located here:\n",
    "{formatted_kallisto_indices}\n",
    "3. Take into consideration this documentation for kallisto quant:\n",
    "\n",
    "{kallisto_docs}\n",
    "\n",
    "4. Provide a SINGLE command that can be used to perform quantification. \n",
    "- This command will be executed via asyncio.create_subprocess_shell\n",
    "- Do not include any placeholder values - ensure all values are accurate based on the information provided.\n",
    "- Input FASTQ files corresponding to each sample\n",
    "- Account for output directories and file management\n",
    "- Account for error handling and output verification\n",
    "- Ensure compatibility with POSIX-compliant shell syntax and Python's `asyncio.create_subprocess_shell`.\n",
    "- Do not assume information which is not provided\n",
    "- Ensure any outputs are created in a scratch subdirectory nested within the current directory\n",
    "5. Justify the inclusion of each Kallisto parameter and any addition command components (e.g. loops, mkdir)\n",
    "\n",
    "## OUTPUT\n",
    "\n",
    "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
    "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
    "2. command logic - for each kallisto parameter that was specified, include the reasoning for enabling the parameter or specifying why the parameter was used. Additionally, if any auxiliary functions are used, such as mkdir or loops, include a justification. Example format:\n",
    "- The -b 100 parameter was specified to generate 100 bootstrap samples, improving reliablility of quantification\n",
    "- The --plaintext option was specified as Kallisto was not built with HDF5\n",
    "- A for loop was constructed to iteratively perform analysis on each pair of samples\n",
    "\n",
    "## INPUT\n",
    "List of FASTQ files: \n",
    "\n",
    "{formatted_fastq_files}\n",
    "\n",
    "\"\"\"\n",
    "    return(kallisto_prompt)\n",
    "\n",
    "temp = CreateInitialPrompt(index_dir=\"/home/myuser/work/data/kallisto_indices\",\n",
    "                           fastq_dir=\"/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/\")\n",
    "print(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ab524ed9-f65c-4bb1-bc1a-d3a48bc9b71e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Executing code\n",
    "async def execute_command(command, timeout=1000):\n",
    "    # Create a process to run the shell command directly\n",
    "    process = await asyncio.create_subprocess_shell(\n",
    "        command,\n",
    "        stdout=asyncio.subprocess.PIPE,\n",
    "        stderr=asyncio.subprocess.PIPE,\n",
    "        shell=True\n",
    "    )\n",
    "    \n",
    "    try:\n",
    "        # Wait for initial output or timeout\n",
    "        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout)\n",
    "        stdout = stdout.decode()\n",
    "        stderr = stderr.decode()\n",
    "        return_code = process.returncode\n",
    "    except asyncio.TimeoutError:\n",
    "        # If we timeout, it means the process is still running\n",
    "        stdout = \"Process started and running in the background.\"\n",
    "        stderr = \"\"\n",
    "        return_code = \"Running\"\n",
    "    \n",
    "    return stdout, stderr, return_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "93eaa461-51a7-4725-9cb8-dc05db08e45e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Self correction mechanism\n",
    "class CodeAssessment(BaseModel):\n",
    "    rewrite_needed: Literal[\"Yes\", \"No\"]\n",
    "    code_analysis: str\n",
    "    suggested_code: str\n",
    "\n",
    "async def analyze_and_correct_code(command, stdout, stderr, return_code, formatted_fastq_files, kallisto_docs):\n",
    "    # Define analysis prompt\n",
    "\n",
    "    analysis_prompt = f\"\"\"\n",
    "\n",
    "## IDENTITY AND PURPOSE\n",
    "\n",
    "You are an expert bioinformatician who specialises in RNA-seq data. You critically assess a Kallisto command provided to you and provide feedback about the validity of the code.\n",
    "\n",
    "Take a deep breath, and carefully follow the below steps to provide the best possible feedback. \n",
    "\n",
    "## STEPS\n",
    "\n",
    "Throughout these steps, do NOT make any assumptions about information that is not explicitly provided.\n",
    "\n",
    "1. Carefully digest the Kallisto command that was given to you, and consider whether the command is likely to represent a reasonable method of quantifying samples. Take into account:\n",
    "- If ALL FASTQ files were correctly analysed. Note that the FASTQ files are here:\n",
    "\n",
    "{formatted_fastq_files}\n",
    "\n",
    "- whether any additional functions, such as loops, would be correctly applied\n",
    "- whether any improvements can be made based on this documentation:\n",
    "\n",
    "{kallisto_docs}\n",
    "\n",
    "- whether the parameters included in the Kallisto command are logical and biologically sound given the input and documentation\n",
    "- whether there are any potential issues regarding validity and reliability\n",
    "\n",
    "2. Carefully digest the command's output, for example taking note of:\n",
    "- whether or not there were errors\n",
    "- whether or not there were warnings\n",
    "- check for compatibility issues with asyncio.create_subprocess_shell execution.\n",
    "- verify that all shell syntax used is POSIX-compliant and avoids bash-specific features.\n",
    "- proper escaping of special characters for both shell and Python string contexts.\n",
    "- potential issues with command length or complexity that might cause problems in subprocess execution.\n",
    "- if any part of the command might be causing the process to hang or timeout\n",
    "\n",
    "3. After considering the above, provide feedback about:\n",
    "- if the command should be rewritten. Note that:\n",
    "    - if there were ANY errors, the command should be rewritten\n",
    "    - if there were warnings, consider if the warning contributes to a biologically meaningful difference: if the warning does not affect the intended function of the comand, it can be ignored. However, if the warning indicates the command's outcome could be biologically compromised, the command should be rewritten to address the issue.\n",
    "- IF the command is to be rewritten, then take into account the stderr and stdout output, and use this to inform about\n",
    "    - the likely source of error(s) and/or warning(s)\n",
    "    - an analysis of ALL parameters being used, and whether they should be changed\n",
    "    - an analysis of whether the exclusion of any parameters might be appropriate\n",
    "    - a suggested code rewrite.\n",
    "\n",
    "## OUTPUT\n",
    "\n",
    "Your output should consist of:\n",
    "- whether or not a rewrite is necessary. Respond either \"Yes\" or \"No\"\n",
    "- if a rewrite is needed, provide a dot point list of all changes that should be made. This should be where the analysis of parameters is included.\n",
    "- a suggested code rewrite\n",
    "\n",
    "If a rewrite is not needed, include \"Not needed\" for points 2 and 3.\n",
    "\n",
    "## INPUT\n",
    "\n",
    "Input code:\n",
    "\n",
    "{command}\n",
    "\n",
    "Stdout:\n",
    "\n",
    "{stdout}\n",
    "\n",
    "Stderr:\n",
    "\n",
    "{stderr}\n",
    "\n",
    "Return code:\n",
    "\n",
    "{return_code}\n",
    "\n",
    "\"\"\"\n",
    "    \n",
    "    # Use OpenAI to analyze the error and suggest a correction\n",
    "    response = client.beta.chat.completions.parse(\n",
    "        model=\"gpt-4o-mini\",  # Adjust the model as needed\n",
    "        messages=[\n",
    "            {\"role\": \"system\", \"content\": \"You are an expert in shell scripting and command-line operations.\"},\n",
    "            {\"role\": \"user\", \"content\": analysis_prompt}\n",
    "        ],\n",
    "        response_format=CodeAssessment\n",
    "    )\n",
    "    \n",
    "    assessment = response.choices[0].message.parsed\n",
    "    return assessment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "00832689-132e-4c81-9cfc-3e87d3e6c052",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main workflow\n",
    "\n",
    "# Define Pydantic models\n",
    "class KallistoCode(BaseModel):\n",
    "    command: str\n",
    "    command_logic: str\n",
    "\n",
    "class CodeAssessment(BaseModel):\n",
    "    rewrite_needed: Literal[\"Yes\", \"No\"]\n",
    "    code_analysis: str\n",
    "    suggested_code: str\n",
    "\n",
    "async def main_workflow(fastq_dir, index_dir, max_iterations=5):\n",
    "    # Preliminary step - prepare the initial prompt\n",
    "\n",
    "    prompt = CreateInitialPrompt(index_dir = index_dir,\n",
    "                                 fastq_dir = fastq_dir)\n",
    "    \n",
    "    for iteration in range(max_iterations):\n",
    "        print(f\"\\n(Iteration {iteration + 1})\\n\\n\")\n",
    "        # Step 1: Generate the command using OpenAI API\n",
    "        chat_completion = client.beta.chat.completions.parse(\n",
    "            messages=[\n",
    "                {\"role\": \"system\", \"content\": \"You are an expert bioinformatician who specialises in generating accurate commands to achieve tasks.\"},\n",
    "                {\"role\": \"user\", \"content\": prompt}\n",
    "            ],\n",
    "            model=\"gpt-4o-mini\",\n",
    "            response_format=KallistoCode\n",
    "        )\n",
    "        # Extract the generated command\n",
    "        generated_command = chat_completion.choices[0].message.parsed.command\n",
    "        print(f\"Command justification: {chat_completion.choices[0].message.parsed.command_logic}\\n\\n\")\n",
    "        # Step 2: Execute the generated command\n",
    "        print(f\"Executing command: {generated_command}\")\n",
    "        stdout, stderr, return_code = await execute_command(generated_command)\n",
    "        # Step 3: Always analyze the command output, and proceed based on the analysis. Provide FASTQ files for checking\n",
    "        fastq_files = list_fastq_gz_files(fastq_dir)\n",
    "        formatted_fastq_files = \"\\n\".join(fastq_files)\n",
    "        kallisto_docs = kallisto_documentation()\n",
    "        assessment = await analyze_and_correct_code(generated_command, stdout, stderr, \n",
    "                                                    return_code, formatted_fastq_files, kallisto_docs)\n",
    "        print(f\"Stdout:\\n\\n{stdout}\")\n",
    "        print(f\"Stderr:\\n\\n{stderr}\")\n",
    "\n",
    "        # Step 4: Rewrite if needed, otherwise finish\n",
    "        if assessment.rewrite_needed == \"No\":\n",
    "            print(\"Command executed successfully and no rewrite is needed.\")\n",
    "            print(f\"Stdout:\\n{stdout}\")\n",
    "            break\n",
    "        else:\n",
    "            print(f\"Command needs improvement:\\n{assessment.error_warnings_source}\")\n",
    "            print(f\"Suggested corrections:\\n{assessment.suggestion}\")\n",
    "            \n",
    "            if iteration == max_iterations - 1:\n",
    "                print(\"Max iterations reached. The command might still have issues.\")\n",
    "                break\n",
    "            \n",
    "            # Update the prompt for the next iteration with suggestions\n",
    "            prompt = f\"\"\"\n",
    "            The previous command had issues when executed via Python's asyncio.create_subprocess_shell. Please generate a new command addressing these problems:\n",
    "            Previous command: {generated_command}\n",
    "            Analysis: {assessment.code_analysis}\n",
    "            Suggestions: {assessment.suggested_code}\n",
    "            \n",
    "            Original requirements:\n",
    "            {prompt}\n",
    "            \"\"\"\n",
    "\n",
    "    return generated_command, stdout, stderr, return_code, assessment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8f6cf53e-048a-462d-9988-920531166c80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "(Iteration 1)\n",
      "\n",
      "\n",
      "Command justification: - The command begins with 'mkdir -p' to create an output directory where results will be stored, ensuring that the command won't fail due to a missing output path.\n",
      "- The 'kallisto quant' command is invoked for each pair of FASTQ files, specifying the index with '-i' pointing to '/home/myuser/work/data/kallisto_indices/human/index.idx'. This is essential for Kallisto to know which index to use for mapping.\n",
      "- Each quantification invocation uses '-o' to define a unique output directory for each sample based on its identifier (e.g., SRR29101291).\n",
      "- The '--threads=4' option is utilized to enable multi-threading for faster processing, assuming a machine that supports at least 4 threads. This can significantly speed up the quantification process.\n",
      "- The command uses backslashes to create a line continuation for readability, accommodating more commands in a single execution. Each command is separated and will execute sequentially.\n",
      "- Kallisto does not require bootstrapping as no parameter is specified for it, therefore it will perform the quantifications with the default setting.\n",
      "\n",
      "\n",
      "Executing command: mkdir -p /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101291 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101292 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101293 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101294 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101295 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101296 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101297 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101298 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101299 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101300 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101301 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz && \\\n",
      "kallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101302 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
      "Stdout:\n",
      "\n",
      "\n",
      "Stderr:\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 30,364 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 209.483\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 732 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 30,242 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 200.645\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 669 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 32,224 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 201.753\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 724 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 30,720 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 212.242\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 675 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 29,669 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 203.446\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 711 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 31,540 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 214.288\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 683 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 33,865 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 212.372\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 675 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 34,001 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 210.536\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 692 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 36,915 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 204.182\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 810 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 37,198 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 211.44\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 709 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 38,817 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 203.794\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 704 rounds\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 227,665\n",
      "[index] number of k-mers: 139,900,295\n",
      "[index] number of D-list k-mers: 5,477,475\n",
      "[quant] running in paired-end mode\n",
      "[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
      "                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 80,000 reads, 38,759 reads pseudoaligned\n",
      "[quant] estimated average fragment length: 219.194\n",
      "[   em] quantifying the abundances ... done\n",
      "[   em] the Expectation-Maximization algorithm ran for 738 rounds\n",
      "\n",
      "\n",
      "Command executed successfully and no rewrite is needed.\n",
      "Stdout:\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('mkdir -p /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101291 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101292 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101293 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101294 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101295 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101296 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101297 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101298 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101299 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101300 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101301 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz && \\\\\\nkallisto quant -i /home/myuser/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/output/SRR29101302 --threads=4 /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz',\n",
       " '',\n",
       " '\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 30,364 reads pseudoaligned\\n[quant] estimated average fragment length: 209.483\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 732 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 30,242 reads pseudoaligned\\n[quant] estimated average fragment length: 200.645\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 669 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 32,224 reads pseudoaligned\\n[quant] estimated average fragment length: 201.753\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 724 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 30,720 reads pseudoaligned\\n[quant] estimated average fragment length: 212.242\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 675 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 29,669 reads pseudoaligned\\n[quant] estimated average fragment length: 203.446\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 711 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 31,540 reads pseudoaligned\\n[quant] estimated average fragment length: 214.288\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 683 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 33,865 reads pseudoaligned\\n[quant] estimated average fragment length: 212.372\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 675 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 34,001 reads pseudoaligned\\n[quant] estimated average fragment length: 210.536\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 692 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 36,915 reads pseudoaligned\\n[quant] estimated average fragment length: 204.182\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 810 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 37,198 reads pseudoaligned\\n[quant] estimated average fragment length: 211.44\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 709 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 38,817 reads pseudoaligned\\n[quant] estimated average fragment length: 203.794\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 704 rounds\\n\\n\\n[quant] fragment length distribution will be estimated from the data\\n[index] k-mer length: 31\\n[index] number of targets: 227,665\\n[index] number of k-mers: 139,900,295\\n[index] number of D-list k-mers: 5,477,475\\n[quant] running in paired-end mode\\n[quant] will process pair 1: /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\\n                             /home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\\n[quant] finding pseudoalignments for the reads ... done\\n[quant] processed 80,000 reads, 38,759 reads pseudoaligned\\n[quant] estimated average fragment length: 219.194\\n[   em] quantifying the abundances ... done\\n[   em] the Expectation-Maximization algorithm ran for 738 rounds\\n\\n',\n",
       " 0,\n",
       " CodeAssessment(rewrite_needed='No', code_analysis=\"The Kallisto command provided processes all paired-end FASTQ files accurately and employs appropriate parameters for quantification. The command uses the correct input paths for well-identified pairs of FASTQ files. The output directories are well specified, and there are no apparent errors or compatibility issues based on the presented output. Furthermore, the average fragment lengths were calculated and reported, which aligns with Kallisto's intended functionality. No biological parameters appear to be violated in this case.\", suggested_code='Not needed'))"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "await main_workflow(index_dir = \"/home/myuser/work/data/kallisto_indices/\",\n",
    "                   fastq_dir=\"/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "0b3ac786-1058-4757-adc1-663ed4d4fd25",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "## IDENTITY AND PURPOSE\n",
      "\n",
      "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
      "\n",
      "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
      "\n",
      "## STEPS\n",
      "\n",
      "1. Identify and pair the provided FASTQ files (e.g., R1 and R2 pairs) for each sample.\n",
      "2. Use the Kallisto index located at ['/home/myuser/work/data/kallisto_indices/human/index.idx'].\n",
      "3. Take into consideration this documentation for kallisto quant:\n",
      "\n",
      "kallisto 0.51.0\n",
      "Computes equivalence classes for reads and quantifies abundances\n",
      "\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "\n",
      "\n",
      "4. Provide a SINGLE command that can be used to perform quantification. \n",
      "- This command should be something that can be copy-pasted into the command line.\n",
      "- Do not include any placeholder values - ensure all values are accurate based on the information provided.\n",
      "- Input FASTQ files corresponding to each sample\n",
      "- Account for output directories and file management\n",
      "- Account for error handling and output verification\n",
      "- Ensure compatibility with POSIX-compliant shell syntax and Python's `asyncio.create_subprocess_shell`.\n",
      "- Do not assume information which is not provided\n",
      "5. Justify the inclusion of each Kallisto parameter and any addition command components (e.g. loops, mkdir)\n",
      "\n",
      "## OUTPUT\n",
      "\n",
      "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
      "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
      "2. command logic - for each kallisto parameter that was specified, include the reasoning for enabling the parameter or specifying why the parameter was used. Additionally, if any auxiliary functions are used, such as mkdir or loops, include a justification. Example format:\n",
      "- The -b 100 parameter was specified to generate 100 bootstrap samples, improving reliablility of quantification\n",
      "- The --plaintext option was specified as Kallisto was not built with HDF5\n",
      "- A for loop was constructed to iteratively perform analysis on each pair of samples\n",
      "\n",
      "## INPUT\n",
      "List of FASTQ files: \n",
      "\n",
      "['/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz', '/home/myuser/work/data/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz']\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6f3b27d-b8ff-4565-aaf4-b054e4d0cd28",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (Poetry)",
   "language": "python",
   "name": "poetry-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
