{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6a14ecbd-e4f4-4b8a-921c-398bce3e1653",
   "metadata": {},
   "source": [
    "# Kallisto quantification\n",
    "\n",
    "The purpose of this notebook is to develop a framework where I can automate the Kallisto quantification. \n",
    "\n",
    "The ground rules that I plan on setting:\n",
    "- I will use Kallisto indices which are already prepared\n",
    "- I am using human data as a test case, however I really think I can easily generalise this to other species\n",
    "- I will specify the directory where the FASTQ files are prepared."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f59a5fa6-5407-48d8-af32-f8bb2dc0f87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "from pydantic import BaseModel\n",
    "from typing import List, Dict\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f1ef49e-43a5-441e-b49d-434c8892a576",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Water is composed of two hydrogen atoms covalently bonded to one oxygen atom, represented by the chemical formula Hâ‚‚O. This molecular structure results in polar molecules, contributing to water's unique properties.\n"
     ]
    }
   ],
   "source": [
    "# Notebookly test\n",
    "\n",
    "load_dotenv('../../.env')\n",
    "\n",
    "openai_api_key = os.getenv('OPENAI_API_KEY')\n",
    "\n",
    "# Test OpenAI API...\n",
    "\n",
    "client = OpenAI(\n",
    "  api_key=openai_api_key,  # this is also the default, it can be omitted\n",
    ")\n",
    "\n",
    "chat_completion = client.chat.completions.create(\n",
    "    messages=[\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": \"In two sentences or fewer, can you describe the molecular composition of water?\",\n",
    "        }\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    ")\n",
    "\n",
    "result = chat_completion.choices[0].message.content\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e471760d-184c-4de7-9b3a-b0ed65762ca5",
   "metadata": {},
   "source": [
    "I appreciate the formatting."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caa25f47-10e6-4e48-904e-28ff55de653f",
   "metadata": {},
   "source": [
    "# Plan of action\n",
    "\n",
    "- Specify the location of the directory containing the FASTQ files\n",
    "- Specify the location of the metadata (EDIT: I think I will develop this later - I need to be linking SRX/SRA IDs if I want this to have any chance of working)\n",
    "- Develop a prompt for Kallisto quantification\n",
    "- The outcomes should be:\n",
    "  - Kallisto code that appropriately identifies: the correct pair of FASTQ files, correct parameters, correct index selection\n",
    "  - Outputs the results in an appropriate directory\n",
    "  - Outputs the code and logic used\n",
    "  - Outputs the... output... which can be used as an evaluation tool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ba65681-b93b-40f3-bd56-1ea2c53cc716",
   "metadata": {},
   "outputs": [],
   "source": [
    "class KallistoCode(BaseModel):\n",
    "    command: str\n",
    "    kallisto_param_logic: str\n",
    "    other_command_logic: str\n",
    "\n",
    "# Intention is to print out logic associated with the Kallisto parameter and any other commands. In my mind, I anticipate a for loop, with the logic behind this decision belonging in other_command\n",
    "\n",
    "prompt = f\"\"\"\n",
    "\n",
    "## IDENTITY AND PURPOSE\n",
    "\n",
    "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
    "\n",
    "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
    "\n",
    "## STEPS\n",
    "\n",
    "1. Consider the list of FASTQ files. Use this to determine which files are matched with eachother - for example, the R1 and R2 files that correspond to eachother.\n",
    "2. Take into account that the path to the index file is ~/work/data/kallisto_indices/human/index.idx\n",
    "3. Take into consideration this documentation for kallisto quant:\n",
    "\n",
    "kallisto 0.50.0\n",
    "Computes equivalence classes for reads and quantifies abundances\n",
    "\n",
    "Usage: kallisto quant [arguments] FASTQ-files\n",
    "\n",
    "Required arguments:\n",
    "-i, --index=STRING            Filename for the kallisto index to be used for\n",
    "                              quantification\n",
    "-o, --output-dir=STRING       Directory to write output to\n",
    "\n",
    "Optional arguments:\n",
    "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
    "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
    "    --plaintext               Output plaintext instead of HDF5\n",
    "    --single                  Quantify single-end reads\n",
    "    --single-overhang         Include reads where unobserved rest of fragment is\n",
    "                              predicted to lie outside a transcript\n",
    "    --fr-stranded             Strand specific reads, first read forward\n",
    "    --rf-stranded             Strand specific reads, first read reverse\n",
    "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
    "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
    "                              (default: -l, -s values are estimated from paired\n",
    "                               end data, but are required when using --single)\n",
    "-t, --threads=INT             Number of threads to use (default: 1)\n",
    "\n",
    "Also note that Kallisto will not take in any of the following legacy options:\n",
    "\n",
    "Legacy options:\n",
    "\n",
    "These options are only supported in kallisto versions before 0.50.0. To use them install an older version of kallisto (recommended: 0.48.0).\n",
    "\n",
    "kallisto quant:\n",
    "\n",
    "--bias learns parameters for a model of sequences specific bias and corrects the abundances accordlingly.\n",
    "\n",
    "--fusion does normal quantification, but additionally looks for reads that do not pseudoalign because they are potentially from fusion genes. All output is written to the file fusion.txt in the output folder.\n",
    "\n",
    "--pseudobam outputs all pseudoalignments to a file pseudoalignments.bam in the output directory. This BAM file contains the pseudoalignments in BAM format, ordered by reads so that each pseudoalignment of a read is adjacent in the BAM file.\n",
    "\n",
    "--genomebam constructs the pseudoalignments to the transcriptome, but projects the transcript alignments to genome coordinates, resulting in split-read alignments. When the --genomebam option is supplied at GTF file must be given with the --gtf option. The GTF file, which can be plain text or gzipped, translates transcripts into genomic coordinates. We recommend downloading a the cdna FASTA files and GTF files from the same data source. The --chromosomes option can provide a length of the genomic chromosomes, this option is not neccessary, but gives a more consistent BAM header, some programs may require this for downstream analysis. kallisto does not require the genome sequence to do pseudoalignment, but downstream tools such as genome browsers will probably need it.\n",
    "\n",
    "4. Carefully consider what you know about using Kallisto for quantification, and provide code to quantify ALL provided FASTQ files in a way that is biologically meaningful. Take into account:\n",
    "- relevant output directories\n",
    "- that all FASTQ files should be quantified in a biologically accurate manner\n",
    "5. Provide a SINGLE command that can be used to perform quantification. This command should be something that can be copy-pasted into the command line\n",
    "6. For each parameter that is specified in the kallisto command, include a justification for why the parameter was included\n",
    "7. For any other part of the command (e.g. wrapping in a for loop), include a justification for why this is necesssary\n",
    "\n",
    "## OUTPUT\n",
    "\n",
    "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
    "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
    "2. kallisto parameter logic - for each kallisto parameter that was specified, include the reasoning for selecting the value that was used\n",
    "3. other command logic - for anything that is used to support the kallisto command (e.g. a for loop), justify why these were needed.\n",
    "\n",
    "## INPUT\n",
    "List of FASTQ files, relative to current directory: \n",
    "\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "be4bf85a-af56-4d30-8315-cd90d02a5851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "command='mkdir -p output && for R1 in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do R2=${R1/_1/_2}; if [ -f \"$R2\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o output/$(basename ${R1%_1.fastq.gz}) --bias --pseudobam --threads=4 $R1 $R2; fi; done' kallisto_param_logic=\"-i ~/work/data/kallisto_indices/human/index.idx: This specifies the path to the Kallisto index for quantification.\\\\n-o output/$(basename ${R1%_1.fastq.gz}): Creates an output directory for each sample using the basename of the R1 file minus the '_1.fastq.gz' suffix. The output structure is important for organizing results.\\\\n--bias: This flag helps to correct for sequence-specific bias, ensuring more accurate quantification.\\\\n--pseudobam: This option generates a BAM file with pseudoalignments, which is useful for further analysis and validation of read alignments in downstream tools.\\\\n--threads=4: Utilizes 4 threads to speed up the quantification process by parallelizing the computation, which is beneficial for large datasets.\" other_command_logic='mkdir -p output: This command creates an output directory to store all the results, ensuring that the workspace is organized.\\\\nfor loop with if condition: This allows us to iterate over each R1 file, find its corresponding R2 file, and run Kallisto only if both files exist, enabling a systematic processing of paired-end reads. This is necessary to ensure that no incomplete pairs are run through the analysis.'\n"
     ]
    }
   ],
   "source": [
    "chat_completion = client.beta.chat.completions.parse(\n",
    "    messages=[\n",
    "        {\"role\": \"system\", \"content\": \"You are an expert bioinformatician who specialises in generating accurate commands to achieve tasks.\"},\n",
    "        {\"role\": \"user\", \"content\": prompt}\n",
    "    ],\n",
    "    model=\"gpt-4o-mini\",\n",
    "    response_format=KallistoCode\n",
    ")\n",
    "print(chat_completion.choices[0].message.parsed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1b873264-0775-4f3a-9e68-d623fc160a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "import subprocess\n",
    "\n",
    "async def execute_command(command, timeout=10):\n",
    "    # Create a process to run the shell command directly\n",
    "    process = await asyncio.create_subprocess_shell(\n",
    "        command,\n",
    "        stdout=asyncio.subprocess.PIPE,\n",
    "        stderr=asyncio.subprocess.PIPE,\n",
    "        shell=True\n",
    "    )\n",
    "    \n",
    "    try:\n",
    "        # Wait for initial output or timeout\n",
    "        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout)\n",
    "        stdout = stdout.decode()\n",
    "        stderr = stderr.decode()\n",
    "        return_code = process.returncode\n",
    "    except asyncio.TimeoutError:\n",
    "        # If we timeout, it means the process is still running\n",
    "        stdout = \"Process started and running in the background.\"\n",
    "        stderr = \"\"\n",
    "        return_code = \"Running\"\n",
    "    \n",
    "    return stdout, stderr, return_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7c0169ff-b41e-4d40-8fd0-506a3b654222",
   "metadata": {},
   "outputs": [],
   "source": [
    "generated_command = chat_completion.choices[0].message.parsed.command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4aa8773c-9b03-4cbd-b05c-5b73c92d7e77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mkdir -p output && for R1 in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do R2=${R1/_1/_2}; if [ -f \"$R2\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o output/$(basename ${R1%_1.fastq.gz}) --bias --pseudobam --threads=4 $R1 $R2; fi; done'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generated_command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "00832689-132e-4c81-9cfc-3e87d3e6c052",
   "metadata": {},
   "outputs": [],
   "source": [
    "class KallistoCode(BaseModel):\n",
    "    command: str\n",
    "    kallisto_param_logic: str\n",
    "    other_command_logic: str\n",
    "\n",
    "# Intention is to print out logic associated with the Kallisto parameter and any other commands. In my mind, I anticipate a for loop, with the logic behind this decision belonging in other_command\n",
    "\n",
    "prompt = f\"\"\"\n",
    "\n",
    "## IDENTITY AND PURPOSE\n",
    "\n",
    "You are an expert bioinformatician who specialises in RNA-seq data. You have great familiarity with the Kallisto tool, and use this in command line prompts to perform quantification.\n",
    "\n",
    "You will be given the paths to FASTQ files and required to construct a command line prompt that uses Kallisto to quantify these FASTQ files. Take a deep breath, and carefully consider the following steps to best achieve this goal.\n",
    "\n",
    "## STEPS\n",
    "\n",
    "1. Identify and pair the provided FASTQ files (e.g., R1 and R2 pairs). Ensure that, for any given Kallisto input, only FASTQ files corresponding to a single sample are included.\n",
    "2. Use the Kallisto index located at `~/work/data/kallisto_indices/human/index.idx`.\n",
    "3. Take into consideration this documentation for kallisto quant:\n",
    "\n",
    "kallisto 0.50.0\n",
    "Computes equivalence classes for reads and quantifies abundances\n",
    "\n",
    "Usage: kallisto quant [arguments] FASTQ-files\n",
    "\n",
    "Required arguments:\n",
    "-i, --index=STRING            Filename for the kallisto index to be used for\n",
    "                              quantification\n",
    "-o, --output-dir=STRING       Directory to write output to\n",
    "\n",
    "Optional arguments:\n",
    "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
    "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
    "    --plaintext               Output plaintext instead of HDF5\n",
    "    --single                  Quantify single-end reads\n",
    "    --single-overhang         Include reads where unobserved rest of fragment is\n",
    "                              predicted to lie outside a transcript\n",
    "    --fr-stranded             Strand specific reads, first read forward\n",
    "    --rf-stranded             Strand specific reads, first read reverse\n",
    "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
    "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
    "                              (default: -l, -s values are estimated from paired\n",
    "                               end data, but are required when using --single)\n",
    "-t, --threads=INT             Number of threads to use (default: 1)\n",
    "\n",
    "Also note that Kallisto will not take in any of the following legacy options:\n",
    "\n",
    "Legacy options:\n",
    "\n",
    "These options are only supported in kallisto versions before 0.50.0. To use them install an older version of kallisto (recommended: 0.48.0).\n",
    "\n",
    "kallisto quant:\n",
    "\n",
    "--bias learns parameters for a model of sequences specific bias and corrects the abundances accordlingly.\n",
    "\n",
    "--fusion does normal quantification, but additionally looks for reads that do not pseudoalign because they are potentially from fusion genes. All output is written to the file fusion.txt in the output folder.\n",
    "\n",
    "--pseudobam outputs all pseudoalignments to a file pseudoalignments.bam in the output directory. This BAM file contains the pseudoalignments in BAM format, ordered by reads so that each pseudoalignment of a read is adjacent in the BAM file.\n",
    "\n",
    "--genomebam constructs the pseudoalignments to the transcriptome, but projects the transcript alignments to genome coordinates, resulting in split-read alignments. When the --genomebam option is supplied at GTF file must be given with the --gtf option. The GTF file, which can be plain text or gzipped, translates transcripts into genomic coordinates. We recommend downloading a the cdna FASTA files and GTF files from the same data source. The --chromosomes option can provide a length of the genomic chromosomes, this option is not neccessary, but gives a more consistent BAM header, some programs may require this for downstream analysis. kallisto does not require the genome sequence to do pseudoalignment, but downstream tools such as genome browsers will probably need it.\n",
    "\n",
    "4. Carefully consider what you know about using Kallisto for quantification, and provide code to quantify ALL provided FASTQ files in a way that is biologically meaningful. Take into account:\n",
    "- relevant output directories\n",
    "- that all FASTQ files should be quantified in a biologically accurate manner\n",
    "5. Provide a SINGLE command that can be used to perform quantification. This command should be something that can be copy-pasted into the command line. Do not include any placeholder values - ensure all values are accurate based on the information provided.\n",
    "- you should ensure checks that any relevant files are included\n",
    "- also include checks that the output files are produced as expected\n",
    "- consider the use of for loops in generating the code\n",
    "- include robust error handling\n",
    "- include informative print messages\n",
    "- The command will be executed within a Python function using `asyncio.create_subprocess_shell`. Ensure that the command is fully compatible with execution in this context.\n",
    "- Pay attention to quoting and escaping characters to avoid issues with Python's string handling.\n",
    "- Consider any differences in behavior when the command is executed via Python versus directly in the shell.\n",
    "6. For each parameter that is specified in the kallisto command, include a justification for why the parameter was included\n",
    "7. For any other part of the command (e.g. wrapping in a for loop), include a justification for why this is necesssary\n",
    "\n",
    "## OUTPUT\n",
    "\n",
    "Your output should be structured with the following components. Do not include any information that does not belong in a component. For example, any output for \"command\" should ONLY include code, and no additional comments or commentary:\n",
    "1. command - the command that can be copy-pasted into the command line to perform the quantification for all FASTQ files\n",
    "2. kallisto parameter logic - for each kallisto parameter that was specified, include the reasoning for selecting the value that was used\n",
    "3. other command logic - for anything that is used to support the kallisto command (e.g. a for loop), justify why these were needed.\n",
    "\n",
    "## INPUT\n",
    "List of FASTQ files, relative to current directory: \n",
    "\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101300_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101296_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101298_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101302_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101297_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101295_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_1.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101294_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101292_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101291_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101299_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101301_2.fastq.gz\n",
    "/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/SRR29101293_2.fastq.gz\"\"\"\n",
    "\n",
    "async def execute_command(command, timeout=300):\n",
    "    # Create a process to run the shell command directly\n",
    "    process = await asyncio.create_subprocess_shell(\n",
    "        command,\n",
    "        stdout=asyncio.subprocess.PIPE,\n",
    "        stderr=asyncio.subprocess.PIPE,\n",
    "        shell=True\n",
    "    )\n",
    "    \n",
    "    try:\n",
    "        # Wait for initial output or timeout\n",
    "        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout)\n",
    "        stdout = stdout.decode()\n",
    "        stderr = stderr.decode()\n",
    "        return_code = process.returncode\n",
    "    except asyncio.TimeoutError:\n",
    "        # If we timeout, it means the process is still running\n",
    "        stdout = \"Process started and running in the background.\"\n",
    "        stderr = \"\"\n",
    "        return_code = \"Running\"\n",
    "    \n",
    "    return stdout, stderr, return_code\n",
    "\n",
    "async def analyze_and_correct_code(command, error_output):\n",
    "    analysis_prompt = f\"\"\"\n",
    "    Analyze the following shell command and its error output. Suggest corrections or improvements to fix the issue.\n",
    "\n",
    "    Command:\n",
    "    {command}\n",
    "\n",
    "    Error Output:\n",
    "    {error_output}\n",
    "\n",
    "    Begin by providing the input command, and stating why this did not work. Then, provide the corrected command or detailed suggestions on what changes are needed.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Use OpenAI to analyze the error and suggest a correction\n",
    "    response = client.chat.completions.create(\n",
    "        model=\"gpt-4o-mini\",  # Adjust the model as needed\n",
    "        messages=[\n",
    "            {\"role\": \"system\", \"content\": \"You are an expert in shell scripting and command-line operations.\"},\n",
    "            {\"role\": \"user\", \"content\": analysis_prompt}\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    suggestions = response.choices[0].message.content\n",
    "    return suggestions\n",
    "\n",
    "async def main_workflow(prompt, max_iterations=3):\n",
    "    for iteration in range(max_iterations):\n",
    "        # Step 1: Generate the command using OpenAI API\n",
    "        chat_completion = client.beta.chat.completions.parse(\n",
    "            messages=[\n",
    "                {\"role\": \"system\", \"content\": \"You are an expert bioinformatician who specialises in generating accurate commands to achieve tasks.\"},\n",
    "                {\"role\": \"user\", \"content\": prompt}\n",
    "            ],\n",
    "            model=\"gpt-4o-mini\",\n",
    "            response_format=KallistoCode\n",
    "        )\n",
    "\n",
    "        # Extract the generated command\n",
    "        generated_command = chat_completion.choices[0].message.parsed.command\n",
    "        \n",
    "        # Step 2: Execute the generated command\n",
    "        print(f\"Executing command (Iteration {iteration + 1}): {generated_command}\")\n",
    "        stdout, stderr, return_code = await execute_command(generated_command)\n",
    "        \n",
    "        if return_code == 0:\n",
    "            print(f\"Command executed successfully:\\n{stdout}\")\n",
    "            break\n",
    "        else:\n",
    "            print(f\"Error executing command:\\n{stderr}\\nStdout:\\n{stdout}\")\n",
    "            # Step 3: Analyze and attempt to correct the command if there is an error\n",
    "            suggestions = await analyze_and_correct_code(generated_command, stderr)\n",
    "            print(f\"Suggested corrections:\\n{suggestions}\")\n",
    "            # Update the prompt for the next iteration with suggestions or the corrected command\n",
    "            prompt = suggestions\n",
    "\n",
    "            if iteration == max_iterations - 1:\n",
    "                print(\"Max iterations reached. The command might still have issues.\")\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8f6cf53e-048a-462d-9988-920531166c80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing command (Iteration 1): for pair in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do name=$(basename \"$pair\" _1.fastq.gz); echo \"Processing sample: $name\"; if [ -f \"$pair\" ] && [ -f \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/results/kallisto_output/$name --bootstrap-samples=100 --threads=4 \"$pair\" \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\"; else echo \"FASTQ pairs for sample $name are missing!\"; fi; done\n",
      "Error executing command:\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101291\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101292\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101293\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101294\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101295\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101296\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101297\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101298\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101299\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101300\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101301\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "[quant] fragment length distribution will be estimated from the data\n",
      "Error: could not create directory /home/myuser/work/results/kallisto_output/SRR29101302\n",
      "Warning: kallisto was not compiled with HDF5 support so no bootstrapping\n",
      "will be performed. Run quant with --plaintext option or recompile with\n",
      "HDF5 support to obtain bootstrap estimates.\n",
      "\n",
      "\n",
      "Stdout:\n",
      "Processing sample: SRR29101291\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101292\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101293\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101294\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101295\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101296\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101297\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101298\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101299\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101300\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101301\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "Processing sample: SRR29101302\n",
      "Usage: kallisto quant [arguments] FASTQ-files\n",
      "\n",
      "Required arguments:\n",
      "-i, --index=STRING            Filename for the kallisto index to be used for\n",
      "                              quantification\n",
      "-o, --output-dir=STRING       Directory to write output to\n",
      "\n",
      "Optional arguments:\n",
      "-b, --bootstrap-samples=INT   Number of bootstrap samples (default: 0)\n",
      "    --seed=INT                Seed for the bootstrap sampling (default: 42)\n",
      "    --plaintext               Output plaintext instead of HDF5\n",
      "    --single                  Quantify single-end reads\n",
      "    --single-overhang         Include reads where unobserved rest of fragment is\n",
      "                              predicted to lie outside a transcript\n",
      "    --fr-stranded             Strand specific reads, first read forward\n",
      "    --rf-stranded             Strand specific reads, first read reverse\n",
      "-l, --fragment-length=DOUBLE  Estimated average fragment length\n",
      "-s, --sd=DOUBLE               Estimated standard deviation of fragment length\n",
      "                              (default: -l, -s values are estimated from paired\n",
      "                               end data, but are required when using --single)\n",
      "-p, --priors                  Priors for the EM algorithm, either as raw counts or as\n",
      "                              probabilities. Pseudocounts are added to raw reads to\n",
      "                              prevent zero valued priors. Supplied in the same order\n",
      "                              as the transcripts in the transcriptome\n",
      "-t, --threads=INT             Number of threads to use (default: 1)\n",
      "    --verbose                 Print out progress information every 1M proccessed reads\n",
      "\n",
      "Suggested corrections:\n",
      "The input shell command is as follows:\n",
      "\n",
      "```bash\n",
      "for pair in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do \n",
      "    name=$(basename \"$pair\" _1.fastq.gz); \n",
      "    echo \"Processing sample: $name\"; \n",
      "    \n",
      "    if [ -f \"$pair\" ] && [ -f \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\" ]; then \n",
      "        kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o /home/myuser/work/results/kallisto_output/$name --bootstrap-samples=100 --threads=4 \"$pair\" \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\"; \n",
      "    else \n",
      "        echo \"FASTQ pairs for sample $name are missing!\"; \n",
      "    fi; \n",
      "done\n",
      "```\n",
      "\n",
      "### Issue Identified\n",
      "\n",
      "The command did not work primarily due to the following reasons, as indicated in the error output:\n",
      "\n",
      "1. **Directory Creation Failure**: The error `could not create directory /home/myuser/work/results/kallisto_output/SRR29101291` indicates that the `kallisto` command was unable to create the output directory specified in the `-o` option. This can occur if the parent directory does not exist or if there are permissions issues.\n",
      "\n",
      "2. **Lack of HDF5 Support**: The warning `kallisto was not compiled with HDF5 support` indicates that the version of `kallisto` being used does not support HDF5 for bootstrapping. This is not a critical error but suggests that bootstrapping won't be performed unless the user recompiles `kallisto` with HDF5 support.\n",
      "\n",
      "### Corrections and Improvements\n",
      "\n",
      "To fix the issue, we need to ensure that the output directory exists before attempting to run `kallisto`. Additionally, we can make optional adjustments regarding the HDF5 warning based on the user's preferences and needs. Here are the steps for corrections:\n",
      "\n",
      "1. **Create Output Directory**: Ensure that the necessary output directory exists before calling `kallisto`. We can use `mkdir -p` to create the output directory if it does not exist.\n",
      "\n",
      "2. **Handle HDF5 Warning**: If bootstrapping is required and `kallisto` was not compiled with HDF5 support, consider recompiling it or running with the `--plaintext` option as indicated in the warning message.\n",
      "\n",
      "### Corrected Command\n",
      "\n",
      "Here is the updated command reflecting these changes:\n",
      "\n",
      "```bash\n",
      "for pair in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do \n",
      "    name=$(basename \"$pair\" _1.fastq.gz); \n",
      "    echo \"Processing sample: $name\"; \n",
      "    \n",
      "    # Define the output directory\n",
      "    output_dir=\"/home/myuser/work/results/kallisto_output/$name\"\n",
      "    \n",
      "    # Create the output directory if it doesn't exist\n",
      "    mkdir -p \"$output_dir\"\n",
      "    \n",
      "    if [ -f \"$pair\" ] && [ -f \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\" ]; then \n",
      "        # Use --plaintext if HDF5 support is not enabled\n",
      "        kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o \"$output_dir\" --bootstrap-samples=100 --threads=4 --plaintext \"$pair\" \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\"; \n",
      "    else \n",
      "        echo \"FASTQ pairs for sample $name are missing!\"; \n",
      "    fi; \n",
      "done\n",
      "```\n",
      "\n",
      "### Summary of Changes\n",
      "- Added `mkdir -p \"$output_dir\"` before calling the `kallisto` command to ensure the output directory exists.\n",
      "- Added the `--plaintext` option to the `kallisto` command to address the warning about HDF5 support (assuming it's acceptable for the user's requirements). If bootstrapping is essential, the user should consider recompiling `kallisto` with HDF5 support.\n",
      "\n",
      "This revised command should help resolve the directory creation issues and handle the HDF5 warning accordingly.\n",
      "Executing command (Iteration 2): for pair in /home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/*_1.fastq.gz; do name=$(basename \"$pair\" _1.fastq.gz); echo \"Processing sample: $name\"; output_dir=\"/home/myuser/work/results/kallisto_output/$name\"; mkdir -p \"$output_dir\"; if [ -f \"$pair\" ] && [ -f \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\" ]; then kallisto quant -i ~/work/data/kallisto_indices/human/index.idx -o \"$output_dir\" --bootstrap-samples=100 --threads=4 --plaintext \"$pair\" \"/home/myuser/work/results/2_extract_data/scratch/FASTQs_GSE268034/${name}_2.fastq.gz\"; else echo \"FASTQ pairs for sample $name are missing!\"; fi; done\n",
      "Command executed successfully:\n",
      "Processing sample: SRR29101291\n",
      "Processing sample: SRR29101292\n",
      "Processing sample: SRR29101293\n",
      "Processing sample: SRR29101294\n",
      "Processing sample: SRR29101295\n",
      "Processing sample: SRR29101296\n",
      "Processing sample: SRR29101297\n",
      "Processing sample: SRR29101298\n",
      "Processing sample: SRR29101299\n",
      "Processing sample: SRR29101300\n",
      "Processing sample: SRR29101301\n",
      "Processing sample: SRR29101302\n",
      "\n"
     ]
    }
   ],
   "source": [
    "await main_workflow(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e606c704-0f1d-42b0-b95d-cdda95e3eaf1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (Poetry)",
   "language": "python",
   "name": "poetry-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
